<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2025-07-03 Thu 11:25 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>&lrm;</title>
<meta name="generator" content="Org Mode" />
<style>
  #content { max-width: 60em; margin: auto; }
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #e6e6e6;
    border-radius: 3px;
    background-color: #f2f2f2;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: auto;
  }
  pre.src:before {
    display: none;
    position: absolute;
    top: -8px;
    right: 12px;
    padding: 3px;
    color: #555;
    background-color: #f2f2f299;
  }
  pre.src:hover:before { display: inline; margin-top: 14px;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-authinfo::before { content: 'Authinfo'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .equation-container {
    display: table;
    text-align: center;
    width: 100%;
  }
  .equation {
    vertical-align: middle;
  }
  .equation-label {
    display: table-cell;
    text-align: right;
    vertical-align: middle;
  }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { }
</style>
<style type="text/css">
    <!--
      body {
        color: #000000;
        background-color: #ffffff;
      }
      .org-bold {
        /* bold */
        font-weight: bold;
      }
      .org-builtin {
        /* font-lock-builtin-face */
        color: #8f0075;
      }
      .org-comment {
        /* font-lock-comment-face */
        color: #505050;
      }
      .org-comment-delimiter {
        /* font-lock-comment-delimiter-face */
        color: #505050;
      }
      .org-constant {
        /* font-lock-constant-face */
        color: #0000c0;
      }
      .org-default {
        /* default */
        color: #000000;
        background-color: #ffffff;
      }
      .org-diff-added {
        /* diff-added */
        color: #002c00;
        background-color: #bbeabb;
      }
      .org-diff-context {
        /* diff-context */
        color: #56576d;
      }
      .org-diff-file-header {
        /* diff-file-header */
        color: #000000;
        font-weight: bold;
      }
      .org-diff-header {
        /* diff-header */
        color: #000000;
      }
      .org-diff-hunk-header {
        /* diff-hunk-header */
        color: #041645;
        background-color: #b7cfe0;
        font-weight: bold;
      }
      .org-diff-indicator-added {
        /* diff-indicator-added */
        color: #005e00;
        background-color: #bbeabb;
        font-weight: bold;
      }
      .org-diff-indicator-removed {
        /* diff-indicator-removed */
        color: #a60000;
        background-color: #efcbcf;
        font-weight: bold;
      }
      .org-diff-removed {
        /* diff-removed */
        color: #4a0000;
        background-color: #efcbcf;
      }
      .org-doc {
        /* font-lock-doc-face */
        color: #2a486a;
      }
      .org-ess-assignment {
        /* ess-assignment-face */
        color: #0000c0;
      }
      .org-ess-constant {
        /* ess-constant-face */
        color: #005a5f;
      }
      .org-ess-keyword {
        /* ess-keyword-face */
        color: #5317ac;
      }
      .org-ess-modifiers {
        /* ess-modifiers-face */
        color: #0000c0;
      }
      .org-function-name {
        /* font-lock-function-name-face */
        color: #721045;
      }
      .org-keyword {
        /* font-lock-keyword-face */
        color: #5317ac;
      }
      .org-nxml-attribute-local-name {
        /* nxml-attribute-local-name */
        color: #00538b;
      }
      .org-nxml-element-local-name {
        /* nxml-element-local-name */
        color: #721045;
      }
      .org-nxml-processing-instruction-content {
        /* nxml-processing-instruction-content */
        color: #093060;
      }
      .org-nxml-processing-instruction-delimiter {
        /* nxml-processing-instruction-delimiter */
        color: #282828;
      }
      .org-nxml-processing-instruction-target {
        /* nxml-processing-instruction-target */
        color: #5317ac;
      }
      .org-nxml-tag-delimiter {
        /* nxml-tag-delimiter */
        color: #282828;
      }
      .org-nxml-tag-slash {
        /* nxml-tag-slash */
        color: #282828;
      }
      .org-nxml-text {
      }
      .org-org-block {
        /* org-block */
        color: #000000;
      }
      .org-org-block-begin-line {
        /* org-block-begin-line */
        color: #505050;
        background-color: #f0f0f0;
      }
      .org-org-block-end-line {
        /* org-block-end-line */
        color: #505050;
        background-color: #f0f0f0;
      }
      .org-org-checkbox-statistics-done {
        /* org-checkbox-statistics-done */
        color: #005e00;
      }
      .org-org-code {
        /* org-code */
        color: #005a5f;
      }
      .org-org-date {
        /* org-date */
        color: #00538b;
        text-decoration: underline;
      }
      .org-org-done {
        /* org-done */
        color: #005e00;
      }
      .org-org-drawer {
        /* org-drawer */
        color: #505050;
      }
      .org-org-headline-done {
        /* org-headline-done */
        color: #004000;
      }
      .org-org-level-1 {
        /* org-level-1 */
        color: #000000;
        font-weight: bold;
      }
      .org-org-level-2 {
        /* org-level-2 */
        color: #5d3026;
        font-weight: bold;
      }
      .org-org-level-3 {
        /* org-level-3 */
        color: #093060;
        font-weight: bold;
      }
      .org-org-level-4 {
        /* org-level-4 */
        color: #184034;
        font-weight: bold;
      }
      .org-org-link {
        /* org-link */
        color: #0000c0;
        text-decoration: underline;
      }
      .org-org-meta-line {
        /* org-meta-line */
        color: #505050;
      }
      .org-org-property-value {
        /* org-property-value */
        color: #093060;
      }
      .org-org-special-keyword {
        /* org-special-keyword */
        color: #505050;
      }
      .org-org-table {
        /* org-table */
        color: #093060;
      }
      .org-org-tag {
        /* org-tag */
        color: #541f4f;
      }
      .org-org-verbatim {
        /* org-verbatim */
        color: #8f0075;
      }
      .org-sh-quoted-exec {
        /* sh-quoted-exec */
        color: #8f0075;
      }
      .org-show-paren-match {
        /* show-paren-match */
        color: #000000;
        background-color: #e0af82;
      }
      .org-string {
        /* font-lock-string-face */
        color: #2544bb;
      }
      .org-type {
        /* font-lock-type-face */
        color: #005a5f;
      }
      .org-underline {
        /* underline */
        text-decoration: underline;
      }
      .org-variable-name {
        /* font-lock-variable-name-face */
        color: #00538b;
      }
      .org-warning {
        /* font-lock-warning-face */
        color: #813e00;
      }

      a {
        color: inherit;
        background-color: inherit;
        font: inherit;
        text-decoration: inherit;
      }
      a:hover {
        text-decoration: underline;
      }
    -->
    </style>
</head>
<body>
<div id="content" class="content">
<div id="table-of-contents" role="doc-toc">
<h2>Table of Contents</h2>
<div id="text-table-of-contents" role="doc-toc">
<ul>
<li><a href="#QUICK_START">1. Quick Start</a>
<ul>
<li><a href="#org6176b4b">1.1. Tutorial</a>
<ul>
<li><a href="#org9fa9cf3">1.1.1. Cloning the directory</a></li>
<li><a href="#org2a66fde">1.1.2. Example Data</a></li>
<li><a href="#org2833113">1.1.3. Convert RAW to mzml</a></li>
<li><a href="#orgf06df2f">1.1.4. Setting up your experiments config file</a></li>
<li><a href="#org1fb691e">1.1.5. Calling gladiator</a></li>
</ul>
</li>
<li><a href="#org8feba18">1.2. Results</a></li>
</ul>
</li>
<li><a href="#org59afd13">2. About this document</a>
<ul>
<li><a href="#org5e27803">2.1. Writing Style</a></li>
<li><a href="#orgb4fba17">2.2. Trouble Shooting</a></li>
<li><a href="#org917fb29">2.3. Tangling</a></li>
<li><a href="#org65077e8">2.4. Building the containers</a></li>
<li><a href="#org5159ba9">2.5. License</a></li>
</ul>
</li>
<li><a href="#orga714a85">3. Overview of this pipeline</a></li>
<li><a href="#orgd37ea86">4. Preprocessing Data</a>
<ul>
<li><a href="#orgc4f8110">4.1. Picking Peaks</a></li>
<li><a href="#org4a1e12d">4.2. Converting Dia Raw with Msconvert</a></li>
</ul>
</li>
<li><a href="#org90732e0">5. Analysis <code>[4/4]</code></a>
<ul>
<li><a href="#org1e918b2">5.1. Headers</a></li>
<li><a href="#org1ed1558">5.2. building database</a>
<ul>
<li><a href="#orgc0e70a2">5.2.1. Combining Fasta Files</a></li>
<li><a href="#org6ddc65b">5.2.2. Adding Decoys</a></li>
</ul>
</li>
<li><a href="#org2d0e377">5.3. Input DIA files</a></li>
<li><a href="#org47af1f4">5.4. Creating Swath window files</a>
<ul>
<li><a href="#org7d35826">5.4.1. Branching if user supplied windows</a></li>
<li><a href="#org894d3d1">5.4.2. Making truncated-swath-windows and swath-windows from user-supplied swath-windows</a></li>
<li><a href="#org3facb06">5.4.3. Inferring Windows from mzml files</a></li>
<li><a href="#orgb22c933">5.4.4. Making the non-overlapping swath-windows</a></li>
</ul>
</li>
<li><a href="#orgc4069b5">5.5. Library Generation</a>
<ul>
<li><a href="#org392471a">5.5.1. Building {Pseudo-,}Spectral library from (Pseudo)-Spectra  <code>[5/5]</code></a>
<ul>
<li><a href="#org6b7746b">5.5.1.1. Using DDA Data</a></li>
<li><a href="#orgc88f836">5.5.1.2. Creating Pseudospectra with DIAumpire</a>
<ul>
<li><a href="#org0fd7577">5.5.1.2.1. Problems you might encounter during this step</a>
<ul>
<li><a href="#orgf53a0bd">5.5.1.2.1.1. Out of Memory in Dia-umpire</a></li>
<li><a href="#orga5f56ad">5.5.1.2.1.2. <code>MzmlToMzxml</code> processing error.</a></li>
</ul>
</li>
<li><a href="#org63b988a">5.5.1.2.2. Steps that are run</a></li>
</ul>
</li>
<li><a href="#org50cf2ec">5.5.1.3. Creating Pseudospectra with diams2pep</a></li>
<li><a href="#org29c9d7c">5.5.1.4. Choosing the MS Sequence database search engine: Comet/Xtandem</a></li>
<li><a href="#orga6e8b6b">5.5.1.5. Comet</a></li>
<li><a href="#org6868709">5.5.1.6. Xtandem</a></li>
<li><a href="#org1a37a38">5.5.1.7. Joining Comet &amp; Xtandem</a></li>
<li><a href="#org7a590b7">5.5.1.8. Building Specral Library</a>
<ul>
<li><a href="#orgff0099e">5.5.1.8.1. Mayu</a></li>
<li><a href="#org57b626e">5.5.1.8.2. Spectrast</a>
<ul>
<li><a href="#org91d58b9">5.5.1.8.2.1. Converting traml into spectrast friendly format</a></li>
<li><a href="#org109f2bb">5.5.1.8.2.2. Running Spectrast</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="#section-deepdia">5.5.2. Building Spectral library from Machine learning</a></li>
<li><a href="#org0d8d164">5.5.3. Supplying a custom spectral library</a></li>
<li><a href="#org86499fa">5.5.4. combining various spectral libraries into one.</a></li>
</ul>
</li>
<li><a href="#org615da83">5.6. OpenSwathDecoys</a></li>
<li><a href="#orgfb5b55c">5.7. Building Dia Matrix</a>
<ul>
<li><a href="#org84e26d6">5.7.1. OpenSwathWorkflow</a>
<ul>
<li><a href="#orge7e2f99">5.7.1.1. What is the relation between <code>irt.txt</code> and <code>iRTAssayLibrary.TraML</code></a></li>
</ul>
</li>
<li><a href="#orgc302ff7">5.7.2. Pyprophet</a>
<ul>
<li><a href="#org2c0b841">5.7.2.1. Legacy Pyprophet</a></li>
<li><a href="#org2a0458e">5.7.2.2. nonlegacy prypophet</a>
<ul>
<li><a href="#orgeb39ba4">5.7.2.2.1. KeyError: 'HOME' in <code>pyprophet_subsample</code></a></li>
</ul>
</li>
<li><a href="#org2c0159d">5.7.2.3. Choosing between legacy and nonlegacy pyprophet</a></li>
</ul>
</li>
<li><a href="#orgf718b08">5.7.3. feature-alignment</a></li>
<li><a href="#org105e35f">5.7.4. Swath2stats</a>
<ul>
<li><a href="#org5e1c307">5.7.4.1. Dependencies</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="#org8a7d27f">6. Configuration for backends</a>
<ul>
<li><a href="#org97679a9">6.1. Defining the config template</a></li>
<li><a href="#org0d0a1bc">6.2. Per backend config files</a></li>
<li><a href="#orgedbdfbf">6.3. Backend Specific Issues</a>
<ul>
<li><a href="#org928fcbb">6.3.1. Podman specific Issues</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#orgf5bf9a7">7. SDRF support</a>
<ul>
<li><a href="#org602b580">7.1. <code>comment[file uri]</code></a></li>
<li><a href="#orge5528da">7.2. <code>comment[precursor mass tolerance]</code>, <code>comment[fragment mass tolerance]</code></a></li>
<li><a href="#org1af7273">7.3. Utility functions</a></li>
<li><a href="#org9fd977a">7.4. Testing</a></li>
<li><a href="#org44cd314">7.5. Updating params</a></li>
</ul>
</li>
<li><a href="#org8de5dc7">8. Putting it together</a>
<ul>
<li><a href="#org6d834a7">8.1. gwl</a></li>
<li><a href="#org2a47adf">8.2. nf</a></li>
</ul>
</li>
<li><a href="#orgac9c4be">9. File local variables</a></li>
<li><a href="#org84da839">10. Note</a></li>
<li><a href="#org186f487">11. Launching</a></li>
<li><a href="#org99e0bbc">12. Troubleshooting</a></li>
</ul>
</div>
</div>


<div id="outline-container-QUICK_START" class="outline-2">
<h2 id="QUICK_START"><span class="section-number-2">1.</span> Quick Start</h2>
<div class="outline-text-2" id="text-QUICK_START">
<p>
This shows section shows you how to quickly get started with gladiator-nf.
</p>
</div>
<div id="outline-container-org6176b4b" class="outline-3">
<h3 id="org6176b4b"><span class="section-number-3">1.1.</span> Tutorial</h3>
<div class="outline-text-3" id="text-1-1">
</div>
<div id="outline-container-org9fa9cf3" class="outline-4">
<h4 id="org9fa9cf3"><span class="section-number-4">1.1.1.</span> Cloning the directory</h4>
<div class="outline-text-4" id="text-1-1-1">
<p>
If you are using the development version of this, you will wnat to clone this directory, and 'tangle' (create the scripts).
This will require 'make', emacs 27 or greater, and the coreutils
</p>
<div class="org-src-container">
<pre class="src src-sh">git clone https://github.com/elolab/gladiator-nf
<span class="org-builtin">cd</span> gladiator-nf 
yes <span class="org-string">'//'</span> | make tangle
</pre>
</div>
</div>
</div>

<div id="outline-container-org2a66fde" class="outline-4">
<h4 id="org2a66fde"><span class="section-number-4">1.1.2.</span> Example Data</h4>
<div class="outline-text-4" id="text-1-1-2">
<p>
For this example we will use MSV000090837 from , just 210820_Grad090_LFQ_A_01.raw and 210820_Grad090_LFQ_B_01.raw, so that you dont need to use that much disk space for trying this out.
While the massive repo provides mzml files, which you could use, we are gonna start from the raw for educational purposes.
Lets say
Download the data
</p>
<div class="org-src-container">
<pre class="src src-sh">mkdir RAW
wget --no-directories  --directory-prefix=RAW ftp://massive-ftp.ucsd.edu/v05/MSV000090837/raw/Exp03_repeat_90minGradient_from_Exp02_same_Mastermix/210820_Grad090_LFQ_{A,B}_01.raw 
</pre>
</div>


<div class="org-src-container">
<pre class="src src-sh">mkdir fasta
wget --directory-prefix=fasta  <span class="org-string">'ftp://massive-ftp.ucsd.edu:/v05/MSV000090837/sequence/fasta/*.fasta'</span> 
</pre>
</div>
</div>
</div>

<div id="outline-container-org2833113" class="outline-4">
<h4 id="org2833113"><span class="section-number-4">1.1.3.</span> Convert RAW to mzml</h4>
<div class="outline-text-4" id="text-1-1-3">
<p>
We will want to convert the raw files to mzml, so we will want to run the following command in the container 
</p>

<div class="org-src-container">
<pre class="src src-sh"><span class="org-variable-name">COMMAND</span>=<span class="org-string">'find . -iname '"'"'*.raw'"'"' -print0 | xargs -P5 -0 -i wine msconvert {} --filter '"'"'titleMaker &lt;RunId&gt;.&lt;ScanNumber&gt;.&lt;ScanNumber&gt;.&lt;ChargeState&gt; File:"&lt;SourcePath&gt;", NativeID:"&lt;Id&gt;"'"'"' -o MZML/'</span>
<span class="org-variable-name">IMAGEID</span>=$(podman pull docker://chambm/pwiz-skyline-i-agree-to-the-vendor-licenses:3.0.21354-9ee14c7)
<span class="org-builtin">echo</span>  $<span class="org-variable-name">COMMAND</span> &gt; ./convert.sh
chmod +x convert.sh
podman run -it  -v $<span class="org-variable-name">PWD</span>:$<span class="org-variable-name">PWD</span> -w $<span class="org-variable-name">PWD</span> $<span class="org-variable-name">IMAGEID</span> /bin/bash  ./convert.sh
</pre>
</div>
</div>
</div>

<div id="outline-container-orgf06df2f" class="outline-4">
<h4 id="orgf06df2f"><span class="section-number-4">1.1.4.</span> Setting up your experiments config file</h4>
<div class="outline-text-4" id="text-1-1-4">
<p>
Create a config file with the following contents, 
adjusting as needed.
Lets name it <code>myconfig.nf</code>
</p>
<div class="org-src-container">
<pre class="src src-nextflow">params {
    <span class="org-comment-delimiter">// </span><span class="org-comment">this is fragment mass tolerance in Dalton, 0.02 is a sensible default</span>
    fragment_mass_tolerance=0.02
    <span class="org-comment-delimiter">// </span><span class="org-comment">This is in parts per million, ppm</span>
    precursor_mass_tolerance=10
    <span class="org-comment-delimiter">// </span><span class="org-comment">this is cutoffrate used by mayu for finding the peptide probability</span>
    protFDR=0.01
    <span class="org-comment-delimiter">// </span><span class="org-comment">retention time information, this traml is a good default.</span>
    irt_traml_file=<span class="org-string">'ftp://PASS00289:XY4524h@ftp.peptideatlas.org:/SGS/assays/OpenSWATH_SM4_iRT_AssayLibrary.TraML'</span>
    use_irt=<span class="org-constant">false</span>
    max_missed_cleavages=1 
    libgen_method=<span class="org-string">'diaumpire'</span>
    pyprophet_use_legacy=<span class="org-constant">false</span>
    pyprophet_fixed_seed=<span class="org-constant">false</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">this would make your subsample ratio 1 / your number of samples</span>
    pyprophet_subsample_ratio=<span class="org-constant">null</span>
}
</pre>
</div>
</div>
</div>

<div id="outline-container-org1fb691e" class="outline-4">
<h4 id="org1fb691e"><span class="section-number-4">1.1.5.</span> Calling gladiator</h4>
<div class="outline-text-4" id="text-1-1-5">
<p>
Note that you will have to quote wildcards, because nextflow's java expects to handle the wildcard expansion,rather than your shell handling the expandsion
</p>

<div class="org-src-container">
<pre class="src src-sh"><span class="org-comment-delimiter"># </span><span class="org-comment">if you are using podman </span>
<span class="org-variable-name">NXF_VER</span>=22.10.1 nextflow -c myconfig.nf <span class="org-sh-escaped-newline">\</span>
         -c config/podman.nf <span class="org-sh-escaped-newline">\</span>
         run gladiator.nf <span class="org-sh-escaped-newline">\</span>
         --fastafiles=<span class="org-string">'fasta/*.fasta'</span> <span class="org-sh-escaped-newline">\</span>
         --diafiles=<span class="org-string">'MZML/*.mzML'</span> 
</pre>
</div>

<p>
if you are using singularity, use <code>config/singularity.nf</code> in lieu of <code>config/podman.nf</code>.
if you have the guix package manager, you can use <code>config/guix.nf</code> .
</p>

<p>
You will want to have java / openjdk 17, 18, or 19.
Java 20 and later will give you a groovy error.
</p>
</div>
</div>
</div>

<div id="outline-container-org8feba18" class="outline-3">
<h3 id="org8feba18"><span class="section-number-3">1.2.</span> Results</h3>
<div class="outline-text-3" id="text-1-2">
<p>
In the results directory , which you can specify with <code>--outdir=/path/to/results</code>, which  defaults to <code>./results</code>,
you will find two files
</p>
<ul class="org-ul">
<li><code>dia/DIA-peptide-matrix.tsv</code>
This contains the intensitities on the peptide level,
the first column is the peptide identifier, and the the other columns are the intensities in that sample.
For example:</li>
</ul>
<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-right" />

<col  class="org-right" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">"ProteinName_FullPeptideName"</th>
<th scope="col" class="org-right">"210820_Grad090_LFQ_A_01.mzML"</th>
<th scope="col" class="org-right">"210820_Grad090_LFQ_B_01.mzML"</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left">"1/A2I7N3_LAVSHVIHK"</td>
<td class="org-right">624895</td>
<td class="org-right">723064</td>
</tr>

<tr>
<td class="org-left">"3/sp¦P08729¦K2C7_HUMAN/Q3KNV1/P08729_VDALNDEINFLR"</td>
<td class="org-right">2037142</td>
<td class="org-right">1283280</td>
</tr>
</tbody>
</table>

<ul class="org-ul">
<li><code>dia/DIA-protein-matrix.tsv</code>
Here the intensities are on the protein group level, with the first column containing the protein group name and size.</li>
</ul>
<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-right" />

<col  class="org-right" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">ProteinName</th>
<th scope="col" class="org-right">210820_Grad090_LFQ_A_01.mzML</th>
<th scope="col" class="org-right">210820_Grad090_LFQ_B_01.mzML</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left">1/sp¦O14561¦ACPM_HUMAN</td>
<td class="org-right">1410100</td>
<td class="org-right">1359820</td>
</tr>

<tr>
<td class="org-left">2/sp¦P0CX49¦RL18A_YEAST/sp¦P0CX50¦RL18B_YEAST</td>
<td class="org-right">99249410</td>
<td class="org-right">44970930</td>
</tr>
</tbody>
</table>
<p>
(read <code>|</code> instead of  <code>¦</code> )
</p>
</div>
</div>
</div>



<div id="outline-container-org59afd13" class="outline-2">
<h2 id="org59afd13"><span class="section-number-2">2.</span> About this document</h2>
<div class="outline-text-2" id="text-2">
<p>
This is a literate programming (org-mode) document that describes the
the nextflow implemantiation of gladiator (<a href="https://github.com/elolab/glaDIAtor">https://github.com/elolab/glaDIAtor</a>)
and all needed template files.
This directory should already contain the tangled files.
To learn more about org-mode, see <a href="https://orgmode.org/">https://orgmode.org/</a>.
</p>
</div>

<div id="outline-container-org5e27803" class="outline-3">
<h3 id="org5e27803"><span class="section-number-3">2.1.</span> Writing Style</h3>
<div class="outline-text-3" id="text-2-1">
<p>
Sentences with the first person plural ("we") as subject
or with implied third person (it reads as "[The program] &#x2026;"),
are notes about the development process or an explanation of the program,
whereas sentences with the  second person as subject ("You" e.g. "You might try setting foo to 3")
are instructions to the end user.
</p>
</div>
</div>

<div id="outline-container-orgb4fba17" class="outline-3">
<h3 id="orgb4fba17"><span class="section-number-3">2.2.</span> Trouble Shooting</h3>
<div class="outline-text-3" id="text-2-2">
<p>
When you encounter an error, go to the section of this document that this error occurred,
there we will describe fixes for errors that occur in that step.
You can also search for the text the error message gave, we will usually include it in that section.
</p>
</div>
</div>
<div id="outline-container-org917fb29" class="outline-3">
<h3 id="org917fb29"><span class="section-number-3">2.3.</span> Tangling</h3>
<div class="outline-text-3" id="text-2-3">
<p>
in order to turn this file into the needed files run, you'll need to have
emacs and gnu make installed, and then run
</p>
<div class="org-src-container">
<pre class="src src-sh">yes <span class="org-string">'//'</span> | make tangle
</pre>
</div>

<p>
alternatively, if you have gnu guix installed, you can run
</p>
<div class="org-src-container">
<pre class="src src-sh">make <span class="org-variable-name">SHELL</span>=guix tangle
</pre>
</div>

<p>
The following is a list of all files this document tangles into
</p>
<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />
</colgroup>
<tbody>
<tr>
<td class="org-left">gwl-gladiator.scm</td>
</tr>

<tr>
<td class="org-left">gladiator.nf</td>
</tr>

<tr>
<td class="org-left">diaumpireconfig.txt</td>
</tr>

<tr>
<td class="org-left">comet_template.txt</td>
</tr>

<tr>
<td class="org-left">xtandem-template.xml</td>
</tr>

<tr>
<td class="org-left">tpp-5.2-fix.diff</td>
</tr>

<tr>
<td class="org-left">irt.txt</td>
</tr>

<tr>
<td class="org-left">pyprophet-legacy-requirements.txt</td>
</tr>

<tr>
<td class="org-left">pyprophet-legacy-standalone.dockerfile</td>
</tr>

<tr>
<td class="org-left">install-R-packages.R</td>
</tr>

<tr>
<td class="org-left">config/singularity.nf</td>
</tr>

<tr>
<td class="org-left">config/singularity-local.nf</td>
</tr>

<tr>
<td class="org-left">config/docker.nf</td>
</tr>

<tr>
<td class="org-left">config/docker-local.nf</td>
</tr>

<tr>
<td class="org-left">config/podman.nf</td>
</tr>

<tr>
<td class="org-left">config/podman-local.nf</td>
</tr>

<tr>
<td class="org-left">nextflow.config</td>
</tr>

<tr>
<td class="org-left">nextflow.tags</td>
</tr>
</tbody>
</table>
</div>
</div>


<div id="outline-container-org65077e8" class="outline-3">
<h3 id="org65077e8"><span class="section-number-3">2.4.</span> Building the containers</h3>
<div class="outline-text-3" id="text-2-4">
<p>
Containers are available from the public registry,
but you can also build them yourself.
The containers are defined in terms guix manifests, rather than Dockerfiles or the like,
so you will need to install <a href="https://guix.gnu.org">guix</a> in order to build the containers yourself.
</p>
<div class="org-src-container">
<pre class="src src-sh"><span class="org-comment-delimiter"># </span><span class="org-comment">if you want singularity images</span>
make <span class="org-variable-name">SHELL</span>=guix singularity-containers
<span class="org-comment-delimiter"># </span><span class="org-comment">if you want docker images</span>
make <span class="org-variable-name">SHELL</span>=guix docker-containers
</pre>
</div>

<p>
If you want to pass arguments to the step that's responsible for building the containers,
(like e.g. guix style transformations), you can specify those with make variable <code>GUIX_PACK_FLAGS</code>.
e.g. 
</p>
<div class="org-src-container">
<pre class="src src-sh">make <span class="org-variable-name">SHELL</span>=guix <span class="org-string">'GUIX_PACK_FLAGS=--with-patch=tpp=/path/to/my/tpp/patchfile.patch'</span> singularity-containers
</pre>
</div>
</div>
</div>


<div id="outline-container-org5159ba9" class="outline-3">
<h3 id="org5159ba9"><span class="section-number-3">2.5.</span> License</h3>
<div class="outline-text-3" id="text-2-5">
<p>
This program is free software, under GPL3 or later.
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">/*</span>
<span class="org-comment"> * Copyright (C) 2025</span>
<span class="org-comment"> *</span>
<span class="org-comment"> * This program is free software: you can redistribute it and/or modify</span>
<span class="org-comment"> * it under the terms of the GNU General Public License as published by</span>
<span class="org-comment"> * the Free Software Foundation, either version 3 of the License, or</span>
<span class="org-comment"> * (at your option) any later version.</span>
<span class="org-comment"> *</span>
<span class="org-comment"> * This program is distributed in the hope that it will be useful,</span>
<span class="org-comment"> * but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="org-comment"> * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="org-comment"> * GNU General Public License for more details.</span>
<span class="org-comment"> *</span>
<span class="org-comment"> * You should have received a copy of the GNU General Public License</span>
<span class="org-comment"> * along with this program.  If not, see <a href="http://www.gnu.org/licenses/">&lt;http://www.gnu.org/licenses/&gt;</a>.</span>
<span class="org-comment"> */</span>
</pre>
</div>
</div>
</div>
</div>
<div id="outline-container-orga714a85" class="outline-2">
<h2 id="orga714a85"><span class="section-number-2">3.</span> Overview of this pipeline</h2>
<div class="outline-text-2" id="text-3">

<div id="orgdb987d1" class="figure">
<p><img src="img/flow.svg" alt="flow.svg" class="org-svg" />
</p>
</div>
</div>
</div>



<div id="outline-container-orgd37ea86" class="outline-2">
<h2 id="orgd37ea86"><span class="section-number-2">4.</span> Preprocessing Data</h2>
<div class="outline-text-2" id="text-4">
<p>
We will not distribute the vendored msconvert,
but if you have DDA-files you need to convert froma propriatry format, to mzmxml,
following the picking peaks step,
and you can use the docker image of <code>dockerhub:chambm/pwiz-skyline-i-agree-to-the-vendor-licenses</code>.
You can convert your DIA-files with the same container following "Converting Dia Raw with Msconvert"
</p>
</div>
<div id="outline-container-orgc4f8110" class="outline-3">
<h3 id="orgc4f8110"><span class="section-number-3">4.1.</span> Picking Peaks</h3>
<div class="outline-text-3" id="text-4-1">
<div class="org-src-container">
<pre class="src src-sh">mkdir -p MZXML-pwiz
<span class="org-keyword">for</span> f<span class="org-keyword"> in</span> RAW/*.wiff; <span class="org-keyword">do</span>
    wine qtofpeakpicker --resolution=2000 --area=1 --threshold=1 --smoothwidth=1.1 --in $<span class="org-variable-name">f</span> --out MZXML-pwiz/$(basename --suffix=.wiff $<span class="org-variable-name">f</span>).mzXML
<span class="org-keyword">done</span>
</pre>
</div>
</div>
</div>
<div id="outline-container-org4a1e12d" class="outline-3">
<h3 id="org4a1e12d"><span class="section-number-3">4.2.</span> Converting Dia Raw with Msconvert</h3>
<div class="outline-text-3" id="text-4-2">
<div class="org-src-container">
<pre class="src src-sh">mkdir -p MZML-pwiz
find . -iname <span class="org-string">'*.wiff'</span> -print0 | xargs -P5 -0 -i wine msconvert {} --filter <span class="org-string">'titleMaker &lt;RunId&gt;.&lt;ScanNumber&gt;.&lt;ScanNumber&gt;.&lt;ChargeState&gt; File:"&lt;SourcePath&gt;", NativeID:"&lt;Id&gt;"'</span> -o MZML-pwiz/
</pre>
</div>
</div>
</div>
</div>


<div id="outline-container-org90732e0" class="outline-2">
<h2 id="org90732e0"><span class="section-number-2">5.</span> Analysis <code>[4/4]</code></h2>
<div class="outline-text-2" id="text-5">
<p>
Gladiator paper: <a href="https://doi.org/10.1038/s43705-022-00137-0">https://doi.org/10.1038/s43705-022-00137-0</a>
diatools: <a href="https://doi.org/10.1021/acs.jproteome.9b00606">https://doi.org/10.1021/acs.jproteome.9b00606</a>
</p>
</div>
<div id="outline-container-org1e918b2" class="outline-3">
<h3 id="org1e918b2"><span class="section-number-3">5.1.</span> Headers</h3>
<div class="outline-text-3" id="text-5-1">
<div class="org-src-container">
<pre class="src src-emacs-lisp">(<span class="org-keyword">setq</span> org-babel-tangle-lang-exts
      (cl-remove-duplicates 
      (append
       '((<span class="org-string">"scheme"</span> . <span class="org-string">"scm"</span>))
       org-babel-tangle-lang-exts
       )
      <span class="org-builtin">:test</span> 'equal))
</pre>
</div>
<div class="org-src-container">
<pre class="src src-scheme" id="orgd9b957a">(<span class="org-keyword">define-module</span> (<span class="org-type">workflow</span>)
  <span class="org-builtin">#:use-module</span> (gwl workflows)
  <span class="org-builtin">#:use-module</span> (gwl processes)
  <span class="org-builtin">#:use-module</span> (gwl utils)
  <span class="org-builtin">#:use-module</span> (gwl sugar))

<span class="org-type">&lt;&lt;gwl-vars&gt;&gt;</span>
</pre>
</div>



<div class="org-src-container">
<pre class="src src-nextflow" id="org41ee01c">&lt;&lt;nf-sdrf-handling&gt;&gt;
&lt;&lt;nf-function-definitions&gt;&gt;
&lt;&lt;nf-vars&gt;&gt;
</pre>
</div>
</div>
</div>


<div id="outline-container-org1ed1558" class="outline-3">
<h3 id="org1ed1558"><span class="section-number-3">5.2.</span> building database</h3>
<div class="outline-text-3" id="text-5-2">
<p>
<b>Overview</b>

</p>


<div id="org26c4b7b" class="figure">
<p><img src="img/dot-build-database.svg" alt="dot-build-database.svg" class="org-svg" />
</p>
</div>
</div>


<div id="outline-container-orgc0e70a2" class="outline-4">
<h4 id="orgc0e70a2"><span class="section-number-4">5.2.1.</span> Combining Fasta Files</h4>
<div class="outline-text-4" id="text-5-2-1">
<div class="org-src-container">
<pre class="src src-python" id="orgf3c4177"><span class="org-keyword">from</span> Bio <span class="org-keyword">import</span> SeqIO
<span class="org-keyword">def</span> <span class="org-function-name">join_fasta_files</span>(input_files, output_file):
    <span class="org-variable-name">IDs</span> = <span class="org-builtin">set</span>()
    <span class="org-variable-name">seqRecords</span> = []
    <span class="org-keyword">for</span> filename <span class="org-keyword">in</span> <span class="org-variable-name">input_files</span>:
        records = SeqIO.index(filename, <span class="org-string">"fasta"</span>)
        <span class="org-keyword">for</span> ID <span class="org-keyword">in</span> records:
            <span class="org-keyword">if</span> ID <span class="org-keyword">not</span> <span class="org-keyword">in</span> IDs:
                seqRecords.append(records[ID])
                IDs.add(ID)
            <span class="org-keyword">else</span>:
                <span class="org-builtin">print</span>(<span class="org-string">"Found duplicated sequence ID "</span> + <span class="org-builtin">str</span>(ID) + <span class="org-string">", skipping this sequence from file "</span> + filename)

    SeqIO.write(seqRecords, output_file, <span class="org-string">"fasta"</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow" id="org9be2cc8"><span class="org-keyword">process</span> <span class="org-function-name">JoinFastaFiles</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">fasta_files</span> from fasta_files_ch.toSortedList()
    output:
    file <span class="org-string">'joined_database.fasta'</span> into joined_fasta_database_ch

    <span class="org-string">"""</span>
<span class="org-string">    #!/usr/bin/env python3</span>
<span class="org-string">    &lt;&lt;py-joinfastafiles&gt;&gt;</span>
<span class="org-string">    join_fasta_files("</span><span class="org-variable-name">$fasta_files</span><span class="org-string">".split(" "), 'joined_database.fasta')</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow">params.fastafiles=<span class="org-string">'fasta/*.fasta'</span>
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-type">Channel</span>.fromPath(params.fastafiles).set{fasta_files_ch}
</pre>
</div>

<p>
This was how we could set the fasta_files_ch to be in the same order as
the original bruderer run
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-type">Channel</span>.from([
    <span class="org-string">"fasta/Q7M135.fasta"</span>,
    <span class="org-string">"fasta/irtfusion.fasta"</span>,
    <span class="org-string">"fasta/trypsin.fasta"</span>,
    <span class="org-string">"fasta/uniprot_human_2017_04_05.fasta"</span>,
    <span class="org-string">"fasta/Bruderer_QS-spike-in-proteins.fasta"</span>])
    .map{file(<span class="org-variable-name">it</span>)}
    .set({fasta_files_ch})
</pre>
</div>
<div class="org-src-container">
<pre class="src src-scheme" id="org203ed5a">(<span class="org-string">"join-fasta-files"</span>
 <span class="org-string">"python"</span>
 <span class="org-string">"biopython"</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-scheme" id="org7e8a6c7">(<span class="org-keyword">define</span> (<span class="org-function-name">join-fasta-files</span> fasta-files)
  (make-process
   (name <span class="org-string">"join-fasta-files"</span>)
   (synopsis <span class="org-string">"Join fasta files into one file"</span>)
   (packages
    (cdr (quote
      <span class="org-type">&lt;&lt;gwl-joinfastafiles-deps&gt;&gt;</span>)))
   (inputs (files fasta-files))
   (outputs <span class="org-string">"joined-fasta.fasta"</span>)
   # python
{
<span class="org-type">&lt;&lt;py-joinfastafiles&gt;&gt;</span>
join_fasta_files({{inputs}}.split(<span class="org-string">" "</span>),{{outputs}})
}))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-scheme">(<span class="org-keyword">define</span> <span class="org-function-name">fasta-files</span>
  '(<span class="org-string">"Q7M135.fasta"</span> <span class="org-string">"trypsin.fasta"</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-scheme">(join-fasta-files fasta-files)
</pre>
</div>
</div>
</div>

<div id="outline-container-org6ddc65b" class="outline-4">
<h4 id="org6ddc65b"><span class="section-number-4">5.2.2.</span> Adding Decoys</h4>
<div class="outline-text-4" id="text-5-2-2">
<div class="org-src-container">
<pre class="src src-nextflow" id="org7039025">fasta_db_with_decoys = <span class="org-type">Channel</span>.value()
<span class="org-keyword">process</span> <span class="org-function-name">BuildFastaDatabase</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">joined_fasta_db</span> from joined_fasta_database_ch
    output:
    file <span class="org-string">"DB_with_decoys.fasta"</span> into joined_fasta_with_decoys_ch
    <span class="org-string">"""</span>
<span class="org-string">    DecoyDatabase -in </span><span class="org-variable-name">$joined_fasta_db</span><span class="org-string"> -out DB_with_decoys.fasta</span>
<span class="org-string">    """</span>
}
</pre>
</div>
<p>
<code>DecoyDatabase</code> package is from <code>OpenMs/utils</code>
<a href="https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/UTILS_DecoyDatabase.html">https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/UTILS_DecoyDatabase.html</a>
<a href="https://github.com/OpenMS/OpenMS.git">https://github.com/OpenMS/OpenMS.git</a>
License: BSD-3 clause
(Not in guix, but uses cmake as build-program,
should be relatively easy to define)
</p>
<div class="org-src-container">
<pre class="src src-scheme" id="org9c6489c">(<span class="org-keyword">define</span> <span class="org-function-name">create-database-with-decoys</span>
  (make-process
   (name <span class="org-string">"create-database-with-database"</span>)
   (synopsis <span class="org-string">"Add decoys to fasta database"</span>)
   (inputs <span class="org-string">"joined-fasta.fasta"</span>)
   (outputs <span class="org-string">"DB-with-decoys.fasta"</span>)
   (packages )
   # sh
     {
      DecoyDatabase -in $inputs -out $outputs
                    })))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-scheme">create-database-with-decoys
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org2d0e377" class="outline-3">
<h3 id="org2d0e377"><span class="section-number-3">5.3.</span> Input DIA files</h3>
<div class="outline-text-3" id="text-5-3">
<p>
Here we redirect the dia files to the
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-type">Channel</span>
    .fromPath(params.diafiles)
    .multiMap{
        <span class="org-variable-name">it</span> -&gt; swath_windows: osw: <span class="org-variable-name">it</span>}
    .set{dia_mzml_files_ch}
</pre>
</div>
</div>
</div>

<div id="outline-container-org47af1f4" class="outline-3">
<h3 id="org47af1f4"><span class="section-number-3">5.4.</span> Creating Swath window files</h3>
<div class="outline-text-3" id="text-5-4">
<p>

outputs files <code>swath-windows.txt</code>,
<code>truncated-swath-windows.txt</code>
</p>

<p>
If you are using FAIMS split MZMLs,
the mzml might not contain <code>isolationWindow</code> elements,
in that case you can provide your own tab-separated file of swathwindows.
</p>
</div>

<div id="outline-container-org7d35826" class="outline-4">
<h4 id="org7d35826"><span class="section-number-4">5.4.1.</span> Branching if user supplied windows</h4>
<div class="outline-text-4" id="text-5-4-1">
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">optional swath windows file thats a tab-separated file</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">where the first column is the isolation window lower offset</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">and the second column  is the isolation window upper offset</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">this file is normally automatically generated in the MakeSwathWindows steps</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">but if your mzML does not provide isolationWindow </span>
params.swath_windows_file=<span class="org-string">''</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span> (params.swath_windows_file) {    
  &lt;&lt;nf-regularize-user-swath-windows&gt;&gt;  
} <span class="org-keyword">else</span>  {
    &lt;&lt;nf-infer-swath-windows&gt;&gt;
}
</pre>
</div>
</div>
</div>


<div id="outline-container-org894d3d1" class="outline-4">
<h4 id="org894d3d1"><span class="section-number-4">5.4.2.</span> Making truncated-swath-windows and swath-windows from user-supplied swath-windows</h4>
<div class="outline-text-4" id="text-5-4-2">
<p>
Here we do some mangling so that the user inputed swathwindows is in the same format as the one that would be generated by us.
</p>

<p>
We keeep <code>FS</code> to the default so that awk will happily accept whitespace as field separator,
(be it normal spaces or tabs),
but we output with tab as separators.
</p>

<div class="org-src-container">
<pre class="src src-nextflow" id="org5cd073a"><span class="org-keyword">process</span> <span class="org-function-name">RegularizeUserSwathWindow</span> {
    input:
    path user_swath_windows, stageAs: <span class="org-string">'userSwathWindow.txt'</span> from <span class="org-type">Channel</span>.fromPath(params.swath_windows_file).first()
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">swath_windows</span> into swath_windows_ch
    script:
    swath_windows=<span class="org-string">"swath-windows.txt"</span>
    <span class="org-string">"""</span>
<span class="org-string">    sort -n </span><span class="org-variable-name">$user_swath_windows</span><span class="org-string"> | awk 'BEGIN {OFS="     "} {print \$1,\$2}' &gt;  </span><span class="org-variable-name">$swath_windows</span>
<span class="org-string">    """</span>
}
</pre>
</div>
</div>
</div>


<div id="outline-container-org3facb06" class="outline-4">
<h4 id="org3facb06"><span class="section-number-4">5.4.3.</span> Inferring Windows from mzml files</h4>
<div class="outline-text-4" id="text-5-4-3">
<p>
If the user didnt supply a swath windows file, we infer it from the mzml file
</p>

<div class="org-src-container">
<pre class="src src-python" id="org96263c2"><span class="org-keyword">import</span> xml.etree.ElementTree <span class="org-keyword">as</span> ET
<span class="org-keyword">import</span> os

<span class="org-keyword">def</span> <span class="org-function-name">read_swath_windows</span>(dia_mzML):

    <span class="org-builtin">print</span> (<span class="org-string">"DEBUG: reading_swath_windows: "</span>, dia_mzML)

    context = ET.iterparse(dia_mzML, events=(<span class="org-string">"start"</span>, <span class="org-string">"end"</span>))

    windows = {}
    <span class="org-keyword">for</span> event, elem <span class="org-keyword">in</span> context:

        <span class="org-keyword">if</span> event == <span class="org-string">"end"</span> <span class="org-keyword">and</span> elem.tag == <span class="org-string">'{http://psi.hupo.org/ms/mzml}precursor'</span>:
            il_target = <span class="org-constant">None</span>
            il_lower = <span class="org-constant">None</span>
            il_upper = <span class="org-constant">None</span>

            isolationwindow = elem.find(<span class="org-string">'{http://psi.hupo.org/ms/mzml}isolationWindow'</span>)
            <span class="org-keyword">if</span> isolationwindow <span class="org-keyword">is</span> <span class="org-constant">None</span>:
                <span class="org-keyword">raise</span> <span class="org-type">RuntimeError</span>(<span class="org-string">"Could not find isolation window; please supply --swath_windows_file to Gladiator."</span>)
            <span class="org-keyword">for</span> cvParam <span class="org-keyword">in</span> isolationwindow.findall(<span class="org-string">'{http://psi.hupo.org/ms/mzml}cvParam'</span>):
                name = cvParam.get(<span class="org-string">'name'</span>)
                value = cvParam.get(<span class="org-string">'value'</span>)

                <span class="org-keyword">if</span> (name == <span class="org-string">'isolation window target m/z'</span>):
                    il_target = value
                <span class="org-keyword">elif</span> (name == <span class="org-string">'isolation window lower offset'</span>):
                    il_lower = value
                <span class="org-keyword">elif</span> (name == <span class="org-string">'isolation window upper offset'</span>):
                    il_upper = value

            ionList = elem.find(<span class="org-string">'{http://psi.hupo.org/ms/mzml}selectedIonList'</span>)

            selectedion = ionList.find(<span class="org-string">'{http://psi.hupo.org/ms/mzml}selectedIon'</span>)

            <span class="org-keyword">if</span> selectedion:

                <span class="org-keyword">for</span> cvParam <span class="org-keyword">in</span> selectedion.findall(<span class="org-string">'{http://psi.hupo.org/ms/mzml}cvParam'</span>):
                    name = cvParam.get(<span class="org-string">'name'</span>)
                    value = cvParam.get(<span class="org-string">'value'</span>)

                    <span class="org-keyword">if</span> (name == <span class="org-string">'selected ion m/z'</span>):
                        <span class="org-keyword">if</span> <span class="org-keyword">not</span> il_target:
                            il_target = value

            <span class="org-keyword">if</span> <span class="org-keyword">not</span> il_target <span class="org-keyword">in</span> windows:
                windows[<span class="org-variable-name">il_target</span>] = (il_lower, il_upper)
            <span class="org-keyword">else</span>:
                lower, upper = windows[il_target]
                <span class="org-keyword">assert</span> (il_lower == lower)
                <span class="org-keyword">assert</span> (il_upper == upper)
                <span class="org-keyword">return</span> windows

    <span class="org-keyword">return</span> windows

<span class="org-keyword">def</span> <span class="org-function-name">create_swath_window_files</span>(cwd, dia_mzML):
    windows = read_swath_windows(dia_mzML)
    swaths = []
    <span class="org-keyword">for</span> x <span class="org-keyword">in</span> windows:
        target_str = x
        lower_str, upper_str = windows[x]
        target = <span class="org-builtin">float</span>(target_str)
        lower = <span class="org-builtin">float</span>(lower_str)
        upper = <span class="org-builtin">float</span>(upper_str)
        <span class="org-keyword">assert</span> (lower &gt; 0)
        <span class="org-keyword">assert</span> (upper &gt; 0)
        swaths.append((target - lower, target + upper))
    swaths.sort(key=<span class="org-keyword">lambda</span> tup: tup[0])
    <span class="org-comment-delimiter"># </span><span class="org-comment">here we use chr(10) (equivalent to slash n), and chr(9) (equivalent to slash t)  because i dont wanna deal with nextflow headaches</span>
    newline_character = <span class="org-builtin">chr</span>(10)
    tab_character = <span class="org-builtin">chr</span>(9)
    <span class="org-keyword">with</span> <span class="org-builtin">open</span>(os.path.join(cwd, <span class="org-string">"swath-windows.txt"</span>), <span class="org-string">"w"</span>) <span class="org-keyword">as</span> fh_swaths:
        <span class="org-keyword">for</span> lower,upper <span class="org-keyword">in</span> swaths:
            fh_swaths.write(<span class="org-builtin">str</span>(lower) + tab_character + <span class="org-builtin">str</span>(upper)  + newline_character)
    <span class="org-keyword">return</span> fh_swaths

</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow" id="org7a62001"><span class="org-keyword">process</span> <span class="org-function-name">InferSwathWindows</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">diafile</span> from dia_mzml_files_ch.swath_windows.first()
    output: 
    file <span class="org-string">"swath-windows.txt"</span> into  swath_windows_ch
    shell:
    <span class="org-string">'''</span>
<span class="org-string">    #!/usr/bin/env python3</span>
<span class="org-string">    &lt;&lt;py-makeswathwindows&gt;&gt;</span>
<span class="org-string">    swaths = create_swath_window_files(".","!{diafile}")</span>
<span class="org-string">    '''</span>
}
</pre>
</div>

<p>
we'll have to get minswath and maxswath by reading <code>"swath-windows.txt"</code>
</p>
</div>
</div>
<div id="outline-container-orgb22c933" class="outline-4">
<h4 id="orgb22c933"><span class="section-number-4">5.4.4.</span> Making the non-overlapping swath-windows</h4>
<div class="outline-text-4" id="text-5-4-4">
<p>
<code>Openswath</code> requires non-overlapping windows,
so we create them here.
</p>

<div class="org-src-container">
<pre class="src src-awk" id="org9f43c4c"><span class="org-keyword">BEGIN</span> {<span class="org-variable-name">OFS</span>=<span class="org-string">"    "</span>}
<span class="org-keyword">function</span> <span class="org-function-name">max</span>(a,b){
    <span class="org-keyword">if</span>(a &gt; b)
        <span class="org-keyword">return</span> a
    <span class="org-keyword">return</span> b
}
<span class="org-variable-name">NR</span>==1 {
    <span class="org-comment-delimiter"># </span><span class="org-comment">we start with the special case that the boundary for the first entry</span>
    <span class="org-comment-delimiter"># </span><span class="org-comment">should be unchanged</span>
    prev_upper=$1
    <span class="org-comment-delimiter"># </span><span class="org-comment">and we add the column names</span>
    <span class="org-preprocessor">print</span> <span class="org-string">"LowerOffset"</span>,<span class="org-string">"HigherOffset"</span>
}
{
    <span class="org-keyword">if</span> (prev_upper &gt; $2)
    {
        <span class="org-preprocessor">print</span> <span class="org-string">"There is a a window thats a subwindow of the previous window"</span>
        <span class="org-keyword">exit</span> 1
    }
    <span class="org-preprocessor">print</span>(max($1,prev_upper),$2)
    prev_upper=$2
}
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">InferNonOverlappingSwathWindows</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">swath_windows</span> from swath_windows_ch.first()
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">truncated_swath_windows</span> into truncated_swath_windows_ch
    script:
    truncated_swath_windows=<span class="org-string">"truncated_swath_windows.txt"</span>
    <span class="org-string">''' awk '</span>
<span class="org-string">    &lt;&lt;awk-infer-non-overlapping-swath-windows&gt;&gt;' '''</span> + <span class="org-string">"</span><span class="org-variable-name">$swath_windows</span><span class="org-string"> &gt; </span><span class="org-variable-name">$truncated_swath_windows</span><span class="org-string">"</span>
}
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-orgc4069b5" class="outline-3">
<h3 id="orgc4069b5"><span class="section-number-3">5.5.</span> Library Generation</h3>
<div class="outline-text-3" id="text-5-5">
<p>
There are various way to generate spectral libraries from DIA data / DDA dat.
Here we make the distinction between deconvolution methods and other library generation methods.
</p>

<p>
The following is a list of the methods we support,
</p>
<div class="org-src-container">
<pre class="src src-nextflow" id="org636bec3">[ <span class="org-string">"dda"</span>,<span class="org-string">"custom"</span>, <span class="org-string">"deepdia"</span>, <span class="org-string">"diaumpire"</span>,<span class="org-string">"diams2pep"</span>]
</pre>
</div>

<p>
And you can adjust the following parameter 
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">one or more of [ "dda","custom", "deepdia", "diaumpire","diams2pep"] seperated by commas</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">will default to "dda" if ddafiles are supplied</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">othewise to "deepdia"</span>
params.libgen_method = <span class="org-constant">null</span>
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">TODO: raise an error if params.libgen_method  is not a supported method</span>
libgen_methods_validate_params(params)
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">returns all libgen methods that we supplor</span>
<span class="org-keyword">def</span> <span class="org-function-name">libgen_methods_get_existing</span> (){
    <span class="org-keyword">return</span> [ <span class="org-string">"dda"</span>,<span class="org-string">"custom"</span>, <span class="org-string">"deepdia"</span>, <span class="org-string">"diaumpire"</span>,<span class="org-string">"diams2pep"</span>]
}

<span class="org-keyword">def</span> <span class="org-function-name">libgen_method_any_pseudospectra_method_is_enabled</span>(params){
    <span class="org-keyword">def</span> <span class="org-variable-name">pseudospectra_methods</span> = [<span class="org-string">"diams2pep"</span>,<span class="org-string">"diaumpire"</span>]
    <span class="org-keyword">return</span> pseudospectra_methods.inject(<span class="org-constant">false</span>) { acc, val -&gt; acc || libgen_method_is_enabled(val, params)}
}


<span class="org-keyword">def</span> <span class="org-function-name">libgen_methods_validate_params</span>(params){
    <span class="org-keyword">if</span>(params.libgen_method != <span class="org-constant">null</span>){
        <span class="org-keyword">def</span> <span class="org-variable-name">invalid_methods</span> = params.libgen_method.split(<span class="org-string">","</span>).findAll({    !libgen_methods_get_existing().contains(<span class="org-variable-name">it</span>)})
        <span class="org-keyword">if</span>(invalid_methods)
            raise <span class="org-type">RunTimeError</span>(<span class="org-string">"Invalid libgen methods specified: "</span> + invalid_methods.join(<span class="org-string">","</span>))
    }
}
<span class="org-keyword">def</span> <span class="org-function-name">libgen_method_is_enabled</span>(method, params){
    <span class="org-comment-delimiter">// </span><span class="org-comment">method to use if the user didnt specify anything</span>
    <span class="org-keyword">def</span> <span class="org-variable-name">fallback_method</span> = <span class="org-string">"diaumpire"</span>;
    <span class="org-keyword">if</span> (params.libgen_method){
        <span class="org-keyword">return</span> params.libgen_method.split(<span class="org-string">","</span>).contains(method)
    }
    <span class="org-keyword">switch</span> (method) {
            <span class="org-keyword">case</span> <span class="org-string">"dda"</span>: <span class="org-keyword">return</span> !!params.ddafiles;
            <span class="org-keyword">case</span> <span class="org-string">"custom"</span>: <span class="org-keyword">return</span> !!params.speclib;
        <span class="org-keyword">default</span>: <span class="org-keyword">return</span> (method == fallback_method) &amp;&amp; !params.ddafiles &amp;&amp; !params.speclib;  
    }
}

<span class="org-keyword">def</span> <span class="org-function-name">libgen_method_is_exclusively_enabled</span>(method, params) {
    <span class="org-keyword">return</span> libgen_methods_get_existing().inject(<span class="org-constant">true</span>) { acc, val -&gt; acc &amp;&amp; ( libgen_method_is_enabled(val, params)  == (val == methods)) }
}
</pre>
</div>
</div>
<div id="outline-container-org392471a" class="outline-4">
<h4 id="org392471a"><span class="section-number-4">5.5.1.</span> Building {Pseudo-,}Spectral library from (Pseudo)-Spectra  <code>[5/5]</code></h4>
<div class="outline-text-4" id="text-5-5-1">
<p>
This section covers seval ways of deconvolution for making spectral libraries for later usage by open swath.
</p>

<p>
The below block handles the logic of dealing with the various deconvolution methods, sending diafiles to all input channels,
and getting output mgf from output channels.
</p>
<div class="org-src-container">
<pre class="src src-nextflow">deconvolution_methods = []
&lt;&lt;nf-deconvolution-handling&gt;&gt;
deconv_input_chs = deconvolution_methods*.input.findAll({<span class="org-variable-name">it</span> != <span class="org-constant">null</span>})
<span class="org-keyword">if</span>(deconv_input_chs){
    <span class="org-type">Channel</span>
        .fromPath(params.diafiles)
        .into(
            deconv_input_chs
                .inject() { acc, val -&gt; acc &lt;&lt; val })
}
deconv_output_chs = deconvolution_methods*.output.findAll({<span class="org-variable-name">it</span> != <span class="org-constant">null</span>})
<span class="org-keyword">for</span>(ch:deconv_output_chs)
    <span class="org-type">Channel</span>.create().set(ch.clone())
<span class="org-keyword">if</span>(deconv_output_chs){
    <span class="org-type">Channel</span>.empty()
        .mix(*(deconv_output_chs*.call()))
        .multiMap{ <span class="org-variable-name">it</span> -&gt; spectrast: comet: xtandem: <span class="org-variable-name">it</span> }
        .set{maybespectra_ch}
}
</pre>
</div>
<div class="org-src-container">
<pre class="src src-nextflow">
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(libgen_method_any_pseudospectra_method_is_enabled(params) || libgen_method_is_enabled(<span class="org-string">"dda"</span>,params)){
</pre>
</div>
</div>
<div id="outline-container-org6b7746b" class="outline-5">
<h5 id="org6b7746b"><span class="section-number-5">5.5.1.1.</span> Using DDA Data</h5>
<div class="outline-text-5" id="text-5-5-1-1">
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(libgen_method_is_enabled(<span class="org-string">"dda"</span>,params)){
    deconvolution_methods += [output: { dda_files_ch } ]
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(libgen_method_is_enabled(<span class="org-string">"dda"</span>,params)){
    <span class="org-type">Channel</span>.fromPath(params.ddafiles).tap(dda_files_ch)
}
</pre>
</div>
</div>
</div>

<div id="outline-container-orgc88f836" class="outline-5">
<h5 id="orgc88f836"><span class="section-number-5">5.5.1.2.</span> Creating Pseudospectra with DIAumpire</h5>
<div class="outline-text-5" id="text-5-5-1-2">
<p>

</p>
<p>
<img src="img/dot-create-pseudospectra.svg" alt="dot-create-pseudospectra.svg" class="org-svg" />
<a href="https://github.com/Nesvilab/DIA-Umpire/tree/master">https://github.com/Nesvilab/DIA-Umpire/tree/master</a>
<a href="https://github.com/Nesvilab/DIA-Umpire/raw/gh-pages/DIA_Umpire_Manual_v2.0.pdf">https://github.com/Nesvilab/DIA-Umpire/raw/gh-pages/DIA_Umpire_Manual_v2.0.pdf</a>
</p>
</div>
<div id="outline-container-org0fd7577" class="outline-6">
<h6 id="org0fd7577"><span class="section-number-6">5.5.1.2.1.</span> Problems you might encounter during this step</h6>
<div class="outline-text-6" id="text-5-5-1-2-1">
</div>
<div id="outline-container-orgf53a0bd" class="outline-7">
<h7 id="orgf53a0bd"><span class="section-number-7">5.5.1.2.1.1.</span> Out of Memory in Dia-umpire</h7>
<div class="outline-text-7" id="text-5-5-1-2-1-1">
<p>
Dia-umpire, which we use here for pseudo-spectra creation,
has pretty extreme memory requirements,
in  your config file you can set the process specific memory (required to be in Gigabyes) e.g. 
</p>
<div class="org-src-container">
<pre class="src src-nextflow">process { 
  withName: <span class="org-string">'GeneratePseudoSpectra'</span>
  {
        time=<span class="org-string">'96h'</span>
        memory=<span class="org-string">'400 GB'</span>
  }
}
</pre>
</div>
<p>
see also <a href="https://www.nextflow.io/docs/latest/process.html#memory">The Nextflow documentation about process memory</a>
</p>
</div>
</div>
<div id="outline-container-orga5f56ad" class="outline-7">
<h7 id="orga5f56ad"><span class="section-number-7">5.5.1.2.1.2.</span> <code>MzmlToMzxml</code> processing error.</h7>
<div class="outline-text-7" id="text-5-5-1-2-1-2">
<p>
If you get an error of
</p>
<div class="org-src-container">
<pre class="src src-fundamental">processing file: RD139_Narrow_UPS1_50fmol_inj3.mzML
[SpectrumList_mzML::create()] Bad istream.
Error processing file RD139_Narrow_UPS1_50fmol_inj3.mzML
</pre>
</div>
<p>
in <code>MzmltoMzxml</code>, that can mean that something went wrong
when you used <code>msconvert</code> to convert from the propriatary format to mzml
</p>
</div>
</div>
</div>
<div id="outline-container-org63b988a" class="outline-6">
<h6 id="org63b988a"><span class="section-number-6">5.5.1.2.2.</span> Steps that are run</h6>
<div class="outline-text-6" id="text-5-5-1-2-2">
<p>
DIAumpire is <code>Apache 2</code> licensed.
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(libgen_method_is_enabled(<span class="org-string">"diaumpire"</span>,params)){
    deconvolution_methods += [output: {diaumpire_pseudospectra_ch},
                              input:  {dia_mzml_files_for_diaumpire_ch}]
}
</pre>
</div>
<div class="org-src-container">
<pre class="src src-nextflow">
<span class="org-keyword">if</span>(libgen_method_is_enabled(<span class="org-string">"diaumpire"</span>,params)){
</pre>
</div>

<div class="org-src-container">
<pre class="src src-conf" id="org3c9f453"><span class="org-comment-delimiter">#</span><span class="org-comment">Number of threads</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">set to the number of cores available</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">In the original gladiator, this was set by replicing this all caps </span>
<span class="org-variable-name">Thread</span> = 4

<span class="org-comment-delimiter">#</span><span class="org-comment">Precursor-fragments grouping parameters</span>
<span class="org-variable-name">RPmax</span> = 25
<span class="org-variable-name">RFmax</span> = 300
<span class="org-variable-name">CorrThreshold</span> = 0.2
<span class="org-variable-name">DeltaApex</span> = 0.6
<span class="org-variable-name">RTOverlap</span> = 0.3

<span class="org-comment-delimiter">#</span><span class="org-comment">Fragment intensity adjustments</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">change BoostComplementaryIon if later using database search results to build libraries for Skyline/OpenSWATH</span>
<span class="org-comment-delimiter">## </span><span class="org-comment">[2023-05-30 Tue]</span>
<span class="org-comment-delimiter">## </span><span class="org-comment">what did the original gladiator author mean by this?</span>
<span class="org-comment-delimiter">## </span><span class="org-comment">he forgot.</span>
<span class="org-comment-delimiter">## </span><span class="org-comment">in dia-umpire repo example BoostComplementaryIon is True.</span>
<span class="org-variable-name">AdjustFragIntensity</span> = true
<span class="org-variable-name">BoostComplementaryIon</span> = true

<span class="org-comment-delimiter">#</span><span class="org-comment">Export detected MS1 features (output feature file can be loaded and mapped to RAW data in BatMass)</span>
<span class="org-variable-name">ExportPrecursorPeak</span> = false

<span class="org-comment-delimiter">#</span><span class="org-comment">Signal extraction: mass accuracy and resolution</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">resolution parameter matters only for data generated in profile mode</span>
<span class="org-variable-name">SE.MS1PPM</span> = 15
<span class="org-variable-name">SE.MS2PPM</span> = 25
<span class="org-variable-name">SE.Resolution</span> = 60000

<span class="org-comment-delimiter">#</span><span class="org-comment">Signal extraction: signal to noise filter</span>
<span class="org-variable-name">SE.SN</span> = 1.1
<span class="org-variable-name">SE.MS2SN</span> = 1.1

<span class="org-comment-delimiter">#</span><span class="org-comment">Signal extraction: minimum signal intensity filter</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">for Thermo data, filtering is usually not necessary. Set SE.EstimateBG to false and SE.MinMSIntensity and SE.MinMSMSIntensity to a low value, e.g. 1</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">for older Q Exactive data, or when too many MS1 features are extracted, set SE.EstimateBG to yes (or apply SE.MinMSIntensity and SE.MinMSMSIntensity values based on BatMass visualization)</span>
<span class="org-variable-name">SE.EstimateBG</span> = false
<span class="org-variable-name">SE.MinMSIntensity</span> = 1
<span class="org-variable-name">SE.MinMSMSIntensity</span> = 1

<span class="org-comment-delimiter">#</span><span class="org-comment">Signal extraction: peak curve detection and isotope grouping</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">for older Q Exactive data, or when too many MS1 features are extracted, set SE.NoMissedScan to 1</span>
<span class="org-variable-name">SE.NoMissedScan</span> = 2
<span class="org-variable-name">SE.MaxCurveRTRange</span> = 2
<span class="org-variable-name">SE.RemoveGroupedPeaks</span> = true
<span class="org-variable-name">SE.RemoveGroupedPeaksRTOverlap</span> = 0.3
<span class="org-variable-name">SE.RemoveGroupedPeaksCorr</span> = 0.3
<span class="org-variable-name">SE.MinNoPeakCluster</span> = 2
<span class="org-variable-name">SE.MaxNoPeakCluster</span> = 4

<span class="org-comment-delimiter">#</span><span class="org-comment">Signal extraction: filtering of MS1 features </span>
<span class="org-comment-delimiter"># </span><span class="org-comment">if interested in modified peptides, increase MassDefectOffset parameter, or set SE.MassDefectFilter to false</span>
<span class="org-variable-name">SE.IsoPattern</span> = 0.3
<span class="org-variable-name">SE.MassDefectFilter</span> = true
<span class="org-variable-name">SE.MassDefectOffset</span> = 0.1

<span class="org-comment-delimiter">#</span><span class="org-comment">Signal extraction: other </span>
<span class="org-variable-name">SE.StartCharge</span> = 1
<span class="org-variable-name">SE.EndCharge</span> = 5
<span class="org-variable-name">SE.MS2StartCharge</span> = 2
<span class="org-variable-name">SE.MS2EndCharge</span> = 5
<span class="org-variable-name">SE.MinFrag</span>=10
<span class="org-variable-name">SE.StartRT</span> = 0
<span class="org-variable-name">SE.EndRT</span> = 9999
<span class="org-variable-name">SE.MinMZ</span> = 200
<span class="org-variable-name">SE.MinPrecursorMass</span> = 600
<span class="org-variable-name">SE.MaxPrecursorMass</span> = 5000

<span class="org-comment-delimiter">#</span><span class="org-comment">Isolation window setting</span>
<span class="org-comment-delimiter">#</span><span class="org-comment">The current version supports the following window type: SWATH (fixed window size), V_SWATH (variable SWATH window), MSX, MSE, pSMART</span>
<span class="org-variable-name">WindowType</span>=SWATH

<span class="org-comment-delimiter">#</span><span class="org-comment">Fix window size (For SWATH)</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">for Thermo data, this will be determined from raw data automatically</span>
<span class="org-comment-delimiter">#</span><span class="org-comment">WindowSize=15</span>

<span class="org-comment-delimiter">#</span><span class="org-comment">Variable SWATH window setting (start m/z, end m/z, separated by Tab)</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">for Thermo data, this will be determined from raw data automatically</span>

<span class="org-comment-delimiter">#</span><span class="org-comment">==window setting begin</span>
<span class="org-comment-delimiter">#</span><span class="org-comment">==window setting end</span>
</pre>
</div>
<p>
MGF = Mascot Generic Format
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3518119/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3518119/</a>
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">create mzxml</span>
<span class="org-keyword">process</span> <span class="org-function-name">MzmlToMzxml</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">diafile</span> from dia_mzml_files_for_diaumpire_ch
    output:
    file <span class="org-string">"*.mzXML"</span> into dia_mzxml_files_for_diaumpire_ch
    <span class="org-string">"""</span>
<span class="org-string">    msconvert </span><span class="org-variable-name">$diafile</span><span class="org-string"> --32 --zlib --filter "peakPicking false 1-" --mzXML</span>
<span class="org-string">    """</span>
}

<span class="org-keyword">process</span> <span class="org-function-name">GeneratePseudoSpectra</span>  {
    memory <span class="org-string">'16 GB'</span> 
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">diafile</span> from dia_mzxml_files_for_diaumpire_ch
    path diaumpireconfig from diaumpireconfig_ch.first()
    output:
    <span class="org-comment-delimiter">// </span><span class="org-comment">we flatten here becuase a single mzxml might result in multiple mgf files</span>
    file <span class="org-string">"*.mgf"</span> into diaumpire_pseudospectra_mgf_ch mode flatten 

    <span class="org-string">"""</span>
<span class="org-string">    # we set \$1 to the number of gigs of memory</span>
<span class="org-string">    set -- </span><span class="org-variable-name">$task.memory</span>
<span class="org-string">    if command -v diaumpire-se; </span>
<span class="org-string">    then</span>
<span class="org-string">        diaumpire-se  -Xmx\$1g -Xms\$1g </span><span class="org-variable-name">$diafile</span><span class="org-string"> </span><span class="org-variable-name">$diaumpireconfig</span>
<span class="org-string">    else </span>
<span class="org-string">        java -Xmx\$1g -Xms\$1g -jar /opt/dia-umpire/DIA_Umpire_SE.jar </span><span class="org-variable-name">$diafile</span><span class="org-string"> </span><span class="org-variable-name">$diaumpireconfig</span>
<span class="org-string">    fi</span>
<span class="org-string">    """</span>
}

<span class="org-keyword">process</span> <span class="org-function-name">DiaUmpireMgfToMzxml</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">mgf</span> from diaumpire_pseudospectra_mgf_ch
    output:
    file <span class="org-string">"*.mzXML"</span> into diaumpire_pseudospectra_ch
    when:
    <span class="org-comment-delimiter">// </span><span class="org-comment">excluding empty files   </span>
    mgf.size()  &gt; 0
    <span class="org-string">"""</span>
<span class="org-string">    msconvert </span><span class="org-variable-name">$mgf</span><span class="org-string"> --mzXML </span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-scheme">(<span class="org-string">"generate-pseudo-spectra"</span>
 <span class="org-string">"dia-umpire"</span> 
 <span class="org-string">"pwiz"</span>) <span class="org-comment-delimiter">;; </span><span class="org-comment">the free one </span>
</pre>
</div>


<p>
though this  might also be done with openms's <code>FileConverter</code> ? which is more conventionally build 
<a href="https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FileConverter.html">https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FileConverter.html</a>
mstools
</p>
<div class="org-src-container">
<pre class="src src-nextflow">params.diaumpireconfig=<span class="org-string">'diaumpireconfig.txt'</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">glob to DIA mzmML files, e.g. "DIA/*.mzML"</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">MANDATORY to be set if not set by SDRF file</span>
params.diafiles = <span class="org-constant">null</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">OPTIONAL glob to mzXML dda files</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">e.g. "DDA/*.mzXML"</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">if left unset, then pseudospectra will be used.</span>
params.ddafiles = <span class="org-constant">null</span> 
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">so that this is a singleton channel</span>
diaumpireconfig_ch = <span class="org-type">Channel</span>.fromPath(params.diaumpireconfig)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow">} <span class="org-comment-delimiter">// </span><span class="org-comment">end of diaumpire guard</span>
</pre>
</div>
</div>
</div>
</div>
<div id="outline-container-org50cf2ec" class="outline-5">
<h5 id="org50cf2ec"><span class="section-number-5">5.5.1.3.</span> Creating Pseudospectra with diams2pep</h5>
<div class="outline-text-5" id="text-5-5-1-3">
<p>
<a href="https://github.com/SS2proteome/DIA-MS2pep">https://github.com/SS2proteome/DIA-MS2pep</a>
<a href="https://doi.org/10.52601/bpr.2022.220011">https://doi.org/10.52601/bpr.2022.220011</a>
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">fragment tolarance for diam2spep in ppm</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">(other tools require it in dalton)</span>
params.diams2pep_fragment_tolerance = <span class="org-constant">null</span>
</pre>
</div>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(libgen_method_is_enabled(<span class="org-string">"diams2pep"</span>,params)){
    deconvolution_methods += [output: { diams2pep_pseudospectra},
                              input: { diams2pep_input_mzml}]
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(libgen_method_is_enabled(<span class="org-string">"diams2pep"</span>,params)){
</pre>
</div>

<p>
Do we need msconvert to convert to a friendly mzml file?
According to DIA-MS2PEP's readme we need
</p>
<pre class="example" id="orgb96f47d">
mzML=true
zlib=true
mz64=true
inten64=true
simAsSpectra=true
filter=”peakPicking vendor msLevel=1-2"
</pre>


<p>
trying with filter "cwt" because we don't ship vendors.
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">convert_for_DIAMS2PEP</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">mzml</span> from diams2pep_input_mzml
    output:
    <span class="org-comment-delimiter">// </span><span class="org-comment">there is no good way in nextflow that makes a UUUID that persists across -resume things</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">task.hash is forgotten in resume, as is task.id.</span>
    tuple val(<span class="org-string">"</span><span class="org-variable-name">${mzml.baseName}</span><span class="org-string">"</span>), path(ofile) into diams2pep_mgf_mzml, diams2pep_window_mzml, diams2pep_for_pseudo_mzml
    script:
    ofile=<span class="org-string">"converted/</span><span class="org-variable-name">${mzml.baseName}</span><span class="org-string">.mzML"</span>
    <span class="org-string">"""</span>
<span class="org-string">    mkdir -p converted</span>
<span class="org-string">    msconvert --mzML --mz64 --zlib --inten64 --simAsSpectra --filter "peakPicking cwt msLevel=1-2" --outdir converted </span><span class="org-variable-name">$mzml</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">convert_mgf_for_DIAMS2PEP</span> {
    input:
    tuple val(hash), path(mzml) from diams2pep_mgf_mzml
    output:
    tuple val(hash), path(<span class="org-string">"</span><span class="org-variable-name">${mzml.baseName}</span><span class="org-string">.mgf"</span>) into diams2pep_mgf

    <span class="org-string">"""</span>
<span class="org-string">    msconvert --mgf </span><span class="org-variable-name">$mzml</span>
<span class="org-string">    """</span>
}
</pre>
</div>


<p>
<code>DIA_acquistion_window_generator.pl</code> is <span class="underline">not</span> a typo, this is how it is in the original repo.
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">DIAMS2PEP_window</span> {
    input:
    tuple val(hash), path(mzml) from  diams2pep_window_mzml
    output:
    tuple val(hash), path(<span class="org-string">"</span><span class="org-variable-name">${mzml}</span><span class="org-string">.DIA_acquisition_window.txt"</span>) into diams2pep_window

    <span class="org-string">"""</span>
<span class="org-string">    DIA_acquistion_window_generator.pl </span><span class="org-variable-name">$mzml</span>
<span class="org-string">    """</span>
}
</pre>
</div>


<p>
Looking at the source, this creates an mgf file for every window that was detected.
which would not be known before running <code>DIA_acquisition_window_generator.pl</code>
</p>

<p>
<code>DIA_pesudo_MS2.pl</code>.pl= is <span class="underline">not</span> a typo, this is how it is in the original repo
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(params.diams2pep_fragment_tolerance == <span class="org-constant">null</span>)
    raise <span class="org-type">RunTimeError</span>(<span class="org-string">"DIAMSM2PEP enabled but no diams2pep_fragment_tolerance specified."</span>)
<span class="org-keyword">process</span> <span class="org-function-name">DIAMS2PEP_generate_pseudo</span> {
    input:
    tuple val(hash), path(mzml), path(mgf), path(acq_window) from diams2pep_for_pseudo_mzml.join(diams2pep_mgf).join(diams2pep_window)
    <span class="org-keyword">val</span> <span class="org-function-name">tolerance</span> from <span class="org-type">Channel</span>.value(params.diams2pep_fragment_tolerance)
    output:
    file <span class="org-string">"mgf-output/*.mgf"</span> into diams2pep_pseudospectra_mgf mode flatten
    <span class="org-string">"""</span>
<span class="org-string">    mkdir -p mgf-output</span>
<span class="org-string">    DIA_pesudo_MS2_multiforks.pl </span><span class="org-variable-name">${mzml.baseName}</span><span class="org-string"> mgf-output </span><span class="org-variable-name">$tolerance</span><span class="org-string"> </span><span class="org-variable-name">${task.cpus}</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">this will use the default container because we need msconvert</span>
<span class="org-keyword">process</span> <span class="org-function-name">MgfToMzml_DIAMS2PEP</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">mgf</span> from diams2pep_pseudospectra_mgf
    output:
    file <span class="org-string">"*.mzXML"</span> into diams2pep_pseudospectra
    <span class="org-string">"""</span>
<span class="org-string">    msconvert --mzXML </span><span class="org-variable-name">$mgf</span>
<span class="org-string">    """</span>
}
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow">} <span class="org-comment-delimiter">// </span><span class="org-comment">end of diams2pep guard</span>
</pre>
</div>
</div>
</div>
<div id="outline-container-org29c9d7c" class="outline-5">
<h5 id="org29c9d7c"><span class="section-number-5">5.5.1.4.</span> Choosing the MS Sequence database search engine: Comet/Xtandem</h5>
<div class="outline-text-5" id="text-5-5-1-4">
<div class="org-src-container">
<pre class="src src-nextflow">params.search_engines = [<span class="org-string">"comet"</span>,<span class="org-string">"xtandem"</span>]
</pre>
</div>


<p>
Depending on your experimental machine, the precursor and fragment tolerances are different.
These are parameters to all search engines unsed.
</p>

<p>
Some scientific papers use <code>mmu</code> which is equal to 1 milidalton
0.001 Dalton
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">Float or Int; in ppm ; eg. params.precursor_mass_tolerance=10</span>
params.precursor_mass_tolerance=<span class="org-constant">null</span>  
<span class="org-comment-delimiter">// </span><span class="org-comment">Float or Int; in Dalton; e.g. parames.fragment_mass_tolerance=0.2</span>
params.fragment_mass_tolerance=<span class="org-constant">null</span>
</pre>
</div>
<p>
Notably, a more stringent (lower) tolerance increases memorary usage by comet.
</p>


<p>
the maximum number of allowed missed cleavages is also passed to all search engines.
<b>Mandatory</b> if you want to use these search engines.
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">Int, if you are using the comet, this can by at maximum 5,</span>
params.max_missed_cleavages=<span class="org-constant">null</span>
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow">max_missed_cleavages = <span class="org-type">Channel</span>.value(params.max_missed_cleavages)
</pre>
</div>

<p>
However, in this developers experience, <code>xtandem</code> will crash when using another <code>max-missed-cleavages</code> then 1 so you would
put e.g. the following in your config file.
</p>
<div class="org-src-container">
<pre class="src src-nextflow">params.search_engines = [<span class="org-string">"comet"</span>]
params.max_missed_cleavages= 2
</pre>
</div>
</div>
</div>

<div id="outline-container-orga6e8b6b" class="outline-5">
<h5 id="orga6e8b6b"><span class="section-number-5">5.5.1.5.</span> Comet</h5>
<div class="outline-text-5" id="text-5-5-1-5">
<p>


<a href="https://github.com/UWPR/Comet">https://github.com/UWPR/Comet</a>
</p>

<p>
The following  is the template file for the parameters passed to comet.
you can change fields for things that we don't give parameters for (so where the value is not "@&#x2026;@"),
in order to change behaviour of comet specific to your use case.
For more information, see comet's documentation: <a href="https://comet-ms.sourceforge.net/parameters/parameters_202101/">https://comet-ms.sourceforge.net/parameters/parameters_202101/</a>
</p>
<div class="org-src-container">
<pre class="src src-conf"><span class="org-comment-delimiter"># </span><span class="org-comment">comet_version 2022.01 rev. 0</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">Comet MS/MS search engine parameters file.</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">Everything following the '#' symbol is treated as a comment.</span>

<span class="org-variable-name">database_name</span> = @DDA_DB_FILE@
<span class="org-variable-name">decoy_search</span> = 0                       <span class="org-comment-delimiter"># </span><span class="org-comment">0=no (default), 1=concatenated search, 2=separate search</span>
<span class="org-variable-name">peff_format</span> = 0                        <span class="org-comment-delimiter"># </span><span class="org-comment">0=no (normal fasta, default), 1=PEFF PSI-MOD, 2=PEFF Unimod</span>
<span class="org-variable-name">peff_obo</span> =                             <span class="org-comment-delimiter"># </span><span class="org-comment">path to PSI Mod or Unimod OBO file</span>

<span class="org-variable-name">num_threads</span> = 0                        <span class="org-comment-delimiter"># </span><span class="org-comment">0=poll CPU to set num threads; else specify num threads directly (max 128)</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">masses</span>
<span class="org-comment-delimiter">#</span>
<span class="org-variable-name">peptide_mass_tolerance</span> = @PRECURSOR_MASS_TOLERANCE@
<span class="org-variable-name">peptide_mass_units</span> = 2                 <span class="org-comment-delimiter"># </span><span class="org-comment">0=amu, 1=mmu, 2=ppm</span>
<span class="org-variable-name">mass_type_parent</span> = 1                   <span class="org-comment-delimiter"># </span><span class="org-comment">0=average masses, 1=monoisotopic masses</span>
<span class="org-variable-name">mass_type_fragment</span> = 1                 <span class="org-comment-delimiter"># </span><span class="org-comment">0=average masses, 1=monoisotopic masses</span>
<span class="org-variable-name">precursor_tolerance_type</span> = 1           <span class="org-comment-delimiter"># </span><span class="org-comment">0=MH+ (default), 1=precursor m/z; only valid for amu/mmu tolerances</span>
<span class="org-variable-name">isotope_error</span> = 3                      <span class="org-comment-delimiter"># </span><span class="org-comment">0=off, 1=0/1 (C13 error), 2=0/1/2, 3=0/1/2/3, 4=-8/-4/0/4/8 (for +4/+8 labeling)</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">search enzyme</span>
<span class="org-comment-delimiter">#</span>
<span class="org-variable-name">search_enzyme_number</span> = 1               <span class="org-comment-delimiter"># </span><span class="org-comment">choose from list at end of this params file</span>
<span class="org-variable-name">search_enzyme2_number</span> = 0              <span class="org-comment-delimiter"># </span><span class="org-comment">second enzyme; set to 0 if no second enzyme</span>
<span class="org-variable-name">num_enzyme_termini</span> = 2                 <span class="org-comment-delimiter"># </span><span class="org-comment">1 (semi-digested), 2 (fully digested, default), 8 C-term unspecific , 9 N-term unspecific</span>
<span class="org-variable-name">allowed_missed_cleavage</span> = @MAX_MISSED_CLEAVAGES@            <span class="org-comment-delimiter"># </span><span class="org-comment">maximum value is 5; for enzyme search</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">Up to 9 variable modifications are supported</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">format:  &lt;mass&gt; &lt;residues&gt; &lt;0=variable/else binary&gt; &lt;max_mods_per_peptide&gt; &lt;term_distance&gt; &lt;n/c-term&gt; &lt;required&gt; &lt;neutral_loss&gt;</span>
<span class="org-comment-delimiter">#     </span><span class="org-comment">e.g. 79.966331 STY 0 3 -1 0 0 97.976896</span>
<span class="org-comment-delimiter">#</span>
<span class="org-variable-name">variable_mod01</span> = 15.9949 M 0 3 -1 0 0 0.0
<span class="org-variable-name">variable_mod02</span> = 0.0 X 0 3 -1 0 0 0.0
<span class="org-variable-name">variable_mod03</span> = 0.0 X 0 3 -1 0 0 0.0
<span class="org-variable-name">variable_mod04</span> = 0.0 X 0 3 -1 0 0 0.0
<span class="org-variable-name">variable_mod05</span> = 0.0 X 0 3 -1 0 0 0.0
<span class="org-variable-name">variable_mod06</span> = 0.0 X 0 3 -1 0 0 0.0
<span class="org-variable-name">variable_mod07</span> = 0.0 X 0 3 -1 0 0 0.0
<span class="org-variable-name">variable_mod08</span> = 0.0 X 0 3 -1 0 0 0.0
<span class="org-variable-name">variable_mod09</span> = 0.0 X 0 3 -1 0 0 0.0
<span class="org-variable-name">max_variable_mods_in_peptide</span> = 5
<span class="org-variable-name">require_variable_mod</span> = 0

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">fragment ions</span>
<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">ion trap ms/ms:  1.0005 tolerance, 0.4 offset (mono masses), theoretical_fragment_ions = 1</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">high res ms/ms:    0.02 tolerance, 0.0 offset (mono masses), theoretical_fragment_ions = 0, spectrum_batch_size = 15000</span>
<span class="org-comment-delimiter">#</span>
<span class="org-variable-name">fragment_bin_tol</span> = @FRAGMENT_MASS_TOLERANCE@              <span class="org-comment-delimiter"># </span><span class="org-comment">binning to use on fragment ions</span>
<span class="org-variable-name">fragment_bin_offset</span> = 0.0              <span class="org-comment-delimiter"># </span><span class="org-comment">offset position to start the binning (0.0 to 1.0)</span>
<span class="org-variable-name">theoretical_fragment_ions</span> = 1          <span class="org-comment-delimiter"># </span><span class="org-comment">0=use flanking peaks, 1=M peak only</span>
<span class="org-variable-name">use_A_ions</span> = 0
<span class="org-variable-name">use_B_ions</span> = 1
<span class="org-variable-name">use_C_ions</span> = 0
<span class="org-variable-name">use_X_ions</span> = 0
<span class="org-variable-name">use_Y_ions</span> = 1
<span class="org-variable-name">use_Z_ions</span> = 0
<span class="org-variable-name">use_Z1_ions</span> = 0
<span class="org-variable-name">use_NL_ions</span> = 0                        <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=yes to consider NH3/H2O neutral loss peaks</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">output</span>
<span class="org-comment-delimiter">#</span>
<span class="org-variable-name">output_sqtfile</span> = 0                     <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=yes  write sqt file</span>
<span class="org-variable-name">output_txtfile</span> = 0                     <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=yes  write tab-delimited txt file</span>
<span class="org-variable-name">output_pepxmlfile</span> = 1                  <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=yes  write pepXML file</span>
<span class="org-variable-name">output_mzidentmlfile</span> = 0               <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=yes  write mzIdentML file</span>
<span class="org-variable-name">output_percolatorfile</span> = 1              <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=yes  write Percolator pin file</span>
<span class="org-variable-name">print_expect_score</span> = 1                 <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=yes to replace Sp with expect in out &amp; sqt</span>
<span class="org-variable-name">num_output_lines</span> = 5                   <span class="org-comment-delimiter"># </span><span class="org-comment">num peptide results to show</span>

<span class="org-variable-name">sample_enzyme_number</span> = 1               <span class="org-comment-delimiter"># </span><span class="org-comment">Sample enzyme which is possibly different than the one applied to the search.</span>
                                       <span class="org-comment-delimiter"># </span><span class="org-comment">Used to calculate NTT &amp; NMC in pepXML output (default=1 for trypsin).</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">mzXML parameters</span>
<span class="org-comment-delimiter">#</span>
<span class="org-variable-name">scan_range</span> = 0 0                       <span class="org-comment-delimiter"># </span><span class="org-comment">start and end scan range to search; either entry can be set independently</span>
<span class="org-variable-name">precursor_charge</span> = 0 0                 <span class="org-comment-delimiter"># </span><span class="org-comment">precursor charge range to analyze; does not override any existing charge; 0 as 1st entry ignores parameter</span>
<span class="org-variable-name">override_charge</span> = 0                    <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=override precursor charge states, 2=ignore precursor charges outside precursor_charge range, 3=see online</span>
<span class="org-variable-name">ms_level</span> = 2                           <span class="org-comment-delimiter"># </span><span class="org-comment">MS level to analyze, valid are levels 2 (default) or 3</span>
<span class="org-variable-name">activation_method</span> = HCD                <span class="org-comment-delimiter"># </span><span class="org-comment">activation method; used if activation method set; allowed ALL, CID, ECD, ETD, ETD+SA, PQD, HCD, IRMPD, SID</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">misc parameters</span>
<span class="org-comment-delimiter">#</span>
<span class="org-variable-name">digest_mass_range</span> = 600.0 5000.0       <span class="org-comment-delimiter"># </span><span class="org-comment">MH+ peptide mass range to analyze</span>
<span class="org-variable-name">peptide_length_range</span> = 5 63            <span class="org-comment-delimiter"># </span><span class="org-comment">minimum and maximum peptide length to analyze (default 1 63; max length 63)</span>
<span class="org-variable-name">num_results</span> = 100                      <span class="org-comment-delimiter"># </span><span class="org-comment">number of search hits to store internally</span>
<span class="org-variable-name">max_duplicate_proteins</span> = 20            <span class="org-comment-delimiter"># </span><span class="org-comment">maximum number of additional duplicate protein names to report for each peptide ID; -1 reports all duplicates</span>
<span class="org-variable-name">max_fragment_charge</span> = 3                <span class="org-comment-delimiter"># </span><span class="org-comment">set maximum fragment charge state to analyze (allowed max 5)</span>
<span class="org-variable-name">max_precursor_charge</span> = 6               <span class="org-comment-delimiter"># </span><span class="org-comment">set maximum precursor charge state to analyze (allowed max 9)</span>
<span class="org-variable-name">nucleotide_reading_frame</span> = 0           <span class="org-comment-delimiter"># </span><span class="org-comment">0=proteinDB, 1-6, 7=forward three, 8=reverse three, 9=all six</span>
<span class="org-variable-name">clip_nterm_methionine</span> = 0              <span class="org-comment-delimiter"># </span><span class="org-comment">0=leave protein sequences as-is; 1=also consider sequence w/o N-term methionine</span>
<span class="org-variable-name">spectrum_batch_size</span> = 15000            <span class="org-comment-delimiter"># </span><span class="org-comment">max. # of spectra to search at a time; 0 to search the entire scan range in one loop</span>
<span class="org-variable-name">decoy_prefix</span> = DECOY_                  <span class="org-comment-delimiter"># </span><span class="org-comment">decoy entries are denoted by this string which is pre-pended to each protein accession</span>
<span class="org-variable-name">equal_I_and_L</span> = 1                      <span class="org-comment-delimiter"># </span><span class="org-comment">0=treat I and L as different; 1=treat I and L as same</span>
<span class="org-variable-name">output_suffix</span> =                        <span class="org-comment-delimiter"># </span><span class="org-comment">add a suffix to output base names i.e. suffix "-C" generates base-C.pep.xml from base.mzXML input</span>
<span class="org-variable-name">mass_offsets</span> =                         <span class="org-comment-delimiter"># </span><span class="org-comment">one or more mass offsets to search (values substracted from deconvoluted precursor mass)</span>
<span class="org-variable-name">precursor_NL_ions</span> =                    <span class="org-comment-delimiter"># </span><span class="org-comment">one or more precursor neutral loss masses, will be added to xcorr analysis</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">spectral processing</span>
<span class="org-comment-delimiter">#</span>
<span class="org-variable-name">minimum_peaks</span> = 10                     <span class="org-comment-delimiter"># </span><span class="org-comment">required minimum number of peaks in spectrum to search (default 10)</span>
<span class="org-variable-name">minimum_intensity</span> = 0                  <span class="org-comment-delimiter"># </span><span class="org-comment">minimum intensity value to read in</span>
<span class="org-variable-name">remove_precursor_peak</span> = 0              <span class="org-comment-delimiter"># </span><span class="org-comment">0=no, 1=yes, 2=all charge reduced precursor peaks (for ETD), 3=phosphate neutral loss peaks</span>
<span class="org-variable-name">remove_precursor_tolerance</span> = 1.5       <span class="org-comment-delimiter"># </span><span class="org-comment">+- Da tolerance for precursor removal</span>
<span class="org-variable-name">clear_mz_range</span> = 0.0 0.0               <span class="org-comment-delimiter"># </span><span class="org-comment">for iTRAQ/TMT type data; will clear out all peaks in the specified m/z range</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">additional modifications</span>
<span class="org-comment-delimiter">#</span>

<span class="org-variable-name">add_Cterm_peptide</span> = 0.0
<span class="org-variable-name">add_Nterm_peptide</span> = 0.0
<span class="org-variable-name">add_Cterm_protein</span> = 0.0
<span class="org-variable-name">add_Nterm_protein</span> = 0.0

<span class="org-variable-name">add_G_glycine</span> = 0.0000                 <span class="org-comment-delimiter"># </span><span class="org-comment">added to G - avg.  57.0513, mono.  57.02146</span>
<span class="org-variable-name">add_A_alanine</span> = 0.0000                 <span class="org-comment-delimiter"># </span><span class="org-comment">added to A - avg.  71.0779, mono.  71.03711</span>
<span class="org-variable-name">add_S_serine</span> = 0.0000                  <span class="org-comment-delimiter"># </span><span class="org-comment">added to S - avg.  87.0773, mono.  87.03203</span>
<span class="org-variable-name">add_P_proline</span> = 0.0000                 <span class="org-comment-delimiter"># </span><span class="org-comment">added to P - avg.  97.1152, mono.  97.05276</span>
<span class="org-variable-name">add_V_valine</span> = 0.0000                  <span class="org-comment-delimiter"># </span><span class="org-comment">added to V - avg.  99.1311, mono.  99.06841</span>
<span class="org-variable-name">add_T_threonine</span> = 0.0000               <span class="org-comment-delimiter"># </span><span class="org-comment">added to T - avg. 101.1038, mono. 101.04768</span>
<span class="org-variable-name">add_C_cysteine</span> = 57.021464             <span class="org-comment-delimiter"># </span><span class="org-comment">added to C - avg. 103.1429, mono. 103.00918</span>
<span class="org-variable-name">add_L_leucine</span> = 0.0000                 <span class="org-comment-delimiter"># </span><span class="org-comment">added to L - avg. 113.1576, mono. 113.08406</span>
<span class="org-variable-name">add_I_isoleucine</span> = 0.0000              <span class="org-comment-delimiter"># </span><span class="org-comment">added to I - avg. 113.1576, mono. 113.08406</span>
<span class="org-variable-name">add_N_asparagine</span> = 0.0000              <span class="org-comment-delimiter"># </span><span class="org-comment">added to N - avg. 114.1026, mono. 114.04293</span>
<span class="org-variable-name">add_D_aspartic_acid</span> = 0.0000           <span class="org-comment-delimiter"># </span><span class="org-comment">added to D - avg. 115.0874, mono. 115.02694</span>
<span class="org-variable-name">add_Q_glutamine</span> = 0.0000               <span class="org-comment-delimiter"># </span><span class="org-comment">added to Q - avg. 128.1292, mono. 128.05858</span>
<span class="org-variable-name">add_K_lysine</span> = 0.0000                  <span class="org-comment-delimiter"># </span><span class="org-comment">added to K - avg. 128.1723, mono. 128.09496</span>
<span class="org-variable-name">add_E_glutamic_acid</span> = 0.0000           <span class="org-comment-delimiter"># </span><span class="org-comment">added to E - avg. 129.1140, mono. 129.04259</span>
<span class="org-variable-name">add_M_methionine</span> = 0.0000              <span class="org-comment-delimiter"># </span><span class="org-comment">added to M - avg. 131.1961, mono. 131.04048</span>
<span class="org-variable-name">add_H_histidine</span> = 0.0000               <span class="org-comment-delimiter"># </span><span class="org-comment">added to H - avg. 137.1393, mono. 137.05891</span>
<span class="org-variable-name">add_F_phenylalanine</span> = 0.0000           <span class="org-comment-delimiter"># </span><span class="org-comment">added to F - avg. 147.1739, mono. 147.06841</span>
<span class="org-variable-name">add_U_selenocysteine</span> = 0.0000          <span class="org-comment-delimiter"># </span><span class="org-comment">added to U - avg. 150.0379, mono. 150.95363</span>
<span class="org-variable-name">add_R_arginine</span> = 0.0000                <span class="org-comment-delimiter"># </span><span class="org-comment">added to R - avg. 156.1857, mono. 156.10111</span>
<span class="org-variable-name">add_Y_tyrosine</span> = 0.0000                <span class="org-comment-delimiter"># </span><span class="org-comment">added to Y - avg. 163.0633, mono. 163.06333</span>
<span class="org-variable-name">add_W_tryptophan</span> = 0.0000              <span class="org-comment-delimiter"># </span><span class="org-comment">added to W - avg. 186.0793, mono. 186.07931</span>
<span class="org-variable-name">add_O_pyrrolysine</span> = 0.0000             <span class="org-comment-delimiter"># </span><span class="org-comment">added to O - avg. 237.2982, mono  237.14773</span>
<span class="org-variable-name">add_B_user_amino_acid</span> = 0.0000         <span class="org-comment-delimiter"># </span><span class="org-comment">added to B - avg.   0.0000, mono.   0.00000</span>
<span class="org-variable-name">add_J_user_amino_acid</span> = 0.0000         <span class="org-comment-delimiter"># </span><span class="org-comment">added to J - avg.   0.0000, mono.   0.00000</span>
<span class="org-variable-name">add_X_user_amino_acid</span> = 0.0000         <span class="org-comment-delimiter"># </span><span class="org-comment">added to X - avg.   0.0000, mono.   0.00000</span>
<span class="org-variable-name">add_Z_user_amino_acid</span> = 0.0000         <span class="org-comment-delimiter"># </span><span class="org-comment">added to Z - avg.   0.0000, mono.   0.00000</span>

<span class="org-comment-delimiter">#</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">COMET_ENZYME_INFO _must_ be at the end of this parameters file</span>
<span class="org-comment-delimiter">#</span>
[<span class="org-type">COMET_ENZYME_INFO</span>]
0.  Cut_everywhere         0      -           -
1.  Trypsin                1      KR          P
2.  Trypsin/P              1      KR          -
3.  Lys_C                  1      K           P
4.  Lys_N                  0      K           -
5.  Arg_C                  1      R           P
6.  Asp_N                  0      D           -
7.  CNBr                   1      M           -
8.  Glu_C                  1      DE          P
9.  PepsinA                1      FL          P
10. Chymotrypsin           1      FWYL        P
11. No_cut                 1      @           @
</pre>
</div>

<p>
Here we set the above as the default parameter template.
If you customized or have your own the comet config,
you can point to it with this.
</p>
<div class="org-src-container">
<pre class="src src-nextflow">params.comet_template=<span class="org-string">"comet_template.txt"</span>
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow">
<span class="org-keyword">process</span> <span class="org-function-name">MakeCometConfig</span> {
    <span class="org-comment-delimiter">// </span><span class="org-comment">should we instead return a tuple here of fastadb and config</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">because the config.txt refers to it?</span>
    input:
    <span class="org-keyword">val</span> <span class="org-function-name">max_missed_cleavages</span>
    <span class="org-keyword">file</span> <span class="org-function-name">fastadb_with_decoy</span> from joined_fasta_with_decoys_ch.first()
    path template from <span class="org-type">Channel</span>.fromPath(params.comet_template)
    output:
    file <span class="org-string">"comet_config.txt"</span> into comet_config_ch
    <span class="org-string">"""</span>
<span class="org-string">    sed 's/@DDA_DB_FILE@/</span><span class="org-variable-name">$fastadb_with_decoy</span><span class="org-string">/g;s/@FRAGMENT_MASS_TOLERANCE@/</span><span class="org-variable-name">$params.fragment_mass_tolerance</span><span class="org-string">/g;s/@PRECURSOR_MASS_TOLERANCE@/</span><span class="org-variable-name">$params.precursor_mass_tolerance</span><span class="org-string">/g;s/@MAX_MISSED_CLEAVAGES@/</span><span class="org-variable-name">$max_missed_cleavages</span><span class="org-string">/g' </span><span class="org-variable-name">$template</span><span class="org-string"> &gt; comet_config.txt </span>
<span class="org-string">    """</span>

}
</pre>
</div>


<p>
setting memory &amp; error strategy like this prevents caching
even with process.cache='lenient'
maybe because the task.attempt = 1
is tried first 
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">Comet</span> {
    <span class="org-comment-delimiter">// </span><span class="org-comment">we probably also want to publish thees</span>
    memory { 5.GB * 2 *  task.attempt }
    errorStrategy { task.exitStatus <span class="org-keyword">in</span> 137..137 ? <span class="org-string">'retry'</span> : <span class="org-string">'terminate'</span> }
    maxRetries 2
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">comet_config</span> from comet_config_ch.first()
    <span class="org-comment-delimiter">// </span><span class="org-comment">future dev: we can .mix with DDA here?</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">though we might need to tag for DDA / Pseudo</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">so that xinteract </span>
    <span class="org-keyword">file</span> <span class="org-function-name">mzxml</span> from maybespectra_ch.comet
    <span class="org-keyword">file</span> <span class="org-function-name">fastadb_with_decoy</span> from joined_fasta_with_decoys_ch.first()
    output:
    file(<span class="org-string">"</span><span class="org-variable-name">${mzxml.baseName}</span><span class="org-string">.pep.xml"</span>) into comet_pepxml_ch
    <span class="org-keyword">file</span> <span class="org-function-name">mzxml</span> into xinteract_comet_mzxml_ch
    when:
    params.search_engines.contains(<span class="org-string">"comet"</span>)

    <span class="org-string">"""</span>
<span class="org-string">    if command -v command-ms;</span>
<span class="org-string">    then</span>
<span class="org-string">      comet-ms -P</span><span class="org-variable-name">$comet_config</span><span class="org-string"> </span><span class="org-variable-name">$mzxml</span>
<span class="org-string">    else</span>
<span class="org-string">      comet -P</span><span class="org-variable-name">$comet_config</span><span class="org-string"> </span><span class="org-variable-name">$mzxml</span>
<span class="org-string">    fi</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">XinteractComet</span> {
    memory <span class="org-string">'16 GB'</span>
    time <span class="org-string">'5h'</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">memory usage scales with the number of input files</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">find the correct usage per input file or size</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">also for xinteractxtandem</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">usage there seems to be a lot smaller</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">as input files seems to be smaller</span>
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">pepxmls</span> from comet_pepxml_ch.toSortedList()
    <span class="org-comment-delimiter">// </span><span class="org-comment">the filename of needed fastdadb was defined in cometcfg</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">and stored in pepxml in the comet-ms step</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">-a suppplies the absulute path to the data directory where the mzxmls</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">rather than reading wherer the mfrom the xmls</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">where the mzxml are, because its not very</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">nextflow to look outside the cwd.</span>
    <span class="org-keyword">file</span> <span class="org-function-name">fastadb_with_decoy</span> from joined_fasta_with_decoys_ch.first()
    <span class="org-keyword">file</span> <span class="org-function-name">mzxmls</span> from  xinteract_comet_mzxml_ch.toSortedList()
    output: 
    file <span class="org-string">"interact_comet.pep.xml"</span> into comet_search_results_ch
    when:
    pepxmls.size() &gt; 0
    <span class="org-string">"""</span>
<span class="org-string">    xinteract -a\$PWD -OARPd -dDECOY_ -Ninteract_comet.pep.xml </span><span class="org-variable-name">$pepxmls</span>
<span class="org-string">    """</span>
}
</pre>
</div>
</div>
</div>



<div id="outline-container-org6868709" class="outline-5">
<h5 id="org6868709"><span class="section-number-5">5.5.1.6.</span> Xtandem</h5>
<div class="outline-text-5" id="text-5-5-1-6">
<p>


</p>

<p>
As with tandem, you an adjust the file
"xtandem-template.xml" to suit your needs,
values with <code>@...@</code> are automatically replaced.
See also the documentation of xtandem here: <a href="https://www.thegpm.org/TANDEM/api/">https://www.thegpm.org/TANDEM/api/</a>
</p>

<div class="org-src-container">
<pre class="src src-xml" id="org72d1e02"><span class="org-nxml-processing-instruction-delimiter">&lt;?</span><span class="org-nxml-processing-instruction-target">xml</span> <span class="org-nxml-processing-instruction-content">version="1.0"</span><span class="org-nxml-processing-instruction-delimiter">?&gt;</span>
<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">bioml</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"x! taxon-to-file matching list"</span><span class="org-nxml-tag-delimiter">&gt;</span>
  <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">taxon</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"DB"</span><span class="org-nxml-tag-delimiter">&gt;</span>
    <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">file</span> <span class="org-nxml-attribute-local-name">format</span>=<span class="org-string">"peptide"</span> <span class="org-nxml-attribute-local-name">URL</span>=<span class="org-string">"%s"</span> <span class="org-nxml-tag-slash">/</span><span class="org-nxml-tag-delimiter">&gt;</span>
  <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">taxon</span><span class="org-nxml-tag-delimiter">&gt;</span>
<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">bioml</span><span class="org-nxml-tag-delimiter">&gt;</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-xml" id="org8fafefd"><span class="org-nxml-processing-instruction-delimiter">&lt;?</span><span class="org-nxml-processing-instruction-target">xml</span> <span class="org-nxml-processing-instruction-content">version="1.0"</span><span class="org-nxml-processing-instruction-delimiter">?&gt;</span>
<span class="org-nxml-processing-instruction-delimiter">&lt;?</span><span class="org-nxml-processing-instruction-target">xml-stylesheet</span> <span class="org-nxml-processing-instruction-content">type="text/xsl" href="tandem-input-style.xsl"</span><span class="org-nxml-processing-instruction-delimiter">?&gt;</span>
<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">bioml</span><span class="org-nxml-tag-delimiter">&gt;</span>
<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">list path parameters</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">spectrum parameters</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, fragment monoisotopic mass error"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">@FRAGMENT_MASS_TOLERANCE@</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, parent monoisotopic mass error plus"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">@PRECURSOR_MASS_TOLERANCE@</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, parent monoisotopic mass error minus"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">@PRECURSOR_MASS_TOLERANCE@</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, parent monoisotopic mass isotope error"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, fragment monoisotopic mass error units"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">Daltons</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The value for this parameter may be 'Daltons' or 'ppm': all other values are ignored</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, parent monoisotopic mass error units"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">ppm</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The value for this parameter may be 'Daltons' or 'ppm': all other values are ignored</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, fragment mass type"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">monoisotopic</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values are monoisotopic|average </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">spectrum conditioning parameters</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, dynamic range"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">100.0</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The peaks read in are normalized so that the most intense peak</span>
<span class="org-nxml-text">                is set to the dynamic range value. All peaks with values of less that</span>
<span class="org-nxml-text">                1, using this normalization, are not used. This normalization has the</span>
<span class="org-nxml-text">                overall effect of setting a threshold value for peak intensities.</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, total peaks"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">50</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span> 
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">If this value is 0, it is ignored. If it is greater than zero (lets say 50),</span>
<span class="org-nxml-text">                then the number of peaks in the spectrum with be limited to the 50 most intense</span>
<span class="org-nxml-text">                peaks in the spectrum. X! tandem does not do any peak finding: it only</span>
<span class="org-nxml-text">                limits the peaks used by this parameter, and the dynamic range parameter.</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, maximum parent charge"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">4</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, use noise suppression"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, minimum parent m+h"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">500.0</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, minimum fragment mz"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">150.0</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, minimum peaks"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">15</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span> 
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, threads"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">40</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, sequence batch size"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">1000</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">residue modification parameters</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"residue, modification mass"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">57.022@C</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The format of this parameter is m@X, where m is the modfication</span>
<span class="org-nxml-text">                mass in Daltons and X is the appropriate residue to modify. Lists of</span>
<span class="org-nxml-text">                modifications are separated by commas. For example, to modify M and C</span>
<span class="org-nxml-text">                with the addition of 16.0 Daltons, the parameter line would be</span>
<span class="org-nxml-text">                +16.0@M,+16.0@C</span>
<span class="org-nxml-text">                Positive and negative values are allowed.</span>
<span class="org-nxml-text">                </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"residue, potential modification mass"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">16@M</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The format of this parameter is the same as the format</span>
<span class="org-nxml-text">                for residue, modification mass (see above).</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"residue, potential modification motif"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The format of this parameter is similar to residue, modification mass,</span>
<span class="org-nxml-text">                with the addition of a modified PROSITE notation sequence motif specification.</span>
<span class="org-nxml-text">                For example, a value of 80@[ST!]PX[KR] indicates a modification</span>
<span class="org-nxml-text">                of either S or T when followed by P, and residue and the a K or an R.</span>
<span class="org-nxml-text">                A value of 204@N!{P}[ST]{P} indicates a modification of N by 204, if it</span>
<span class="org-nxml-text">                is NOT followed by a P, then either an S or a T, NOT followed by a P.</span>
<span class="org-nxml-text">                Positive and negative values are allowed.</span>
<span class="org-nxml-text">                </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">protein parameters</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, taxon"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">other mammals</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">This value is interpreted using the information in taxonomy.xml.</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, cleavage site"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">[RK]|{P}</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">this setting corresponds to the enzyme trypsin. The first characters</span>
<span class="org-nxml-text">                in brackets represent residues N-terminal to the bond - the '|' pipe -</span>
<span class="org-nxml-text">                and the second set of characters represent residues C-terminal to the</span>
<span class="org-nxml-text">                bond. The characters must be in square brackets (denoting that only</span>
<span class="org-nxml-text">                these residues are allowed for a cleavage) or french brackets (denoting</span>
<span class="org-nxml-text">                that these residues cannot be in that position). Use UPPERCASE characters.</span>
<span class="org-nxml-text">                To denote cleavage at any residue, use [X]|[X] and reset the </span>
<span class="org-nxml-text">                scoring, maximum missed cleavage site parameter (see below) to something like 50.</span>
<span class="org-nxml-text">                </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, modified residue mass file"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, cleavage C-terminal mass change"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">+17.002735</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, cleavage N-terminal mass change"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">+1.007825</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, N-terminal residue modification mass"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">0.0</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, C-terminal residue modification mass"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">0.0</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, homolog management"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">if yes, an upper limit is set on the number of homologues kept for a particular spectrum</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, quick acetyl"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, quick pyrolidone"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">model refinement parameters</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, modification mass"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, sequence path"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, tic percent"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">20</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, spectrum synthesis"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, maximum valid expectation value"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">0.1</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, potential N-terminus modifications"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">+42.010565@[</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>


        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, potential C-terminus modifications"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, unanticipated cleavage"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, potential modification mass"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, point mutations"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, use potential modifications for full refinement"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, point mutations"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"refine, potential modification motif"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The format of this parameter is similar to residue, modification mass,</span>
<span class="org-nxml-text">                with the addition of a modified PROSITE notation sequence motif specification.</span>
<span class="org-nxml-text">                For example, a value of 80@[ST!]PX[KR] indicates a modification</span>
<span class="org-nxml-text">                of either S or T when followed by P, and residue and the a K or an R.</span>
<span class="org-nxml-text">                A value of 204@N!{P}[ST]{P} indicates a modification of N by 204, if it</span>
<span class="org-nxml-text">                is NOT followed by a P, then either an S or a T, NOT followed by a P.</span>
<span class="org-nxml-text">                Positive and negative values are allowed.</span>
<span class="org-nxml-text">                </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">scoring parameters</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, minimum ion count"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">4</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, maximum missed cleavage sites"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">@MAX_MISSED_CLEAVAGES@</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, x ions"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, y ions"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, z ions"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, a ions"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, b ions"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, c ions"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, cyclic permutation"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">if yes, cyclic peptide sequence permutation is used to pad the scoring histograms</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, include reverse"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">if yes, then reversed sequences are searched at the same time as forward sequences</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, cyclic permutation"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"scoring, include reverse"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">output parameters</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, log path"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, message"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">testing 1 2 3</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, one sequence copy"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, sequence path"</span><span class="org-nxml-tag-delimiter">&gt;&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, path"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">output.xml</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, sort results by"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">protein</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = protein|spectrum (spectrum is the default)</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, path hashing"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = yes|no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, xsl path"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">tandem-style.xsl</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, parameters"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = yes|no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, performance"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = yes|no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, spectra"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = yes|no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, histograms"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = yes|no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, proteins"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = yes|no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, sequences"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">yes</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = yes|no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, one sequence copy"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">no</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = yes|no, set to yes to produce only one copy of each protein sequence in the output xml</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, results"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">valid</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values = all|valid|stochastic</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, maximum valid expectation value"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">0.1</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">value is used in the valid|stochastic setting of output, results</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, histogram column width"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">30</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
                <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">values any integer greater than 0. Setting this to '1' makes cutting and pasting histograms</span>
<span class="org-nxml-text">                into spread sheet programs easier.</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"description"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">ADDITIONAL EXPLANATIONS</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"description"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">Each one of the parameters for X! tandem is entered as a labeled note</span>
<span class="org-nxml-text">                        node. In the current version of X!, keep those note nodes</span>
<span class="org-nxml-text">                        on a single line.</span>
<span class="org-nxml-text">        </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"description"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The presence of the type 'input' is necessary if a note is to be considered</span>
<span class="org-nxml-text">                        an input parameter.</span>
<span class="org-nxml-text">        </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"description"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">Any of the parameters that are paths to files may require alteration for a </span>
<span class="org-nxml-text">                        particular installation. Full path names usually cause the least trouble,</span>
<span class="org-nxml-text">                        but there is no reason not to use relative path names, if that is the</span>
<span class="org-nxml-text">                        most convenient.</span>
<span class="org-nxml-text">        </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"description"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">Any parameter values set in the 'list path, default parameters' file are</span>
<span class="org-nxml-text">                        reset by entries in the normal input file, if they are present. Otherwise,</span>
<span class="org-nxml-text">                        the default set is used.</span>
<span class="org-nxml-text">        </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"description"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The 'list path, taxonomy information' file must exist.</span>
<span class="org-nxml-text">                </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"description"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The directory containing the 'output, path' file must exist: it will not be created.</span>
<span class="org-nxml-text">                </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"description"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">The 'output, xsl path' is optional: it is only of use if a good XSLT style sheet exists.</span>
<span class="org-nxml-text">                </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span> 
<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">bioml</span><span class="org-nxml-tag-delimiter">&gt;</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-xml" id="org23ae3ea"><span class="org-nxml-processing-instruction-delimiter">&lt;?</span><span class="org-nxml-processing-instruction-target">xml</span> <span class="org-nxml-processing-instruction-content">version="1.0"</span><span class="org-nxml-processing-instruction-delimiter">?&gt;</span>
<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">bioml</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
<span class="org-nxml-text">        Each one of the parameters for x! tandem is entered as a labeled note node. </span>
<span class="org-nxml-text">        Any of the entries in the default_input.xml file can be over-ridden by</span>
<span class="org-nxml-text">        adding a corresponding entry to this file. This file represents a minimum</span>
<span class="org-nxml-text">        input file, with only entries for the default settings, the output file</span>
<span class="org-nxml-text">        and the input spectra file name. </span>
<span class="org-nxml-text">        See the taxonomy.xml file for a description of how FASTA sequence list </span>
<span class="org-nxml-text">        files are linked to a taxon name.</span>
<span class="org-nxml-text">        </span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"list path, default parameters"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">%s</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"list path, taxonomy information"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">%s</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"protein, taxon"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">DB</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"spectrum, path"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">%s</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>

        <span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-element-local-name">note</span> <span class="org-nxml-attribute-local-name">type</span>=<span class="org-string">"input"</span> <span class="org-nxml-attribute-local-name">label</span>=<span class="org-string">"output, path"</span><span class="org-nxml-tag-delimiter">&gt;</span><span class="org-nxml-text">%s</span><span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">note</span><span class="org-nxml-tag-delimiter">&gt;</span>
<span class="org-nxml-tag-delimiter">&lt;</span><span class="org-nxml-tag-slash">/</span><span class="org-nxml-element-local-name">bioml</span><span class="org-nxml-tag-delimiter">&gt;</span>
</pre>
</div>


<p>
We are making the <code>xtandem_taxonomy</code> xml in the same process because its kinda a pseudo dependency 
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">MakeXtandemConfig</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">template</span> from <span class="org-type">Channel</span>.fromPath(params.xtandem_template)
    <span class="org-keyword">file</span> <span class="org-function-name">fastadb_with_decoy</span> from joined_fasta_with_decoys_ch.first()
    <span class="org-keyword">val</span> <span class="org-function-name">max_missed_cleavages</span>
    output:
    file <span class="org-string">"xtandem_config.xml"</span> into xtandem_config_ch
    <span class="org-string">"""</span>
<span class="org-string">    sed 's/@DDA_DB_FILE@/</span><span class="org-variable-name">$fastadb_with_decoy</span><span class="org-string">/g;s/@FRAGMENT_MASS_TOLERANCE@/</span><span class="org-variable-name">$params.fragment_mass_tolerance</span><span class="org-string">/g;s/@PRECURSOR_MASS_TOLERANCE@/</span><span class="org-variable-name">$params.precursor_mass_tolerance</span><span class="org-string">/g;s/@MAX_MISSED_CLEAVAGES@/</span><span class="org-variable-name">$max_missed_cleavages</span><span class="org-string">/g' </span><span class="org-variable-name">$template</span><span class="org-string"> &gt; xtandem_config.xml</span>
<span class="org-string">    """</span>
}


<span class="org-keyword">process</span> <span class="org-function-name">XTandem</span> {
    when:
    params.search_engines.contains(<span class="org-string">"xtandem"</span>)

    input:
    <span class="org-keyword">file</span> <span class="org-function-name">mzxml</span> from maybespectra_ch.xtandem
    <span class="org-keyword">file</span> <span class="org-function-name">tandem_config</span> from xtandem_config_ch.first()
    <span class="org-keyword">file</span> <span class="org-function-name">fastadb_with_decoy</span> from joined_fasta_with_decoys_ch.first()
    output:
    file(<span class="org-string">"</span><span class="org-variable-name">${mzxml.baseName}</span><span class="org-string">.tandem.pep.xml"</span>) into xtandem_pepxml_ch
    <span class="org-keyword">file</span> <span class="org-function-name">mzxml</span> into xinteract_xtandem_mzxml_ch
    <span class="org-string">"""</span>
<span class="org-string">    printf '</span>
<span class="org-string">    &lt;&lt;taxonomy-template&gt;&gt;'  </span><span class="org-variable-name">$fastadb_with_decoy</span><span class="org-string"> | tail -n+2 &gt; xtandem_taxonomy.xml</span>

<span class="org-string">    printf '</span>
<span class="org-string">    &lt;&lt;xtandem-input-template&gt;&gt;' </span><span class="org-variable-name">$tandem_config</span><span class="org-string"> xtandem_taxonomy.xml </span><span class="org-variable-name">$mzxml</span><span class="org-string"> </span><span class="org-variable-name">${mzxml.baseName}</span><span class="org-string">.TANDEM.OUTPUT.xml | tail -n+2 &gt; input.xml</span>
<span class="org-string">    tandem input.xml</span>
<span class="org-string">    Tandem2XML </span><span class="org-variable-name">${mzxml.baseName}</span><span class="org-string">.TANDEM.OUTPUT.xml </span><span class="org-variable-name">${mzxml.baseName}</span><span class="org-string">.tandem.pep.xml </span>
<span class="org-string">    """</span>
}

<span class="org-keyword">process</span> <span class="org-function-name">XinteractXTandem</span> {
    memory <span class="org-string">'16 GB'</span>
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">pepxmls</span> from xtandem_pepxml_ch.toSortedList()
    <span class="org-comment-delimiter">// </span><span class="org-comment">the filename of needed fastdadb was defined in cometcfg</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">and stored in pepxml in the comet-ms step</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">-a suppplies the absulute path to the data directory where the mzxmls</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">rather than reading wherer the mfrom the xmls</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">where the mzxml are, because its not very</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">nextflow to look outside the cwd.</span>
    <span class="org-keyword">file</span> <span class="org-function-name">fastadb_with_decoy</span> from joined_fasta_with_decoys_ch.first()
    <span class="org-keyword">file</span> <span class="org-function-name">mzxmls</span> from  xinteract_xtandem_mzxml_ch.toSortedList()
    output: 
    file <span class="org-string">"interact_xtandem.pep.xml"</span> into xtandem_search_results_ch
    when:
    pepxmls.size() &gt; 0 
    <span class="org-string">"""</span>
<span class="org-string">    xinteract -a\$PWD -OARPd -dDECOY_ -Ninteract_xtandem.pep.xml </span><span class="org-variable-name">$pepxmls</span>
<span class="org-string">    """</span>
}

</pre>
</div>

<p>
If you customized or have your own the xtandem template,
you can point to it with this.
</p>
<div class="org-src-container">
<pre class="src src-nextflow">params.xtandem_template=<span class="org-string">"xtandem-template.xml"</span>
</pre>
</div>
<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-right" />

<col  class="org-right" />
</colgroup>
<tbody>
<tr>
<td class="org-right"># pepxml</td>
<td class="org-right">size pepxml (GiB)</td>
</tr>

<tr>
<td class="org-right">69</td>
<td class="org-right">0.7890625</td>
</tr>

<tr>
<td class="org-right">&#xa0;</td>
<td class="org-right">&#xa0;</td>
</tr>
</tbody>
</table>
<p>
Why is this much smaller than comet?
</p>


<p>
We patch tpp to accept comet 2019015 
</p>

<div class="org-src-container">
<pre class="src src-diff"><span class="org-diff-header">diff -Naur release_5-2-0/extern/Makefile release_5-2-0_mod/extern/Makefile</span>
<span class="org-diff-header">--- </span><span class="org-diff-header"><span class="org-diff-file-header">release_5-2-0/extern/Makefile</span></span><span class="org-diff-header">       2020-07-29 23:43:45.483620066 +0300</span>
<span class="org-diff-header">+++ </span><span class="org-diff-header"><span class="org-diff-file-header">release_5-2-0_mod/extern/Makefile</span></span><span class="org-diff-header">   2020-07-29 23:47:41.796860274 +0300</span>
<span class="org-diff-hunk-header">@@ -339,7 +339,7 @@</span>
<span class="org-diff-context"> #</span>
<span class="org-diff-context"> # http://comet-ms.sourceforge.net/</span>
<span class="org-diff-context"> #</span>
<span class="org-diff-indicator-removed">-</span><span class="org-diff-removed">COMET_VER := 2018014</span>
<span class="org-diff-indicator-added">+</span><span class="org-diff-added">COMET_VER := 2019015</span>
<span class="org-diff-context"> COMET_ZIP := $(TPP_EXT)/comet_source_$(COMET_VER).zip</span>
<span class="org-diff-context"> COMET_SRC := $(BUILD_SRC)/comet_source_$(COMET_VER)</span>
 .PHONY: comet comet-source comet-clean
</pre>
</div>
</div>
</div>

<div id="outline-container-org1a37a38" class="outline-5">
<h5 id="org1a37a38"><span class="section-number-5">5.5.1.7.</span> Joining Comet &amp; Xtandem</h5>
<div class="outline-text-5" id="text-5-5-1-7">
<p>


</p>
<div class="org-src-container">
<pre class="src src-nextflow">
</pre>
</div>

<p>
the tap seems to hap after nextflow has stopped,
look more into this.
</p>

<p>
possible causes:
<a href="https://github.com/nextflow-io/nextflow/issues/2502">Queue remains open when data is staged from an external source · Issue #2502 · nextflow-io/nextflow · GitHub</a>
<a href="https://github.com/nextflow-io/nextflow/issues/1230">Parent nextflow process doesn't exit after all compute tasks are complete · Issue #1230 · nextflow-io/nextflow · GitHub</a>
</p>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">we handle the one or two engines case</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">DSL2 incompat</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">would be in workflow body</span>

<span class="org-keyword">if</span> (params.search_engines.size() &gt; 1) {  
    <span class="org-keyword">process</span> <span class="org-function-name">CombineSearchResults</span> {
        publishDir <span class="org-string">"</span><span class="org-variable-name">${params.outdir}</span><span class="org-string">/speclib"</span>
        when:

        input:
        <span class="org-keyword">file</span> <span class="org-function-name">xtandem_search_results</span> from xtandem_search_results_ch
        <span class="org-keyword">file</span> <span class="org-function-name">comet_search_results</span> from comet_search_results_ch
        output:
        file <span class="org-string">"lib_iprophet.peps.xml"</span> into combined_search_results_ch
        <span class="org-string">"""</span>
<span class="org-string">        InterProphetParser DECOY=DECOY_ THREADS=</span><span class="org-variable-name">${task.cpus}</span><span class="org-string"> </span><span class="org-variable-name">$xtandem_search_results</span><span class="org-string"> </span><span class="org-variable-name">$comet_search_results</span><span class="org-string"> lib_iprophet.peps.xml</span>
<span class="org-string">        """</span>
    }
} <span class="org-keyword">else</span> <span class="org-keyword">if</span> (params.search_engines.contains(<span class="org-string">"comet"</span>)) {
    combined_search_results_ch = comet_search_results_ch
} <span class="org-keyword">else</span> <span class="org-keyword">if</span> (params.search_engines.contains(<span class="org-string">"xtandem"</span>)) {
    combined_search_results_ch =xtandem_search_results_ch
} <span class="org-keyword">else</span> {
    combined_search_results_ch = <span class="org-type">Channel</span>.create()
}
</pre>
</div>
</div>
</div>
<div id="outline-container-org7a590b7" class="outline-5">
<h5 id="org7a590b7"><span class="section-number-5">5.5.1.8.</span> Building Specral Library</h5>
<div class="outline-text-5" id="text-5-5-1-8">
<p>


Inputs from <a href="#org47af1f4">Creating Swath window files</a>
<a href="http://www.openswath.org/en/latest/docs/openswath.html">http://www.openswath.org/en/latest/docs/openswath.html</a>
</p>
</div>

<div id="outline-container-orgff0099e" class="outline-6">
<h6 id="orgff0099e"><span class="section-number-6">5.5.1.8.1.</span> Mayu</h6>
<div class="outline-text-6" id="text-5-5-1-8-1">
<p>
<a href="https://doi.org/10.5167/uzh-28712">https://doi.org/10.5167/uzh-28712</a>
<a href="https://doi.org/10.1074/mcp.M900317-MCP200">https://doi.org/10.1074/mcp.M900317-MCP200</a>
</p>
<blockquote>
<p>
GENERAL:
Mayu is a software package to determine protein
identification false discovery rates (protFDR) and
peptide identification false discovery rates (pepFDR)
additionally to the peptide-spectrum match false discovery 
rate (mFDR).
</p>
</blockquote>

<p>
Here is what happens in mayu:
For a pepxml file with peptide-spectrum-matches <code>PSM</code>
(type of <code>(spectrum,peptide,probability)</code>, where the probality is based on the similarity of the theoratical spectrum,
mayu determines the peptide-spectrum-match False Detection Rate (<code>mFDR</code>),
and protein identification false discovery rates (<code>protFDR</code>).
We select a <code>protFDR</code> for which mayu finds a matching <code>mFDR</code> level (no higher than the <code>-G</code> flag) and it will filter
everything with a higher mFDR level
In the output csv the <code>score</code> column is the the <code>probability</code> in PSM (in mayu documentation "discrimant")
</p>

<p>
We find the lowest <code>probability</code> that still has an <code>mFDR</code> that matched the above,
and that is what we use as the filtering criterian in spectrast
</p>

<p>
This is what we will than filter on with specrtrast
</p>

<p>
Hmhf why can't mayu return deterministic filenames.
(It incorporates the mayu version number in the filename grumbl),
it follows the pattern
</p>
<div class="org-src-container">
<pre class="src src-perl"><span class="org-keyword">my</span> $<span class="org-variable-name">psm_file_base</span> = $<span class="org-variable-name">out_base</span> . <span class="org-string">'_psm_'</span>;
<span class="org-keyword">my</span> $<span class="org-variable-name">id_csv_file</span> = $<span class="org-variable-name">psm_file_base</span>
                . $<span class="org-variable-name">fdr_type</span>
                . $<span class="org-variable-name">fdr_value</span> . <span class="org-string">'_'</span>
                . $<span class="org-variable-name">target_decoy</span> . <span class="org-string">'_'</span>
                . $<span class="org-variable-name">version</span> . <span class="org-string">'.csv'</span>;

</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow">process  <span class="org-type">FindMinimumPeptideProbability</span> {
    input: 
    <span class="org-keyword">file</span> <span class="org-function-name">combined_search_results</span> from combined_search_results_ch.first()
    <span class="org-keyword">file</span> <span class="org-function-name">fastadb_with_decoy</span> from joined_fasta_with_decoys_ch.first()
    <span class="org-keyword">val</span> <span class="org-function-name">max_missed_cleavages</span>
    output:
    env PEPTIDEPROBABILITY into  minimum_peptide_probability
    <span class="org-comment-delimiter">/* </span><span class="org-comment">explanation of paramaters</span>
<span class="org-comment">     -G  $params.protFDR            | maximum allowed mFDR of $params.protFDR </span>
<span class="org-comment">     -P protFDR=$params.protFDR:t   |  print out PSMs of targets who have a protFDR of $params.protFDR</span>
<span class="org-comment">     -</span>
<span class="org-comment">     -H | defines the resolution of error analysis (mFDR steps)</span>
<span class="org-comment">     -I number of missed cleavages used for database search</span>
<span class="org-comment">     -M | file name base</span>
<span class="org-comment">     */</span>
    script:
    prefix=<span class="org-string">"filtered"</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">you can change this to a glob-pattern (e.g. "*") for future-proofing</span>
    mayu_version=<span class="org-string">"1.07"</span>
    psm_csv=<span class="org-string">"</span><span class="org-variable-name">${prefix}</span><span class="org-string">_psm_protFDR</span><span class="org-variable-name">${params.protFDR}</span><span class="org-string">_t_</span><span class="org-variable-name">${mayu_version}</span><span class="org-string">.csv"</span>
    <span class="org-string">"""</span>
<span class="org-string">    Mayu.pl -verbose -A </span><span class="org-variable-name">$combined_search_results</span><span class="org-string"> -C </span><span class="org-variable-name">$fastadb_with_decoy</span><span class="org-string"> -E DECOY_ -G </span><span class="org-variable-name">$params.protFDR</span><span class="org-string"> -P protFDR=</span><span class="org-variable-name">${params.protFDR}</span><span class="org-string">:t -H 51 -I </span><span class="org-variable-name">$max_missed_cleavages</span><span class="org-string"> -M </span><span class="org-variable-name">$prefix</span>
<span class="org-string">    # test if psm_csv was made</span>
<span class="org-string">    test -e </span><span class="org-variable-name">$psm_csv</span><span class="org-string"> || exit 1</span>
<span class="org-string">    # test if the results arent empty</span>
<span class="org-string">    test `wc -l </span><span class="org-variable-name">$psm_csv</span><span class="org-string"> | cut -d' ' -f1` -gt 1 || exit 1 </span>
<span class="org-string">    PEPTIDEPROBABILITY=`cat </span><span class="org-variable-name">$psm_csv</span><span class="org-string"> | cut -f 5 -d ',' |tail -n+2 |sort -u | head -n1`</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<p>
Note that sort requires <code>$TMPDIR</code> to actually exists and be writable,
<code>$TMPDIR</code> (the envvar) is inherited from the parent env when run in a container,
but not mounted (at least not in Singularity), so if <code>$TMPDIR</code> does not exist in the container, this will crash.
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">sensible values = floats between 0 and 1 </span>
<span class="org-comment-delimiter">// </span><span class="org-comment">target FDR for mayu</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">this is equivalent to the "pvalue" parameter in the original (python) gladiator implementation</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">which is labeld as "Spectral library building FDR" in the UI</span>
params.protFDR=0.01
</pre>
</div>
</div>
</div>

<div id="outline-container-org57b626e" class="outline-6">
<h6 id="org57b626e"><span class="section-number-6">5.5.1.8.2.</span> Spectrast</h6>
<div class="outline-text-6" id="text-5-5-1-8-2">
<p>
<a href="http://tools.proteomecenter.org/wiki/index.php?title=Software:SpectraST">http://tools.proteomecenter.org/wiki/index.php?title=Software:SpectraST</a>
</p>

<p>
Spectrast in <code>SpectraSTPepXMLLibImporter.cpp</code> <code>readFromFile</code> <code>processSearchHit</code> will read
the mzxmls contained in the pepxml. It defaults to looking for the mzxml in the CWD
otherwise it checks the path in the <code>base_name</code> property of msms_run_summary element in <code>&lt;search_summary</code>
so we need again give the <code>maybespectra_ch</code> on.
From the above url
</p>
<blockquote>
<ul class="org-ul">
<li>Creating Consensus Libraries</li>
<li>Importing the raw spectra into SpectraST</li>
</ul>
<p>
[ &#x2026; ]
Remember that the .mzXML files must be in the same directories as their
corresponding .pepXML files. 
</p>
</blockquote>

<p>
Spectrast is from <code>tpp</code>
Note that spectrast flags are single-dash multilettered underscored argument-concatenated.thanks.
Its argument-parser is <b>very</b> funky, so be careful here.
It also doesn't check if illegal flags are given, they will pass silently instead, <b>grumble.</b>
</p>
<div class="org-src-container">
<pre class="src src-fundamental">(\_/) .~~ 
(._.)/~~~
(_ _)     
</pre>
</div>
</div>

<div id="outline-container-org91d58b9" class="outline-7">
<h7 id="org91d58b9"><span class="section-number-7">5.5.1.8.2.1.</span> Converting traml into spectrast friendly format</h7>
<div class="outline-text-7" id="text-5-5-1-8-2-1">
<div class="org-src-container">
<pre class="src src-awk" id="org6ddc044"><span class="org-keyword">BEGIN</span> {<span class="org-variable-name">FS</span>=<span class="org-string">"     "</span>; <span class="org-variable-name">OFS</span>=<span class="org-string">"        "</span>}
<span class="org-variable-name">NR</span>==1 {
    <span class="org-keyword">for</span> (i=1; i&lt;=<span class="org-variable-name">NF</span>; i++) {
        f[$i] = i
    }
}
<span class="org-variable-name">NR</span>&gt;1 { <span class="org-preprocessor">print</span> $(f[<span class="org-string">"PeptideSequence"</span>]), $(f[<span class="org-string">"NormalizedRetentionTime"</span>]) }
</pre>
</div>

<div class="org-src-container">
<pre class="src src-awk" id="org2ace004"><span class="org-keyword">BEGIN</span> {<span class="org-variable-name">FS</span>=<span class="org-string">"     "</span>; <span class="org-variable-name">OFS</span>=<span class="org-string">"        "</span>}
<span class="org-comment-delimiter"># </span><span class="org-comment">we set the column names so that we can look them up later</span>
<span class="org-variable-name">NR</span>==1 {
    <span class="org-keyword">for</span> (i=1; i&lt;=<span class="org-variable-name">NF</span>; i++) {
        f[$i] = i
    }
}
<span class="org-comment-delimiter"># </span><span class="org-comment">use only the last entry in the table per peptide sequence</span>
<span class="org-variable-name">NR</span>&gt;1 {
    irt_by_sequence[$(f[<span class="org-string">"PeptideSequence"</span>])] = $(f[<span class="org-string">"NormalizedRetentionTime"</span>])
    peptide_sequences[$(f[<span class="org-string">"PeptideSequence"</span>])]=$(f[<span class="org-string">"PeptideSequence"</span>])
}
<span class="org-keyword">END</span> {
    <span class="org-keyword">for</span> (sequence <span class="org-keyword">in</span> peptide_sequences)
        <span class="org-preprocessor">print</span>(sequence,irt_by_sequence[sequence])
}
</pre>
</div>

<p>
TargetedFileConverter from <code>OpenMS</code>
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">CreateSpectrastIrtFile</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">irt_traml</span> from <span class="org-type">Channel</span>.fromPath(params.irt_traml_file)
    output:
    file (<span class="org-string">"irt.txt"</span>) into irt_txt_ch
    script:
    intermediate_tsv=<span class="org-string">"intermediate_irt.tsv"</span>
    <span class="org-string">"""</span>
<span class="org-string">    TargetedFileConverter -in </span><span class="org-variable-name">$irt_traml</span><span class="org-string"> -out_type tsv -out </span><span class="org-variable-name">$intermediate_tsv</span>
<span class="org-string">    """</span> + <span class="org-string">'''  awk '</span>
<span class="org-string">    BEGIN {FS=" "; OFS="        "}</span>
<span class="org-string">    NR==1 {</span>
<span class="org-string">        for (i=1; i&lt;=NF; i++) {</span>
<span class="org-string">            f[$i] = i</span>
<span class="org-string">        }</span>
<span class="org-string">    }</span>
<span class="org-string">    NR&gt;1 { print $(f["PeptideSequence"]), $(f["NormalizedRetentionTime"]) }' '''</span> + <span class="org-string">"</span><span class="org-variable-name">$intermediate_tsv</span><span class="org-string"> &gt; irt.txt"</span>
}
</pre>
</div>
</div>
</div>
<div id="outline-container-org109f2bb" class="outline-7">
<h7 id="org109f2bb"><span class="section-number-7">5.5.1.8.2.2.</span> Running Spectrast</h7>
<div class="outline-text-7" id="text-5-5-1-8-2-2">
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">spectrast will create *.splib, *.spidx, *.pepidx, </span>
<span class="org-comment-delimiter">// </span><span class="org-comment">note that where-ever a splib goes, so must its spidx and pepidx</span>
<span class="org-comment-delimiter">//</span><span class="org-comment">/and they must have the same part</span>
<span class="org-keyword">process</span> <span class="org-function-name">SpectrastCreateSpecLib</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">irtfile</span> from irt_txt_ch
    <span class="org-keyword">file</span> <span class="org-function-name">combined_search_results</span> from combined_search_results_ch.first()
    <span class="org-keyword">file</span> <span class="org-function-name">fastadb_with_decoy</span> from joined_fasta_with_decoys_ch.first()
    <span class="org-keyword">file</span> <span class="org-function-name">spectra</span> from maybespectra_ch.spectrast.toSortedList()
    <span class="org-keyword">val</span> <span class="org-function-name">cutoff</span> from minimum_peptide_probability
    output:
    tuple file (<span class="org-string">"</span><span class="org-variable-name">${prefix}</span><span class="org-string">_cons.splib"</span>), file(<span class="org-string">"</span><span class="org-variable-name">${prefix}</span><span class="org-string">_cons.spidx"</span>) into spectrast_ch
    file(<span class="org-string">"</span><span class="org-variable-name">${prefix}</span><span class="org-string">_cons.sptxt"</span>) into consensus_lib_sptxt_ch
    script:
    prefix = <span class="org-string">"SpecLib"</span>
    to_run = <span class="org-string">"spectrast -cN</span><span class="org-variable-name">${prefix}</span><span class="org-string"> -cIHCD -cf\"Protein! ~ DECOY_\" -cP</span><span class="org-variable-name">$cutoff</span><span class="org-string"> -c_IRR "</span>
    <span class="org-keyword">if</span> (params.use_irt)
        to_run += <span class="org-string">"-c_IRT</span><span class="org-variable-name">$irtfile</span><span class="org-string"> "</span>
    to_run +=  <span class="org-string">"</span><span class="org-variable-name">$combined_search_results</span><span class="org-string">"</span> <span class="org-comment-delimiter">// </span><span class="org-comment">spectrast really wants its input-files last.</span>
    to_run += <span class="org-string">"\n spectrast -cN</span><span class="org-variable-name">${prefix}</span><span class="org-string">_cons -cD</span><span class="org-variable-name">$fastadb_with_decoy</span><span class="org-string"> -cIHCD -cAC </span><span class="org-variable-name">${prefix}</span><span class="org-string">.splib"</span>
}
</pre>
</div>

<p>
from original gladiator implementation
source is unclear;
author forgot.
</p>

<div class="org-src-container">
<pre class="src src-fundamental">LGGNEQVTR   -28.308
GAGSSEPVTGLDAK  0.227
VEATFGVDESNAK   13.1078
YILAGVENSK  22.3798
TPVISGGPYEYR    28.9999
TPVITGAPYEYR    33.6311
DGLDAASYYAPVR   43.2819
ADVTPADFSEWSK   54.969
GTFIIDPGGVIR    71.3819
GTFIIDPAAVIR    86.7152
LFLQFGAQGSPFLK  98.0897
</pre>
</div>

<p>
<a id="orgb6857c4"></a>
</p>
<p>
<a id="orgd63b5b4"></a>
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">white-space-delimited file of peptide-sequences and internal retention times</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">whether or not to use the retention-</span>
params.use_irt=<span class="org-constant">true</span>
params.irt_traml_file = <span class="org-string">"iRTAssayLibrary.TraML"</span>
</pre>
</div>

<p>
Here we forward declary  <code>consensus_pseudospectra_openswath_library_tsv</code>
so that we can later redirect it.
</p>
<div class="org-src-container">
<pre class="src src-nextflow">consensus_pseudospectra_openswath_library_tsv = <span class="org-type">Channel</span>.create()
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">Spectrast2OpenSwathTsv</span> {
 <span class="org-comment-delimiter">/*</span>
<span class="org-comment">     Choice parts of sprectrast2.tsv --help</span>

<span class="org-comment">     spectrast2tsv.py</span>
<span class="org-comment">     ---------------</span>
<span class="org-comment">     This script is used as filter from spectraST files to swath input files.</span>
<span class="org-comment">     python spectrast2tsv.py [options] spectrast_file(s)</span>

<span class="org-comment">     -d                  Remove duplicate masses from labeling</span>
<span class="org-comment">     -e                  Use theoretical mass</span>
<span class="org-comment">     -k    output_key    Select the output provided. Keys available: openswath, peakview. Default: peakview</span>
<span class="org-comment">     -l    mass_limits   Lower and upper mass limits of fragment ions. Example: -l 400,2000</span>
<span class="org-comment">     -s    ion_series    List of ion series to be used. Example: -s y,b</span>

<span class="org-comment">     -w    swaths_file   File containing the swath ranges. This is used to remove transitions with Q3 falling in the swath mass range. (line breaks in windows/unix format)</span>
<span class="org-comment">     -n    int           Max number of reported ions per peptide/z. Default: 20</span>
<span class="org-comment">     -o    int           Min number of reported ions per peptide/z. Default: 3</span>
<span class="org-comment">     -a    outfile       Output file name (default: appends _peakview.txt)</span>
<span class="org-comment">     */</span>
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">swath_windows</span> from swath_windows_ch.first()
    <span class="org-keyword">file</span> <span class="org-function-name">sptxt</span> from consensus_lib_sptxt_ch.first()
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">consensus_pseudospectra_openswath_library_tsv</span>
    script:
    consensus_pseudospectra_openswath_library_tsv=<span class="org-string">"SpecLib_cons_openswath.tsv"</span>
    <span class="org-string">"""</span>
<span class="org-string">    MINWINDOW=`head -n1 </span><span class="org-variable-name">$swath_windows</span><span class="org-string"> | cut -d'        ' -f1`</span>
<span class="org-string">    MAXWINDOW=`tail -n1 </span><span class="org-variable-name">$swath_windows</span><span class="org-string"> | cut -d'        ' -f2`</span>
<span class="org-string">    spectrast2tsv.py -l \$MINWINDOW,\$MAXWINDOW -s y,b -d -e -o 6 -n 6 -w </span><span class="org-variable-name">$swath_windows</span><span class="org-string"> -k openswath -a </span><span class="org-variable-name">$consensus_pseudospectra_openswath_library_tsv</span><span class="org-string"> </span><span class="org-variable-name">$sptxt</span>
<span class="org-string">    """</span>
}
</pre>
</div>
</div>
</div>
</div>
</div>
</div>

<div id="outline-container-section-deepdia" class="outline-4">
<h4 id="section-deepdia"><span class="section-number-4">5.5.2.</span> Building Spectral library from Machine learning</h4>
<div class="outline-text-4" id="text-section-deepdia">
<p>
<a href="https://github.com/lmsac/DeepDIA">DeepDIA</a> can predict the spectral library from peptide lists
See also its documentation
<a href="https://github.com/lmsac/DeepDIA/raw/master/README.md">https://github.com/lmsac/DeepDIA/raw/master/README.md</a>
<a href="https://github.com/lmsac/DeepDIA/raw/master/docs/predict_detectability.md">https://github.com/lmsac/DeepDIA/raw/master/docs/predict_detectability.md</a>
<a href="https://doi.org/10.1038/s41467-019-13866-z">https://doi.org/10.1038/s41467-019-13866-z</a>
</p>

<div class="org-src-container">
<pre class="src src-nextflow">
<span class="org-keyword">if</span>(libgen_method_is_enabled(<span class="org-string">"deepdia"</span>,params))  { 
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow">params._deepdia_url = <span class="org-string">"https://github.com/lmsac/DeepDIA/raw/c5ad2aa50218fcdfd1d441714702e605fdb00bb3"</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">float or null</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">if null, do not use minimum detectability filtering</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">if a float, filter</span>
params.deepdia_min_detectability = <span class="org-constant">null</span>
params.deepdia_detectability_model = <span class="org-string">"</span><span class="org-variable-name">${params._deepdia_url}</span><span class="org-string">/data/models/detectability/epoch_004.hdf5"</span>
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">list tuples in the form of</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">[charge, model, peptidelist]</span>
params.deepdia_ms2_entries = [
    [<span class="org-string">"2"</span>,
     <span class="org-string">"</span><span class="org-variable-name">${params._deepdia_url}</span><span class="org-string">/data/models/charge2/epoch_035.hdf5"</span>,
     ],
    [<span class="org-string">"3"</span>,
     <span class="org-string">"</span><span class="org-variable-name">${params._deepdia_url}</span><span class="org-string">/data/models/charge3/epoch_034.hdf5"</span>,
     ]]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-type">Channel</span>
    .from(params.deepdia_ms2_entries)
    .map( {
            charge, model -&gt;
            tuple(charge, file(model))})
    .set{deepdia_ms2_models}
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">DeepDIADigestProtein</span>
{
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">joined_fasta</span> from joined_fasta_database_ch
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">deepdia_peptide_list</span>
    script:
    deepdia_peptide_list=<span class="org-string">"deepdia_peptide_list.csv"</span>
    <span class="org-string">"""</span>
<span class="org-string">    digest_proteins.py --in </span><span class="org-variable-name">$joined_fasta</span><span class="org-string"> --out </span><span class="org-variable-name">$deepdia_peptide_list</span><span class="org-string"> --no-group_duplicated</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<p>
If we do detectability filtering we mix the filtered peptides with the models,
otherwise the unfiltered.
</p>

<div class="org-src-container">
<pre class="src src-nextflow">deepdia_peptide_list = <span class="org-type">Channel</span>.create()
<span class="org-keyword">if</span> (params.deepdia_min_detectability != <span class="org-constant">null</span>){
    deepdia_peptide_list.set{deepdia_prefilt_peptide_list}
    deepdia_filtered_peptide_list = <span class="org-type">Channel</span>.create()
    deepdia_filtered_peptide_list
        .tap{deepdia_peptides_for_retention_pred}
        .tap{deepdia_peptides_for_library}
        .combine(deepdia_ms2_models)
        .set{deepdia_ms2_inputs_ch}
} <span class="org-keyword">else</span> {
    deepdia_peptide_list
        .tap{deepdia_peptides_for_retention_pred}
        .tap{deepdia_peptides_for_library}
        .combine(deepdia_ms2_models)
        .set{deepdia_ms2_inputs_ch}
}
</pre>
</div>

<p>
So here we predict detectability of peptides and filter by them, if requested
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span> (params.deepdia_min_detectability != <span class="org-constant">null</span>){
    <span class="org-comment-delimiter">// </span><span class="org-comment">we seperate these two so that --resume allows for easy tweaking of --minimum-detectability</span>
    <span class="org-keyword">process</span> <span class="org-function-name">DeepDIATrainDetectibility</span> {
        memory <span class="org-string">'64 GB'</span>
        input:
        <span class="org-keyword">file</span> <span class="org-function-name">model</span> from <span class="org-type">Channel</span>.fromPath(params.deepdia_detectability_model)
        <span class="org-keyword">file</span> <span class="org-function-name">deepdia_prefilt_peptide_list</span>
        output:
        set file(deepdia_detectability_prediction), file(deepdia_prefilt_peptide_list) into deepdia_detectability
        script:
        deepdia_detectability_prediction=<span class="org-string">"</span><span class="org-variable-name">${deepdia_prefilt_peptide_list.baseName}</span><span class="org-string">.detectability.csv"</span>
        <span class="org-string">"predict_detectability.py --in </span><span class="org-variable-name">$deepdia_prefilt_peptide_list</span><span class="org-string"> --model </span><span class="org-variable-name">$model</span><span class="org-string"> --out </span><span class="org-variable-name">$deepdia_detectability_prediction</span><span class="org-string">"</span>
    }

    <span class="org-keyword">process</span> <span class="org-function-name">DeepDIAMinimumDetectabilityFiltering</span>
    {
        input:
        set file(detectability_prediction), file(prefilt_peptide_list) from deepdia_detectability
        <span class="org-keyword">val</span> <span class="org-function-name">min_detectability</span> from <span class="org-type">Channel</span>.value(params.deepdia_min_detectability)
        output:
        <span class="org-keyword">file</span> <span class="org-function-name">deepdia_filtered_peptide_list</span>
        script:
        deepdia_filtered_peptide_list=<span class="org-string">"deepdia_filtered_peptide_list.csv"</span>
        <span class="org-string">"""</span>
<span class="org-string">        filter_peptide_by_detectability.py --peptide </span><span class="org-variable-name">$prefilt_peptide_list</span><span class="org-string"> --detect </span><span class="org-variable-name">$detectability_prediction</span><span class="org-string"> --min_detectability </span><span class="org-variable-name">$min_detectability</span><span class="org-string"> --out </span><span class="org-variable-name">${deepdia_filtered_peptide_list}</span>
<span class="org-string">        """</span>
    }
}

</pre>
</div>

<p>
Then we predict the ms2 
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">DeepDIAPredictCharge</span> {
    memory <span class="org-string">'64 GB'</span>
    input:
    set file(peptides),charge,file(model)  from deepdia_ms2_inputs_ch
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">deepdia_ions</span>
    script:
    deepdia_ions=<span class="org-string">"predictions.charge.</span><span class="org-variable-name">${charge}</span><span class="org-string">.ions.json"</span>
    <span class="org-string">"""</span>
<span class="org-string">    predict_ms2.py --charge </span><span class="org-variable-name">$charge</span><span class="org-string"> --in </span><span class="org-variable-name">$peptides</span><span class="org-string"> --model </span><span class="org-variable-name">$model</span><span class="org-string"> --out </span><span class="org-variable-name">$deepdia_ions</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow">params.deepdia_irt_model =
    <span class="org-string">"</span><span class="org-variable-name">${params._deepdia_url}</span><span class="org-string">/data/models/irt/epoch_082.hdf5"</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">params.deepdia_peptides =</span>
<span class="org-comment-delimiter">//     </span><span class="org-comment">"${params._deepdia_url}/data/peptides/Pan_human.peptide.csv"</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow">deepdia_irt_model = <span class="org-type">Channel</span>.fromPath(params.deepdia_irt_model)
deepdia_speclib = <span class="org-type">Channel</span>.create()
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">DeepDIAPredictRetention</span> {
    memory <span class="org-string">'32 GB'</span>
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">deepdia_irt_model</span>
    <span class="org-keyword">file</span> <span class="org-function-name">deepdia_peptides_for_retention_pred</span>
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">predicted_rt</span>
    script:
    predicted_rt=<span class="org-string">"prediction.irt.csv"</span>
    <span class="org-string">"""</span>
<span class="org-string">    predict_rt.py --in </span><span class="org-variable-name">$deepdia_peptides_for_retention_pred</span><span class="org-string"> --model </span><span class="org-variable-name">$deepdia_irt_model</span><span class="org-string"> --out </span><span class="org-variable-name">$predicted_rt</span>
<span class="org-string">    """</span>
}
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">DeepDIAPredictionsToLibrary</span> {
    memory <span class="org-string">'32 GB'</span>
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">predicted_rt</span>
    <span class="org-keyword">file</span> <span class="org-function-name">ions</span> from deepdia_ions.toSortedList()
    <span class="org-keyword">file</span> <span class="org-function-name">deepdia_peptides_for_library</span>
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">deepdia_speclib</span>
    script:
    deepdia_speclib=<span class="org-string">"speclib.tsv"</span>
    <span class="org-string">"""</span>
<span class="org-string">    build_assays_from_prediction.py --peptide </span><span class="org-variable-name">${deepdia_peptides_for_library}</span><span class="org-string"> --rt </span><span class="org-variable-name">${predicted_rt}</span><span class="org-string"> --ions </span><span class="org-variable-name">${ions}</span><span class="org-string"> --out prediction.assay.pickle</span>
<span class="org-string">    convert_assays_to_openswath.py --in prediction.assay.pickle --out </span><span class="org-variable-name">${deepdia_speclib}</span>
<span class="org-string">    """</span>
}
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow">deepdia_speclib.set{speclib_tsv_for_decoys}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow">}  <span class="org-comment-delimiter">// </span><span class="org-comment">end of deepdia guard</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org0d8d164" class="outline-4">
<h4 id="org0d8d164"><span class="section-number-4">5.5.3.</span> Supplying a custom spectral library</h4>
<div class="outline-text-4" id="text-5-5-3">
<p>
You can also supply a custom spectral library by setting <code>=--speclib</code> to a glob/path
of your customly generated library.
This should be in a format that openms's TargetedFileConverte understand,
so MaxQuant or OpenMS works.
</p>

<p>
To check, one can manually inspect whether  <code>TargetedFileConverter --in yourlibrary --out speclib.tsv</code> looks proper.
Pay special attention to the modified peptides field, if this is being parsed correctly.
One bug observed by this developer is that this field is repeated after not being parsed correctly.
See also <a href="https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/release/2.7.0/html/classOpenMS_1_1TransitionTSVFile.html">https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/release/2.7.0/html/classOpenMS_1_1TransitionTSVFile.html</a>
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">if params.deconvolution method is set</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">set this to to spectral libraries tsvs in maxquant / openms / any input format that openms TargetedConverter understands</span>
params.speclib = <span class="org-constant">null</span>
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span> (libgen_method_is_enabled(<span class="org-string">"custom"</span>, params)){
    <span class="org-type">Channel</span>.fromPath(params.speclib).set{speclib_tsv_for_decoys}
}
</pre>
</div>
</div>
</div>

<div id="outline-container-org86499fa" class="outline-4">
<h4 id="org86499fa"><span class="section-number-4">5.5.4.</span> combining various spectral libraries into one.</h4>
<div class="outline-text-4" id="text-5-5-4">
<p>
would need to use openms TargetedFileConverter to convert to openswath like tsv,
then msproteomicstools <a href="file://ssh:tamatoa:/data/epouta1/B22003_Elixir_DIA_Nextflow/src-for-inspection/msproteomicstools/analysis/spectral_libs/tsv2spectrast.py">tsv2spectrast.py</a> to turn into spectrast.
Then merge merge with spectrast with either -cJA  or -cJU <a href="http://tools.proteomecenter.org/wiki/index.php?title=Software:SpectraST#Creating_Consensus_Libraries">http://tools.proteomecenter.org/wiki/index.php?title=Software:SpectraST#Creating_Consensus_Libraries</a>
See aso schubert et al (<a href="https://doi.org/10.1038/nprot.2015.015">https://doi.org/10.1038/nprot.2015.015</a>)
</p>
</div>
</div>
</div>
<div id="outline-container-org615da83" class="outline-3">
<h3 id="org615da83"><span class="section-number-3">5.6.</span> OpenSwathDecoys</h3>
<div class="outline-text-3" id="text-5-6">
<p>
specrast2tsv.py is from <code>msproteomicstools</code>
OpenSwathDecoyGenerator from <code>OpenMS</code> <code>topp</code>
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">optional </span>
params.openswath_transitions = <span class="org-string">""</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">Minimum decoy fraction for open swath decoy generator</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">if left unset, gladiator might pick an appropriate one depending on your deconvolution method,</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">should be a fraction between 0.0 and 1.0</span>
params.oswdg_min_decoy_fraction = <span class="org-constant">null</span>
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(params.oswdg_min_decoy_fraction != <span class="org-constant">null</span>) { 
    <span class="org-type">Channel</span>.value(<span class="org-string">"-min_decoy_fraction </span><span class="org-variable-name">${params.oswdg_min_decoy_fraction}</span><span class="org-string">"</span>).set{oswdg_args} 
} <span class="org-keyword">else</span> <span class="org-keyword">if</span> (libgen_method_is_enabled(<span class="org-string">"deepdia"</span>,params)){
 <span class="org-type">Channel</span>.value(<span class="org-string">"-min_decoy_fraction 0.0"</span>).set{oswdg_args}
} <span class="org-keyword">else</span> {
  <span class="org-type">Channel</span>.value(<span class="org-string">""</span>).set{oswdg_args}
}
</pre>
</div>
<p>
With <a href="#section-deepdia">deepDIA</a> method used there  seems to be some problems with generating decoys,
so we set <code>-min_decoy_fraction</code> to  <code>0.0</code> in this case.
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">AddDecoysToOpenSwathTransitions</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">speclib_tsv</span> from speclib_tsv_for_decoys.first()
    <span class="org-keyword">val</span> <span class="org-function-name">oswdg_args</span>
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">outputfile</span> into openswath_transitions_ch
    script:
    outputfile=<span class="org-string">"SpecLib_cons_decoys.pqp"</span>

    <span class="org-string">"""</span>
<span class="org-string">    TargetedFileConverter -in </span><span class="org-variable-name">$speclib_tsv</span><span class="org-string"> -out SpecLib_cons.TraML</span>
<span class="org-string">    OpenSwathDecoyGenerator -decoy_tag DECOY_ -in SpecLib_cons.TraML -out </span><span class="org-variable-name">$outputfile</span><span class="org-string"> -method reverse </span><span class="org-variable-name">$oswdg_args</span>
<span class="org-string">    """</span>
}
</pre>
</div>


<p>
<code>TargetedFileConverter</code> from <code>openms</code>
</p>
<blockquote>
<p>
/usr/bin/TargetedFileConverter: error while loading shared libraries: libQt5Core.so.5: cannot open shared object file: No such file or directory
</p>
</blockquote>

<p>
Here we might actually not need TargetedFileConverter,
can give tsv directly to OpenSwathDecoyGenerator.
and pass result tsv to OpenSwathWorkflow as <code>-tr</code>.
</p>
</div>
</div>

<div id="outline-container-orgfb5b55c" class="outline-3">
<h3 id="orgfb5b55c"><span class="section-number-3">5.7.</span> Building Dia Matrix</h3>
<div class="outline-text-3" id="text-5-7">
<p>
<a href="https://openswath.org/_/downloads/en/latest/pdf/">https://openswath.org/_/downloads/en/latest/pdf/</a>
<a href="https://openswath.org/_/downloads/en/latest/htmlzip/">https://openswath.org/_/downloads/en/latest/htmlzip/</a>
</p>
</div>
<div id="outline-container-org84e26d6" class="outline-4">
<h4 id="org84e26d6"><span class="section-number-4">5.7.1.</span> OpenSwathWorkflow</h4>
<div class="outline-text-4" id="text-5-7-1">
<p>
Creates tsv with <code>-out_tsv</code>

</p>

<p>
Using a the cache will decrease memomry usage at the cost of writes &amp; time
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">wheter to use -readOptions cacheWorkingInMemory in OSW</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">this actually crashes so disabled</span>
params.osw_use_cache = <span class="org-constant">false</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">extra flags to pass to OSW</span>
params.osw_extra_flags =  <span class="org-string">""</span>
</pre>
</div>

<p>
The transitions size is larger if the deconvolution method is <a href="#section-deepdia">deepDIA</a>,
which will consume more memory.
</p>

<p>
to use <code>-out_osw</code>, <code>-tr</code> needs to be a pqp file,
</p>
<p>
If we are using legacy pyprophet we will need to create a tsv
</p>
<div class="org-src-container">
<pre class="src src-nextflow" id="org11413c8"><span class="org-keyword">process</span> <span class="org-function-name">OpenSwathWorkflow_legacy</span> {
        memory { 16.GB * (libgen_method_is_enabled(<span class="org-string">"deepdia"</span>,params) ? 2 : 1 )}
        input:
        <span class="org-keyword">file</span> <span class="org-function-name">dia_mzml_file</span> from dia_mzml_files_ch.osw
        <span class="org-comment-delimiter">// </span><span class="org-comment">file openswath_transitions from Channel.fromPath("data_from_original/bruderer-pwiz-no-dda/SpecLib_cons_decoy.TraML").first()</span>
        <span class="org-keyword">file</span> <span class="org-function-name">openswath_transitions</span> from openswath_transitions_ch_for_legacy.first()
        <span class="org-keyword">file</span> <span class="org-function-name">swath_truncated_windows</span> from truncated_swath_windows_ch.first()
        <span class="org-keyword">file</span> <span class="org-function-name">irt_traml</span> from <span class="org-type">Channel</span>.fromPath(params.irt_traml_file).first()
        output:
        <span class="org-keyword">file</span> <span class="org-function-name">dia_tsv_file</span>  into openswath_tsv_ch
        script:
        dia_tsv_file = <span class="org-string">"</span><span class="org-variable-name">${dia_mzml_file.baseName}</span><span class="org-string">-DIA.tsv"</span>
        to_execute =
            <span class="org-string">"OpenSwathWorkflow "</span> +
            <span class="org-string">"-force "</span> +
            <span class="org-string">"-in </span><span class="org-variable-name">$dia_mzml_file</span><span class="org-string"> "</span> +
            <span class="org-string">"-tr </span><span class="org-variable-name">$openswath_transitions</span><span class="org-string"> "</span> +
            <span class="org-string">"-threads </span><span class="org-variable-name">${task.cpus}</span><span class="org-string"> "</span> +
            <span class="org-string">"-min_upper_edge_dist 1 "</span> +
            <span class="org-string">"-sort_swath_maps "</span> +
            <span class="org-string">"-out_tsv </span><span class="org-variable-name">${dia_tsv_file}</span><span class="org-string"> "</span> + 
            <span class="org-string">"-swath_windows_file </span><span class="org-variable-name">$swath_truncated_windows</span><span class="org-string"> "</span> +
            params.osw_extra_flags + <span class="org-string">" "</span>
        <span class="org-keyword">if</span> (params.use_irt)
            to_execute +=  <span class="org-string">"-tr_irt </span><span class="org-variable-name">$irt_traml</span><span class="org-string"> "</span>
        to_execute
}
</pre>
</div>



<p>
If we are using nonlegacy pyprophet we will need to create an osw
</p>
<div class="org-src-container">
<pre class="src src-nextflow" id="org4ac5369">openswath_osw_indirect_ch = <span class="org-type">Channel</span>.create()
openswath_osw_indirect_ch.multiMap{ <span class="org-variable-name">it</span> -&gt;
    pyprophet_subsample: pyprophet_score  : <span class="org-variable-name">it</span>}.set{openswath_osw_ch}
<span class="org-keyword">process</span> <span class="org-function-name">OpenSwathWorkflow</span> {
    memory { 16.GB * (libgen_method_is_enabled(<span class="org-string">"deepdia"</span>,params) ? 2 : 1 )}
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">dia_mzml_file</span> from dia_mzml_files_ch.osw
    <span class="org-comment-delimiter">// </span><span class="org-comment">file openswath_transitions from Channel.fromPath("data_from_original/bruderer-pwiz-no-dda/SpecLib_cons_decoy.TraML").first()</span>
    <span class="org-keyword">file</span> <span class="org-function-name">openswath_transitions</span> from openswath_transitions_ch_for_nonlegacy.first()
    <span class="org-keyword">file</span> <span class="org-function-name">swath_truncated_windows</span> from truncated_swath_windows_ch.first()
    <span class="org-keyword">file</span> <span class="org-function-name">irt_traml</span> from <span class="org-type">Channel</span>.fromPath(params.irt_traml_file).first()
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">dia_osw_file</span>  into openswath_osw_indirect_ch
    script:
    dia_osw_file = <span class="org-string">"</span><span class="org-variable-name">${dia_mzml_file.baseName}</span><span class="org-string">-DIA.osw"</span>
    to_execute =
        <span class="org-string">"OpenSwathWorkflow "</span> +
        <span class="org-string">"-force "</span> +
        <span class="org-string">"-in </span><span class="org-variable-name">$dia_mzml_file</span><span class="org-string"> "</span> +
        <span class="org-string">"-tr </span><span class="org-variable-name">$openswath_transitions</span><span class="org-string"> "</span> +
        <span class="org-string">"-threads </span><span class="org-variable-name">${task.cpus}</span><span class="org-string"> "</span> +
        <span class="org-string">"-min_upper_edge_dist 1 "</span> +
        <span class="org-string">"-sort_swath_maps "</span> +
        <span class="org-string">"-out_osw </span><span class="org-variable-name">${dia_osw_file}</span><span class="org-string"> "</span> + 
        <span class="org-string">"-swath_windows_file </span><span class="org-variable-name">$swath_truncated_windows</span><span class="org-string"> "</span> +
        params.osw_extra_flags + <span class="org-string">" "</span>

    <span class="org-keyword">if</span> (params.use_irt)
        to_execute +=  <span class="org-string">"-tr_irt </span><span class="org-variable-name">$irt_traml</span><span class="org-string"> "</span>
    to_execute
}
</pre>
</div>

<p>
Then here we choose which one to use
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">we will need the osw to go to various processes</span>
<span class="org-keyword">if</span> (params.pyprophet_use_legacy){
    openswath_transitions_ch.into{openswath_transitions_ch_for_legacy}
 &lt;&lt;nf-openswathworkflow-<span class="org-keyword">for</span>-pyprophet-legacy&gt;&gt;
} <span class="org-keyword">else</span> {
    openswath_transitions_ch.into{openswath_transitions_ch_for_nonlegacy;openswath_transitions_ch_for_pyprophet}
  &lt;&lt;nf-openswathworkflow-<span class="org-keyword">for</span>-pyprophet-nonlegacy&gt;&gt;
}
</pre>
</div>
<p>
(Apparently these two cant have the same name, even if they are conditionally declared
</p>


<blockquote>
<p>
Extraction windows have a gap. Will abort (override with -force)
</p>
</blockquote>

<p>
OpenSwathWorkflow invocation can output tsv XOR osw, not both.
You will get exit status 8 
</p>
<blockquote>
<p>
Error: Unexpected internal error (Either out_features, out_tsv or out_osw needs to be set (but not two or three at the same time))
</p>
</blockquote>
</div>

<div id="outline-container-orge7e2f99" class="outline-5">
<h5 id="orge7e2f99"><span class="section-number-5">5.7.1.1.</span> What is the relation between <code>irt.txt</code> and <code>iRTAssayLibrary.TraML</code></h5>
<div class="outline-text-5" id="text-5-7-1-1">
<p>
The <code>irt.txt</code> seems to contain pairs of 
<code>CompoundList/Peptide@sequence</code> and <code>Compoundlist/Peptide/RetentionTimeList/cvParam[@name="normalized retention time"]@value</code>
of the traml file, except
</p>
<blockquote>
<p>
LFLQFGAQGSPFLK  98.0897
</p>
</blockquote>
<p>
is not present in the traml file.
</p>

<p>
traml was <span class="underline">not</span> gotten from here:
<a href="https://db.systemsbiology.net/sbeams/cgi/PeptideAtlas/PASS_View?identifier=PASS00779">https://db.systemsbiology.net/sbeams/cgi/PeptideAtlas/PASS_View?identifier=PASS00779</a>
<a href="file://ftp:PASS00779@ftp.peptideatlas.org:/">file://ftp:PASS00779@ftp.peptideatlas.org:/</a>
(but is comparable)
</p>

<blockquote>
<p>
&lt;OA&gt; The files have been downloaded from net and for now they are
	      intended to be used "as is". If there becomes a need to modify
	      those, then there will be a need figure out how to do it. So, it
	      is very much unexplored how to generate those files for now.
								        [10:56]
&lt;NA&gt; looks like the .txt contains pairs of "sequence - Normalized retention
       time" from the traml. The only one in the .txt that isnt in the traml
       seems to be LFLQFGAQGSPFLK  [10:59]
&lt;NA&gt; where did you download them froM?  [11:00]
&lt;OA&gt; That is a good question. I guess it was some example dataset for
	      openswath, but I don't remember which. If you have time and
	      interest, you could try to figure out how iRT assay library
	      should be built.  [11:07]
</p>
</blockquote>

<blockquote>
<p>

</p>

<p>
Retention time normalization¶
</p>

<p>
The retention time normalization peptides are provided using the optional parameter tr_irt in TraML format. We
suggest to use the iRTassays.TraML file provided in the tutorial dataset, if the Biognosys iRT-kit was used during sample
preparation.
</p>

<p>
If the iRT-kit was not used, it is highly recommended to use or generate a set of endogenous peptides for RT
normalization. A recent publication [5] provides such a set of CiRT peptides suitable for many eukaryotic samples. The
TraML file from the supplementary information can be used as input for tr_irt. Since not all CiRT peptides might be
found, the flag RTNormalization:estimateBestPeptides should be set to improve initial filtering of poor signals.
Further parameters for optimization can be found when invoking OpenSwathWorkflow &#x2013;helphelp under the
RTNormalization section. Those do not require adjustment for most common sample types and LC-MS/MS setups, but
might be useful to tweak for specific scenarios.
</p>
</blockquote>
<blockquote>
<p>
5 Röst HL, Liu Y, D’Agostino G, Zanella M, Navarro P, Rosenberger G, Collins BC, Gillet L, Testa G, Malmström L, Aebersold R. TRIC:
an automated alignment strategy for reproducible protein quantification in targeted proteomics. Nat Methods. 2016 Sep;13(9):777-83. doi:
10.1038/nmeth.3954. Epub 2016 Aug 1. PMID: 27479329
</p>
</blockquote>
<p>
<a href="https://doi.org/10.1038/nmeth.3954">https://doi.org/10.1038/nmeth.3954</a>
<a href="file://ftp:PASS00788@ftp.peptideatlas.org:/">file://ftp:PASS00788@ftp.peptideatlas.org:/</a>
</p>

<p>
also not from here
</p>

<p>
It seems the traml is based on
<a href="file://ftp:PASS00289@ftp.peptideatlas.org:/SGS/assays/OpenSWATH_SM4_iRT_AssayLibrary.TraML">file://ftp:PASS00289@ftp.peptideatlas.org:/SGS/assays/OpenSWATH_SM4_iRT_AssayLibrary.TraML</a>
which has also has the same irt times as iRT.txt,
except its still missign <code>LFLQFGAQGSPFLK</code>,
The same irt.txt can be found in <a href="https://github.com/CaronLab/Allele-specific-library-scripts/blob/main/iRT.txt">https://github.com/CaronLab/Allele-specific-library-scripts/blob/main/iRT.txt</a>,
published before gladiator, with the retention times
also hardcoded in <a href="https://github.com/msproteomicstools/msproteomicstools/raw/master/analysis/spectral_libs/spectrast_updateiRTs.py">https://github.com/msproteomicstools/msproteomicstools/raw/master/analysis/spectral_libs/spectrast_updateiRTs.py</a>
</p>

<p>
<code>OpenMs/src/tests/class_tests/openms/data/MRMDecoyGenerator_input.TraML</code> has the same irts and also contains
<code>LFLQFGAQGSPFLK</code> so that might also be a good target.
</p>
</div>
</div>
</div>


<div id="outline-container-orgc302ff7" class="outline-4">
<h4 id="orgc302ff7"><span class="section-number-4">5.7.2.</span> Pyprophet</h4>
<div class="outline-text-4" id="text-5-7-2">
<p>
<a href="https://github.com/PyProphet/pyprophet">https://github.com/PyProphet/pyprophet</a>
<code>pyprophet</code>
License: 3-clause BSD
<a href="https://doi.org/10.1093/bioinformatics/btu686">https://doi.org/10.1093/bioinformatics/btu686</a>
<a href="https://doi.org/10.1038/s42003-023-04977-x">https://doi.org/10.1038/s42003-023-04977-x</a>
Pyprophet aggregates various quality scores into one score
<code>d_score</code> (discriminant_score)
</p>

<p>
is based on mprophet <a href="https://doi.org/10.1038/nmeth.1584">https://doi.org/10.1038/nmeth.1584</a>
</p>
</div>

<div id="outline-container-org2c0b841" class="outline-5">
<h5 id="org2c0b841"><span class="section-number-5">5.7.2.1.</span> Legacy Pyprophet</h5>
<div class="outline-text-5" id="text-5-7-2-1">
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span> (params.pyprophet_use_legacy)
    <span class="org-keyword">process</span> <span class="org-function-name">pyprophet_legacy</span> {
    publishDir <span class="org-string">"</span><span class="org-variable-name">${params.outdir}</span><span class="org-string">/pyprophet/"</span>, pattern: <span class="org-string">"*.csv"</span>
    publishDir <span class="org-string">"</span><span class="org-variable-name">${params.outdir}</span><span class="org-string">/reports/pyprophet/"</span>, pattern: <span class="org-string">"*.pdf"</span>
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">dia_tsv</span> from openswath_tsv_ch
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">dscore_csv</span> into pyprophet_legacy_ch
    <span class="org-comment-delimiter">// </span><span class="org-comment">just for publishing</span>
    file <span class="org-string">"</span><span class="org-variable-name">${dia_tsv.baseName}</span><span class="org-string">_report.pdf"</span> 
    script:
    seed=<span class="org-string">"928418756933579397"</span>

    dscore_csv=<span class="org-string">"</span><span class="org-variable-name">${dia_tsv.baseName}</span><span class="org-string">_with_dscore.csv"</span>
    <span class="org-string">"""</span>
<span class="org-string">    pyprophet --random_seed=</span><span class="org-variable-name">${seed}</span><span class="org-string"> --delim=tab --export.mayu </span><span class="org-variable-name">${dia_tsv}</span><span class="org-string"> --ignore.invalid_score_columns</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<blockquote>
<p>
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_summary_stat.csv
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_full_stat.csv
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_with_dscore.csv
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_with_dscore_filtered.csv
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_report.pdf
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_cutoffs.txt
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_svalues.txt
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_qvalues.txt
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_dscores_top_target_peaks.txt
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_dscores_top_decoy_peaks.txt
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_mayu.cutoff
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_mayu.fasta
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_mayu.csv
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_scorer.bin
WRITTEN:  B_D140314_SGSDSsample1_R01_MHRM_T0.mzML-DIA_weights.txt
</p>
</blockquote>

<p>
If you get an error here as
</p>
<blockquote>
<p>
raise Exception("got empty input file")
</p>
</blockquote>
<p>
try running with <code>params.use_irt=false</code>
</p>
</div>
</div>
<div id="outline-container-org2a0458e" class="outline-5">
<h5 id="org2a0458e"><span class="section-number-5">5.7.2.2.</span> nonlegacy prypophet</h5>
<div class="outline-text-5" id="text-5-7-2-2">
<p>
<a href="http://www.openswath.org/en/latest/docs/pyprophet.html">http://www.openswath.org/en/latest/docs/pyprophet.html</a>
In order to make the pyrophet step less intensive,
by default we subsample to <code>1 / nr_samples</code>, as suggested in the pyprophet guide.
However you can supply your own here
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">The ratio (0,1] to subsample by in pyprophet.</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">leave to null to use 1 / nr_samples</span>
params.pyprophet_subsample_ratio = <span class="org-constant">null</span> 
</pre>
</div>
<p>
we calculate the subsample ratio from the number of runs
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span>(params.pyprophet_subsample_ratio == <span class="org-constant">null</span>){
    <span class="org-type">Channel</span>.value(1 / <span class="org-type">Math</span>.max(<span class="org-type">Channel</span>.fromPath(params.diafiles).toSortedList().size().getVal(), 1))
        .set{subsample_ratio}
} <span class="org-keyword">else</span> {
    <span class="org-type">Channel</span>.value(params.pyprophet_subsample_ratio).set{subsample_ratio}
}
</pre>
</div>

<p>
In <code>pyprophet subsample</code> and <code>pyprophet score</code> (of nonlegacy pypropet),
we will need to pass <code>--test</code> to not have random behaviour.
This will make the sql behaviour behaviour non-random.
Because sql doesn't allow for settign a seed, pyprophet will just select the first <code>subsample_ratio</code> of transitions,
This can result in no decoy peptides being in the subsampled osw's.
This is indicated by the error of pyprophet:
</p>
<pre class="example" id="orge2b9cf3">
Error: At least 10 decoy groups and 10 target groups are required.
</pre>

<p>
This leaves you with one of three choices:
</p>

<ol class="org-ol">
<li>passing <code>=--pyprophet_subsample_ratio=1.0</code>, sacrificing runtime.</li>
<li>passing <code>=--pyprophet_fixed_seed=false</code>, sacrificing reproducibility.</li>
<li>passing <code>=--pyprophet_use_legacy=true</code>, sacrificing "up-to-dateness".</li>
</ol>

<div class="org-src-container">
<pre class="src src-nextflow">params.pyprophet_fixed_seed=<span class="org-constant">true</span>
params.pyprophet_use_legacy=<span class="org-constant">false</span>
</pre>
</div>

<p>
Below we are following the steps of <a href="http://www.openswath.org/en/latest/docs/pyprophet.html#scaling-up">http://www.openswath.org/en/latest/docs/pyprophet.html#scaling-up</a>
</p>

<p>
Anything in pyprophet that is not invoked with an <code>--out</code> flag
will overwrite the <code>--in</code> file,
here we only do that when the <code>--in</code>-file is created in the process
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span> (!params.pyprophet_use_legacy)
    {
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">pyprophet_subsample</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">dia_osw_file</span> from openswath_osw_ch.pyprophet_subsample
    <span class="org-keyword">val</span> <span class="org-function-name">subsample_ratio</span>
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">subsampled_osw</span>
    script:
    subsampled_osw=<span class="org-string">"</span><span class="org-variable-name">${dia_osw_file.baseName}</span><span class="org-string">.osws"</span>
    pyprophet_seed_flag=(params.pyprophet_fixed_seed ? <span class="org-string">"--test"</span> : <span class="org-string">"--no-test"</span>)
    <span class="org-string">"""</span>
<span class="org-string">    pyprophet subsample </span><span class="org-variable-name">$pyprophet_seed_flag</span><span class="org-string"> --in=</span><span class="org-variable-name">$dia_osw_file</span><span class="org-string"> --out=</span><span class="org-variable-name">$subsampled_osw</span><span class="org-string"> --subsample_ratio=</span><span class="org-variable-name">$subsample_ratio</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<p>
<code>pyprophet score</code> requires significantly more memory than <code>pyprophet merge</code>
should we split this?
</p>

<p>
If <code>pyprophet merge</code> or <code>pyprophet score</code> complains about no decoys being found,
you can try without passing <code>=--pyprophet_fixed_seed</code> to gladiator.
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">pyprophet_learn_classifier</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">subsampled_osws</span> from subsampled_osw.toSortedList()
    <span class="org-keyword">file</span> <span class="org-function-name">openswath_transitions</span> from openswath_transitions_ch_for_pyprophet.first()
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">osw_model</span>
    script:
    pyprophet_seed_flag=(params.pyprophet_fixed_seed ? <span class="org-string">"--test"</span> : <span class="org-string">"--no-test"</span>)
    osw_model=<span class="org-string">"model.osw"</span>
    <span class="org-string">"""</span>
<span class="org-string">    pyprophet merge --template=</span><span class="org-variable-name">$openswath_transitions</span><span class="org-string"> --out=</span><span class="org-variable-name">$osw_model</span><span class="org-string"> </span><span class="org-variable-name">$subsampled_osws</span>
<span class="org-string">    pyprophet score  </span><span class="org-variable-name">$pyprophet_seed_flag</span><span class="org-string"> --in=</span><span class="org-variable-name">$osw_model</span><span class="org-string"> --level=ms1ms2</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow">scored_osw_indirect_ch =<span class="org-type">Channel</span>.create()
scored_osw_indirect_ch.multiMap{<span class="org-variable-name">it</span> -&gt;
    reduce: backpropagate: <span class="org-variable-name">it</span>}.set{scored_osw_ch}

<span class="org-keyword">process</span> <span class="org-function-name">pyprophet_apply_classifier</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">osw_model</span> from osw_model.first()
    <span class="org-keyword">file</span> <span class="org-function-name">osw</span> from openswath_osw_ch.pyprophet_score
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">scored_osw</span> into scored_osw_indirect_ch
    script:
    pyprophet_seed_flag=(params.pyprophet_fixed_seed ? <span class="org-string">"--test"</span> : <span class="org-string">"--no-test"</span>)
    scored_osw=<span class="org-string">"</span><span class="org-variable-name">${osw.baseName}</span><span class="org-string">.scored.</span><span class="org-variable-name">${osw.Extension}</span><span class="org-string">"</span>
    <span class="org-string">"""</span>
<span class="org-string">    pyprophet score </span><span class="org-variable-name">$pyprophet_seed_flag</span><span class="org-string"> --in=</span><span class="org-variable-name">$osw</span><span class="org-string"> --out=</span><span class="org-variable-name">$scored_osw</span><span class="org-string"> --apply_weights=</span><span class="org-variable-name">$osw_model</span><span class="org-string"> --level=ms1ms2</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">pyprophet_reduce</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">scored_osw</span> from scored_osw_ch.reduce
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">reduced_scored_osw</span> 
    script:
    reduced_scored_osw=<span class="org-string">"</span><span class="org-variable-name">${file(scored_osw.baseName).baseName}</span><span class="org-string">.</span><span class="org-variable-name">${scored_osw.Extension}</span><span class="org-string">r"</span>
    <span class="org-string">"""</span>
<span class="org-string">    pyprophet reduce --in=</span><span class="org-variable-name">$scored_osw</span><span class="org-string"> --out=</span><span class="org-variable-name">$reduced_scored_osw</span>
<span class="org-string">    """</span>
}
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">pyprophet_control_error</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">reduced_scored_osws</span> from reduced_scored_osw.toSortedList()
    <span class="org-keyword">file</span> <span class="org-function-name">osw_model</span> from osw_model.first()
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">osw_global_model</span>
    script:
    osw_global_model=<span class="org-string">"model_global.osw"</span>
    <span class="org-string">"""</span>
<span class="org-string">    pyprophet merge --template=</span><span class="org-variable-name">$osw_model</span><span class="org-string"> --out=</span><span class="org-variable-name">$osw_global_model</span><span class="org-string"> </span><span class="org-variable-name">$reduced_scored_osws</span>
<span class="org-string">    pyprophet peptide --context=global --in=</span><span class="org-variable-name">$osw_global_model</span>
<span class="org-string">    pyprophet protein --context=global --in=</span><span class="org-variable-name">$osw_global_model</span>
<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">pyprophet_backpropagate</span> {
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">osw_scored</span> from scored_osw_ch.backpropagate
    <span class="org-keyword">file</span> <span class="org-function-name">osw_global_model</span> from osw_global_model.first()
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">dscore_tsv</span> into pyprophet_nonlegacy_ch
    script:
    base=<span class="org-string">"</span><span class="org-variable-name">${file(osw_scored.baseName).baseName}</span><span class="org-string">"</span>
    backproposw=<span class="org-string">"</span><span class="org-variable-name">${base}</span><span class="org-string">.backprop.osw"</span>
    dscore_tsv=<span class="org-string">"</span><span class="org-variable-name">${base}</span><span class="org-string">.tsv"</span>
    <span class="org-string">"""</span>
<span class="org-string">    pyprophet backpropagate --in="</span><span class="org-variable-name">$osw_scored</span><span class="org-string">" --apply_scores="</span><span class="org-variable-name">$osw_global_model</span><span class="org-string">" --out=</span><span class="org-variable-name">$backproposw</span>
<span class="org-string">    # we supply --format=legacy_merged so that pyprophet export respect the --out parameter</span>
<span class="org-string">    pyprophet export --in=</span><span class="org-variable-name">$backproposw</span><span class="org-string"> --format=legacy_merged --out=</span><span class="org-variable-name">$dscore_tsv</span>
<span class="org-string">    """</span>
}

</pre>
</div>
<p>
<a href="https://github.com/PyProphet/pyprophet/issues/49">https://github.com/PyProphet/pyprophet/issues/49</a>
</p>
<div class="org-src-container">
<pre class="src src-nextflow">}
</pre>
</div>
</div>
<div id="outline-container-orgeb39ba4" class="outline-6">
<h6 id="orgeb39ba4"><span class="section-number-6">5.7.2.2.1.</span> KeyError: 'HOME' in <code>pyprophet_subsample</code></h6>
<div class="outline-text-6" id="text-5-7-2-2-1">
<p>
If in this section you get the error
</p>
<blockquote>
<p>
Command error:
    File "/gnu/store/lvip6h5pamjwmvnkwg60sjb63ph8698k-python-3.9.9/lib/python3.9/os.py", line 679, in <span class="underline"><span class="underline">getitem</span></span>
      raise KeyError(key) from None
  KeyError: 'HOME'
</p>
</blockquote>
<p>
and you are using podman or docker,  you will want to add to your config file the following:
</p>
<div class="org-src-container">
<pre class="src src-nextflow">process {
    withName: <span class="org-string">'pyprophet_.*'</span> {
        containerOptions= <span class="org-string">'--env HOME="$PWD"'</span>
    }}
</pre>
</div>

<p>
This will set your home directory to the running directory in pyprophet processes
(Note, single quotes ( <code>'</code> ), rather than double-quotes ( <code>"</code> ), to prevent expansion
</p>
</div>
</div>
</div>


<div id="outline-container-org2c0159d" class="outline-5">
<h5 id="org2c0159d"><span class="section-number-5">5.7.2.3.</span> Choosing between legacy and nonlegacy pyprophet</h5>
<div class="outline-text-5" id="text-5-7-2-3">
<p>
Here we choose between legacy or nonlegacy pyprophet
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span> (params.pyprophet_use_legacy)
    pyprophet_legacy_ch.set{pyprophet_ch}
<span class="org-keyword">else</span>
    pyprophet_nonlegacy_ch.set{pyprophet_ch}
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-orgf718b08" class="outline-4">
<h4 id="orgf718b08"><span class="section-number-4">5.7.3.</span> feature-alignment</h4>
<div class="outline-text-4" id="text-5-7-3">
<p>
<a href="http://www.openswath.org/en/latest/docs/tric.html">http://www.openswath.org/en/latest/docs/tric.html</a>
of <code>msproteomicstools/analysis/alignment/feature_alignment.py</code>
</p>

<p>
Note: If things fail here,
because the right fdr cannot be reached,
you can try changing <code>precursor_mass_tolerance</code> and <code>fragment_mass_tolerance</code>
earlier upstream.
</p>

<p>
Additionally you can try setting <code>=--use_irt</code> to false, if you are getting a zero-division errors because the input tsvs are empty.
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">feature_alignment</span>
{
    publishDir <span class="org-string">"</span><span class="org-variable-name">${params.outdir}</span><span class="org-string">/dia/"</span>
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">dscore_csvs</span> from pyprophet_ch.toSortedList()
    output:
    <span class="org-keyword">file</span> <span class="org-function-name">outfile</span> into feature_alignment_ch
    script:
    outfile = <span class="org-string">"DIA-analysis-results.csv"</span>

    <span class="org-keyword">if</span> (params.use_irt) {
        realign_method = <span class="org-string">"diRT"</span> 
    } <span class="org-keyword">else</span> {
        realign_method = <span class="org-string">"linear"</span>
    }

    <span class="org-string">"feature_alignment.py "</span> +
        <span class="org-string">"--method best_overall "</span> +
        <span class="org-string">"--max_rt_diff 90 "</span> +
        <span class="org-string">"--target_fdr </span><span class="org-variable-name">$params.tric_target_fdr</span><span class="org-string"> "</span> +
        <span class="org-string">"--max_fdr_quality </span><span class="org-variable-name">$params.tric_max_fdr</span><span class="org-string"> "</span> +
        <span class="org-string">"--in </span><span class="org-variable-name">$dscore_csvs</span><span class="org-string"> "</span> +         <span class="org-comment-delimiter">// </span><span class="org-comment">will this break on filenames with spaces</span>
        <span class="org-string">" --realign_method </span><span class="org-variable-name">$realign_method</span><span class="org-string"> "</span> + 
        <span class="org-string">"--out </span><span class="org-variable-name">$outfile</span><span class="org-string">"</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">Target FDR used in TRIC alignment in dirT mode [default 0.01]</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">This was "trig_target_pvalue" in the original python gladiator implementation</span>
params.tric_target_fdr=0.01
<span class="org-comment-delimiter">// </span><span class="org-comment">Maximum FDR for TRIC alignment in dirT mode [default 0.05]</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">This was "trig_max_pvalue" in the original python gladiator implementation</span>
params.tric_max_fdr=0.05
</pre>
</div>
</div>
</div>

<div id="outline-container-org105e35f" class="outline-4">
<h4 id="org105e35f"><span class="section-number-4">5.7.4.</span> Swath2stats</h4>
<div class="outline-text-4" id="text-5-7-4">
<div class="org-src-container">
<pre class="src src-R" id="org8d79a37">suppressPackageStartupMessages(library(SWATH2stats))
suppressPackageStartupMessages(library(data.table))

remove_irt &lt;- function(df)
  df[grep("iRT", df[["ProteinName"]], invert=TRUE, fixed=TRUE),, drop=FALSE]

## original gladiator decoy removing behaviour
remove_decoy_strict &lt;- function(df,decoyprefix)
  df[grep(decoyprefix, df[["ProteinName"]], invert=TRUE, fixed=TRUE),, drop=FALSE]


remove_decoy_loose &lt;- function(df)
  df[!df[["decoy"]],, drop = FALSE]


basename_sans_ext &lt;- function(f)
  unlist(strsplit(basename(f), ".",fixed=TRUE))[[1]]


main &lt;- function(diafile,
		 strict_checking=FALSE,
		 peptideoutputfile="",
		 proteinoutputfile="",
		 decoyprefix="DECOY_")
{
  remove_decoy &lt;- `if`(strict_checking,
		       function(df) remove_decoy_strict(df,decoyprefix),
		       remove_decoy_loose)
  filtered_data &lt;-
    data.table::fread(diafile,header=TRUE) |&gt; 
    data.frame(stringsAsFactors = FALSE) |&gt;
    within(run_id &lt;- basename(filename)) |&gt;
    SWATH2stats::reduce_OpenSWATH_output() |&gt;
    remove_irt() |&gt;
    remove_decoy()

  # Writing output
  filtered_data |&gt;
    SWATH2stats::write_matrix_peptides(filename = basename_sans_ext(peptideoutputfile)) |&gt;
    write.table(sep="\t",file=peptideoutputfile,row.names = FALSE)

  filtered_data |&gt;
    SWATH2stats::write_matrix_proteins(filename = basename_sans_ext(proteinoutputfile)) |&gt;
    write.table(sep="\t",file=proteinoutputfile,row.names = FALSE)
}
</pre>
</div>

<p>
Where the field "Decoy" is 1, thats a decoy generated by <code>OpenSwathDecoyGenerator</code>,
rather than from the fastadataasee of <code>DecoyDatabase</code>
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">process</span> <span class="org-function-name">swath2stats</span> {
    publishDir <span class="org-string">"</span><span class="org-variable-name">${params.outdir}</span><span class="org-string">/dia/"</span>
    input:
    <span class="org-keyword">file</span> <span class="org-function-name">dia_score</span> from feature_alignment_ch

    output:
    <span class="org-keyword">file</span> <span class="org-function-name">peptide_matrix</span>
    <span class="org-keyword">file</span> <span class="org-function-name">protein_matrix</span>

    script:
    strict_checking=params.swath2stats_strict_checking
    peptide_matrix=<span class="org-string">"DIA-peptide-matrix.tsv"</span>
    protein_matrix=<span class="org-string">"DIA-protein-matrix.tsv"</span>

    <span class="org-string">"""</span>
<span class="org-string">    #!/usr/bin/env Rscript</span>
<span class="org-string">    """</span> +
        <span class="org-string">'''</span>
<span class="org-string">&lt;&lt;r-swath2stats&gt;&gt;</span>
<span class="org-string">        '''</span> +
        <span class="org-string">"""</span>
<span class="org-string">    main("</span><span class="org-variable-name">$dia_score</span><span class="org-string">", strict_checking = as.logical("</span><span class="org-variable-name">$strict_checking</span><span class="org-string">"),</span>
<span class="org-string">        peptideoutputfile="</span><span class="org-variable-name">$peptide_matrix</span><span class="org-string">",</span>
<span class="org-string">        proteinoutputfile="</span><span class="org-variable-name">$protein_matrix</span><span class="org-string">",</span>
<span class="org-string">        decoyprefix="DECOY_")</span>

<span class="org-string">    """</span>
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">whether to exclude in final DIA matrices</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">proteins of which</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">any peptide can be a decoy.</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">this is the default behaviour of original gladiator implementation.</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">if set to false, just instead remove anything</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">that has the "decoy" column set to false</span>
params.swath2stats_strict_checking=<span class="org-constant">true</span>
</pre>
</div>
</div>


<div id="outline-container-org5e1c307" class="outline-5">
<h5 id="org5e1c307"><span class="section-number-5">5.7.4.1.</span> Dependencies</h5>
<div class="outline-text-5" id="text-5-7-4-1">
<div class="org-src-container">
<pre class="src src-scheme" id="orgca5aeb1">(<span class="org-string">"swath2stats"</span>
 <span class="org-string">"r-minimal"</span>
 <span class="org-string">"r-swath2stats"</span>)
</pre>
</div>


<p>
The above as an R-script
</p>

<div class="org-src-container">
<pre class="src src-R">#!/usr/bin/env Rscript

.libPaths("/opt/Rlibs/")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.15",ask=FALSE,update=FALSE)
BiocManager::install("SWATH2stats", ask=FALSE,update=FALSE)

</pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="outline-container-org8a7d27f" class="outline-2">
<h2 id="org8a7d27f"><span class="section-number-2">6.</span> Configuration for backends</h2>
<div class="outline-text-2" id="text-6">
</div>
<div id="outline-container-org97679a9" class="outline-3">
<h3 id="org97679a9"><span class="section-number-3">6.1.</span> Defining the config template</h3>
<div class="outline-text-3" id="text-6-1">
<p>
Here we define what the registry is that hosts our images
</p>
<pre class="example">
docker.io/elolabfi/
</pre>


<p>
This is the template that we fill with the container manager program to be used (<code>%p</code>),
the gladiator container uri to be used (<code>%G</code>),
and the pyprophet legacy container to be used (<code>%P</code>).
</p>

<pre class="example" id="orge334db2">
%p.enabled=true
process {
   container='%G'
   withName: 'DeepDIA.*' {
      container='%D'
   }
   
   // for nonlegacy pyprophet processes
   withName: 'pyprophet_.*' {
      container='%N'
   }

   // because this is the more specific rule
   // we apply it last, so that it overrides the 'pyprophet_.*' rule if
   // this rule also applies
    withName: pyprophet_legacy {
	container='%P'
   }
   // for perl-diamspep
   withName: 'DIAMS2PEP_.*'{
   // there is really no good single letter here
   	container='%V'
   }
}
</pre>

<p>
And here we write the logic that fills the template
</p>

<div class="org-src-container">
<pre class="src src-emacs-lisp" id="orgd92d0cf">(<span class="org-keyword">defvar</span> <span class="org-variable-name">container-property-alist</span>
  (<span class="org-keyword">let</span> ((remote-registry registry))
    `((<span class="org-string">"remote"</span>
       (<span class="org-string">"singularity"</span>
        (<span class="org-string">"access-protocol"</span> . <span class="org-string">"docker://"</span>)
        (<span class="org-string">"registry"</span> . ,remote-registry)
        (<span class="org-string">"container-suffix"</span> . <span class="org-string">":0.1.4-0-f7abd8"</span>))
       (<span class="org-string">"docker"</span>
        (<span class="org-string">"access-protocol"</span> . <span class="org-string">""</span>)
        (<span class="org-string">"registry"</span> . ,remote-registry)
        (<span class="org-string">"container-suffix"</span> . <span class="org-string">":0.1.4-0-f7abd8"</span>))
       (<span class="org-string">"podman"</span>
        (<span class="org-string">"access-protocol"</span> . <span class="org-string">"docker://"</span>)
        (<span class="org-string">"registry"</span> . ,remote-registry)
        (<span class="org-string">"container-suffix"</span> . <span class="org-string">":0.1.4-0-f7abd8"</span>)))
      (<span class="org-string">"local"</span>
       (<span class="org-string">"singularity"</span>
        (<span class="org-string">"access-protocol"</span> . <span class="org-string">"file://"</span>)
        (<span class="org-string">"registry"</span> . <span class="org-string">"containers/"</span>)
        (<span class="org-string">"container-suffix"</span> . <span class="org-string">".simg"</span>))
       (<span class="org-string">"docker"</span>
        (<span class="org-string">"access-protocol"</span> . <span class="org-string">""</span>)
        (<span class="org-string">"registry"</span> . <span class="org-string">"localhost/"</span>)
        (<span class="org-string">"container-suffix"</span> . <span class="org-string">""</span>))
       (<span class="org-string">"podman"</span>
        (<span class="org-string">"access-protocol"</span> . <span class="org-string">""</span>)
        (<span class="org-string">"registry"</span> . <span class="org-string">"localhost/"</span>)
        (<span class="org-string">"container-suffix"</span> . <span class="org-string">""</span>))))))

(<span class="org-keyword">defun</span> <span class="org-function-name">construct-container-uri</span> (basename program-name locality)
  (<span class="org-keyword">let</span> ((properties
         (alist-get
          program-name
          (alist-get locality container-property-alist nil nil 'equal)
          nil nil 'equal)))
    (<span class="org-keyword">unless</span> properties
      (<span class="org-warning">error</span> <span class="org-string">"Could not find properties for '</span><span class="org-string"><span class="org-constant">%s</span></span><span class="org-string">' '</span><span class="org-string"><span class="org-constant">%s</span></span><span class="org-string">'"</span> program-name locality))
    (format
     <span class="org-string">"%s%s%s%s"</span>
     (alist-get <span class="org-string">"access-protocol"</span> properties nil nil 'equal)
     (alist-get <span class="org-string">"registry"</span> properties  nil nil 'equal)
     basename
     (alist-get <span class="org-string">"container-suffix"</span> properties nil nil 'equal))))


(<span class="org-keyword">cl-defun</span> <span class="org-function-name">construct-container-config</span> (template program-name locality <span class="org-type">&amp;key</span> (gladiator-image <span class="org-string">"gladiator"</span>))
  (<span class="org-keyword">let</span> ((program-name-key
         program-name))
    (format-spec template
               `((?p . ,program-name)
                 (?G . 
                     ,(construct-container-uri gladiator-image
                                               program-name-key
                                               locality))
                 (?D .
                     ,(construct-container-uri <span class="org-string">"deepdia"</span>
                                               program-name-key
                                               locality))
                 (?N .
                     ,(construct-container-uri <span class="org-string">"pyprophet"</span>
                                               program-name-key
                                               locality))
                 (?V .
                     ,(construct-container-uri
                       <span class="org-string">"diams2pep"</span>
                       program-name-key
                       locality))

                 (?P .
                     ,(construct-container-uri
                       <span class="org-string">"pyprophet-legacy"</span>
                       program-name-key
                       locality))))))

(construct-container-config template program-name locality <span class="org-builtin">:gladiator-image</span> gladiator-image)

</pre>
</div>



<pre class="example">
docker://registry.gitlab.utu.fi/elixirdianf/gladiator-notes/test
</pre>


<pre class="example">
https://registry.gitlab.utu.fi/elixirdianf/gladiator-notes/gladiator
</pre>
</div>
</div>
<div id="outline-container-org0d0a1bc" class="outline-3">
<h3 id="org0d0a1bc"><span class="section-number-3">6.2.</span> Per backend config files</h3>
<div class="outline-text-3" id="text-6-2">
<p>
So here we fill the templates that we defined in the previous section.
If you are reading this as an org-mode file,
this headings default tangle argument makes each blocks tangle output file
based on the name of the block
</p>

<p>
This block is whats common between singularity
</p>
<div class="org-src-container">
<pre class="src src-nextflow" id="org03985f5">singularity.runOptions = <span class="org-string">'-B $TMPDIR:/tmp'</span>
singularity.autoMounts=<span class="org-constant">true</span>
</pre>
</div>


<p>
This block is whats common between singularity and podman.
On certain machines (reported on Ubuntu 22.04.5 LTS, 6.2.0-26-generic, with podman 3.4.4)
</p>

<p>
matplotlib in pyprophet in podman tries to read the home directory.
which does not exist in the container, which will result in the error described <a href="#orgeb39ba4">here</a>. 
</p>
<div class="org-src-container">
<pre class="src src-nextflow" id="orgd350159">process {
    withName: <span class="org-string">'pyprophet_.*'</span> {
        containerOptions= <span class="org-string">'--env HOME="$PWD"'</span>
    }
}
</pre>
</div>


<div class="org-src-container">
<pre class="src src-nextflow" id="org1ecf0c2">nil
singularity.runOptions = <span class="org-string">'-B $TMPDIR:/tmp'</span>
singularity.autoMounts=<span class="org-constant">true</span>
</pre>
</div>

<p>
#+end_src
</p>
<div class="org-src-container">
<pre class="src src-nextflow" id="orga183107">&lt;&lt;construct-container-config(<span class="org-string">"singularity"</span>,<span class="org-string">"local"</span>)&gt;&gt;
&lt;&lt;singularity-general-options&gt;&gt;
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow" id="org745a3ca">nil
singularity.runOptions = <span class="org-string">'-B $TMPDIR:/tmp'</span>
singularity.autoMounts=<span class="org-constant">true</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow" id="orgf05b81c">&lt;&lt;construct-container-config(<span class="org-string">"docker"</span>,<span class="org-string">"remote"</span>)&gt;&gt;
&lt;&lt;dockerlike-general-options&gt;&gt;    
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow" id="org71ea3ce">&lt;&lt;construct-container-config(<span class="org-string">"docker"</span>,<span class="org-string">"local"</span>)&gt;&gt;
&lt;&lt;dockerlike-general-options&gt;&gt;
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow" id="org2d18108">&lt;&lt;construct-container-config(<span class="org-string">"podman"</span>,<span class="org-string">"remote"</span>)&gt;&gt;
&lt;&lt;dockerlike-general-options&gt;&gt;
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow" id="org8b8528e">&lt;&lt;construct-container-config(<span class="org-string">"podman"</span>,<span class="org-string">"local"</span>)&gt;&gt;
&lt;&lt;dockerlike-general-options&gt;&gt;
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow" id="orgdde2141">process {
   beforeScript=<span class="org-string">"source &lt;(guix time-machine -C </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/gladiator-guix-channels.scm -- shell --search-paths -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/gladiator.scm -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/nextflow-trace.scm)"</span>
   withName: <span class="org-string">"DeepDIA.*"</span> {
      beforeScript=<span class="org-string">"source &lt;(guix time-machine -C </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/deepdia-channels.scm -- shell --search-paths -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/deepdia.scm -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/nextflow-trace.scm)"</span>
   }

   <span class="org-comment-delimiter">// </span><span class="org-comment">for nonlegacy pyprophet processes</span>
   withName: <span class="org-string">"pyprophet_.*"</span> {
      beforeScript=<span class="org-string">"source &lt;(guix time-machine -C </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/gladiator-guix-channels.scm -- shell --search-paths -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/pyprophet.scm -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/nextflow-trace.scm)"</span>
   }

   <span class="org-comment-delimiter">// </span><span class="org-comment">because this is the more specific rule</span>
   <span class="org-comment-delimiter">// </span><span class="org-comment">we apply it last, so that it overrides the "pyprophet_.*" rule if</span>
   <span class="org-comment-delimiter">// </span><span class="org-comment">this rule also applies</span>
    withName: pyprophet_legacy {
        beforeScript=<span class="org-string">"source &lt;(guix time-machine -C </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/pyprophet-legacy-channels.scm -- shell --search-paths -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/pyprophet-legacy.scm -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/nextflow-trace.scm)"</span>
   }
   <span class="org-comment-delimiter">// </span><span class="org-comment">for perl-diamspep</span>
   withName: <span class="org-string">"DIAMS2PEP_.*"</span>{
   <span class="org-comment-delimiter">// </span><span class="org-comment">there is really no good single letter here</span>
        beforeScript=<span class="org-string">"source &lt;(guix time-machine -C </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/diams2pep-channels.scm -- shell --search-paths -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/diams2pep.scm -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/nextflow-trace.scm)"</span>
   }

   withName: <span class="org-string">"DIAtracer"</span> {
     beforeScript=<span class="org-string">"source &lt;(guix time-machine -C </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/diatracer-channels.scm -- shell --search-paths -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/diatracer.scm -m </span><span class="org-variable-name">$projectDir</span><span class="org-string">/ci/guix/manifests/nextflow-trace.scm)"</span>
   }
}
</pre>
</div>
</div>
</div>
<div id="outline-container-orgedbdfbf" class="outline-3">
<h3 id="orgedbdfbf"><span class="section-number-3">6.3.</span> Backend Specific Issues</h3>
<div class="outline-text-3" id="text-6-3">
</div>
<div id="outline-container-org928fcbb" class="outline-4">
<h4 id="org928fcbb"><span class="section-number-4">6.3.1.</span> Podman specific Issues</h4>
<div class="outline-text-4" id="text-6-3-1">
<p>
If you get the error in any process
</p>
<blockquote>
<p>
Error: OCI runtime error: the requested cgroup controller `cpu` is not available
</p>
</blockquote>
<p>
This is because nextflow sets the number of CPU's podman can use, but you do not have the cpu cgroup in podman.
</p>

<p>
If, when you do
</p>
<div class="org-src-container">
<pre class="src src-sh">podman info  --format={{<span class="org-string">".Host.CgroupControllers"</span>}}
</pre>
</div>
<p>
you do <span class="underline">not</span> see <code>cpu</code> listed, but when you do
</p>
<div class="org-src-container">
<pre class="src src-sh">podman --cgroup-manager cgroupfs info  --format={{<span class="org-string">".Host.CgroupControllers"</span>}}
</pre>
</div>
<p>
you <span class="underline">do</span> see <code>cpu</code> listed,
you will want to add to your nextflow config file:
</p>

<div class="org-src-container">
<pre class="src src-nextflow">podman.engineOptions=<span class="org-string">"--cgroup-manager cgroupfs"</span>
</pre>
</div>
</div>
</div>
</div>
</div>


<div id="outline-container-orgf5bf9a7" class="outline-2">
<h2 id="orgf5bf9a7"><span class="section-number-2">7.</span> SDRF support</h2>
<div class="outline-text-2" id="text-7">
<p>
You can supply a path to an sdrf file with <code>--sdrf=filename</code>
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">path to a sdrf file </span>
params.sdrf = <span class="org-constant">null</span>
</pre>
</div>

<p>
The standard is defined here <a href="https:://github.com/bigbio/proteomics-metadata-standard">https:://github.com/bigbio/proteomics-metadata-standard</a>
See the <code>annoted-projects</code> directory for examples,
Good examples is <code>PXD003977</code>
</p>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">import</span> nextflow.splitter.<span class="org-type">CsvSplitter</span>

<span class="org-keyword">def</span> <span class="org-function-name">readSDRF</span>(filename,params)
{
    <span class="org-keyword">def</span> <span class="org-variable-name">sdrf_file</span>=file(filename)
    <span class="org-keyword">def</span> <span class="org-variable-name">options</span>=[:]
    <span class="org-keyword">def</span> <span class="org-variable-name">errors</span>=[]
    <span class="org-keyword">def</span> <span class="org-variable-name">warnings</span>=[]
    <span class="org-keyword">def</span> <span class="org-variable-name">handlers</span> =[
        <span class="org-keyword">this</span>.&amp;parseSDRFFileUri,
        {_sdrf_fields,_params -&gt;
            parseSDRFTolerance(_sdrf_fields, _params,
                                                    field_name = <span class="org-string">"comment[fragment mass tolerance]"</span>,  flag_name=<span class="org-string">"fragment_mass_tolerance"</span>,supported_units=[<span class="org-string">"Da"</span>])},
        {_sdrf_fields,_params -&gt;
            parseSDRFTolerance(_sdrf_fields, _params,
                                                    field_name = <span class="org-string">"comment[precursor mass tolerance]"</span>,  flag_name=<span class="org-string">"precursor_mass_tolerance"</span>,supported_units=[<span class="org-string">"ppm"</span>])},
    ]
    <span class="org-keyword">def</span> <span class="org-variable-name">sdrf_fields</span> = <span class="org-keyword">new</span> <span class="org-type">CsvSplitter</span>().target(sdrf_file.text).options(header:<span class="org-constant">true</span>,sep:<span class="org-string">'        '</span>).list()
    sdrf_fields = sdrf_fields.collect(
        {entry -&gt;
            entry.collectEntries(
                {key, row -&gt;
                    [key,
                     (row ==<span class="org-string">"not available"</span> ||
                      row == <span class="org-string">"not applicable"</span>) ? 
                     <span class="org-constant">null</span> :
                     row]})})


    <span class="org-keyword">for</span>(handler <span class="org-keyword">in</span> handlers) {
        <span class="org-keyword">def</span> <span class="org-variable-name">val</span> = handler(sdrf_fields, params);
        errors += val.errors
        warnings+= val.warnings
        options+=val.values
    }
    <span class="org-keyword">for</span>(warning <span class="org-keyword">in</span> warnings)
<span class="org-keyword">        print</span>(<span class="org-string">"WARNING: "</span> + warnings)
    <span class="org-keyword">for</span>(error <span class="org-keyword">in</span> errors)
<span class="org-keyword">        print</span>(<span class="org-string">"ERROR: "</span> + error)
    <span class="org-keyword">if</span>(errors)
        <span class="org-keyword">throw</span> <span class="org-keyword">new</span> <span class="org-type">Exception</span>(<span class="org-string">"Errors in SDRF file, see above messages"</span>)
    <span class="org-keyword">return</span> options
}
</pre>
</div>

<p>
Fields we handle:
</p>
</div>
<div id="outline-container-org602b580" class="outline-3">
<h3 id="org602b580"><span class="section-number-3">7.1.</span> <code>comment[file uri]</code></h3>
<div class="outline-text-3" id="text-7-1">
<p>
We can handle remote and local files in <code>comment[associated file uri]</code>
as long as they are in mzML format.
This can be used i.l.o <code>params.diafiles</code>
(though other libre formats would also be possible if we add an msconvert step first)
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">this</span>.&amp;parseSDRFFileUri,
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">def</span> <span class="org-function-name">parseSDRFFileUri</span>(sdrf_fields, params){
    <span class="org-keyword">def</span> <span class="org-variable-name">raise_error_on_non_mzml</span>=<span class="org-constant">true</span>
    <span class="org-keyword">def</span> <span class="org-variable-name">retval</span>= [values:[:],warnings:[],errors:[]]
    <span class="org-keyword">def</span> <span class="org-variable-name">field_name</span> = <span class="org-string">'comment[file uri]'</span>
    <span class="org-keyword">def</span> <span class="org-variable-name">flag_name</span> = <span class="org-string">'--diafiles'</span>
    <span class="org-keyword">def</span> <span class="org-variable-name">files</span> = sdrf_fields*.get(field_name)
    <span class="org-comment-delimiter">// </span><span class="org-comment">here we intentionally don't break the switch statement so that we can accumulate retval.errors,</span>
    <span class="org-comment-delimiter">// </span><span class="org-comment">so that the user can know all at once.</span>
    <span class="org-keyword">switch</span> (<span class="org-constant">true</span>) {
        <span class="org-keyword">case</span> (params.diafiles):
            retval.values+=[diafiles:params.diafiles]
            <span class="org-keyword">break</span>;
    <span class="org-keyword">case</span>(files == <span class="org-constant">null</span>):
            retval.errors += [<span class="org-string">"No column named '</span><span class="org-variable-name">$field_name</span><span class="org-string">' in sdrf. Add one or supply files with  </span><span class="org-variable-name">$flag_name</span><span class="org-string">"</span>];
            <span class="org-keyword">break</span>;
        <span class="org-keyword">case</span> (isallnull(files) || !files):
            retval.errors += [<span class="org-string">"All retval.values are missing in  supplied in sdrf '</span><span class="org-variable-name">$field_name</span><span class="org-string">'. Enter these or supply CLI </span><span class="org-variable-name">$flag_name</span><span class="org-string">];"</span>]
        <span class="org-keyword">case</span> (!isallnull(files) &amp;&amp; issomenull(files)):
            retval.errors += [<span class="org-string">"Some entries in sdrf '</span><span class="org-variable-name">$field_name</span><span class="org-string">'  are not given, fill in or supply </span><span class="org-variable-name">$flag_name</span><span class="org-string">"</span>];
        <span class="org-keyword">case</span> (!isallnull(files) &amp;&amp; raise_error_on_non_mzml &amp;&amp; files.any({x -&gt; x  &amp;&amp; file(x).getExtension() != <span class="org-string">"mzML"</span>})):
            retval.errors += [<span class="org-string">"Some entries in sdrf  '</span><span class="org-variable-name">$field_name</span><span class="org-string">' are not mzML files"</span>];
            <span class="org-comment-delimiter">// </span><span class="org-comment">if there were no retval.errors</span>
            <span class="org-comment-delimiter">// </span><span class="org-comment">then the files field in the sdrf is correct</span>
        <span class="org-keyword">case</span>(!retval.errors):
            retval.values+=[diafiles:files]
            <span class="org-keyword">break</span>;
        <span class="org-keyword">case</span>(retval.errors):
            retval.values+=[diafiles:<span class="org-constant">null</span>]
            <span class="org-keyword">break</span>;
    }
    <span class="org-keyword">return</span> retval
}
</pre>
</div>
</div>
</div>

<div id="outline-container-orge5528da" class="outline-3">
<h3 id="orge5528da"><span class="section-number-3">7.2.</span> <code>comment[precursor mass tolerance]</code>, <code>comment[fragment mass tolerance]</code></h3>
<div class="outline-text-3" id="text-7-2">
<p>
We require (for now)
precursor mass tolerance to be ppm, and to be the same for all the samples.
</p>

<div class="org-src-container">
<pre class="src src-nextflow">{_sdrf_fields,_params -&gt;
    parseSDRFTolerance(_sdrf_fields, _params,
                                            field_name = <span class="org-string">"comment[fragment mass tolerance]"</span>,  flag_name=<span class="org-string">"fragment_mass_tolerance"</span>,supported_units=[<span class="org-string">"Da"</span>])},
{_sdrf_fields,_params -&gt;
    parseSDRFTolerance(_sdrf_fields, _params,
                                            field_name = <span class="org-string">"comment[precursor mass tolerance]"</span>,  flag_name=<span class="org-string">"precursor_mass_tolerance"</span>,supported_units=[<span class="org-string">"ppm"</span>])},
</pre>
</div>

<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">def</span> <span class="org-function-name">parseSDRFTolerance</span>(sdrf_fields, params,  field_name = <span class="org-string">"comment[precursor mass tolerance]"</span>,  flag_name=<span class="org-string">"precursor_mass_tolerance"</span>,supported_units=[<span class="org-string">"ppm"</span>]){
    <span class="org-keyword">def</span> <span class="org-variable-name">retval</span> = [values:[:], warnings:[], errors:[]]
    <span class="org-keyword">def</span> <span class="org-variable-name">tolerances</span> = sdrf_fields*.get(field_name)
    <span class="org-keyword">switch</span>(<span class="org-constant">true</span>){
        <span class="org-keyword">case</span>(params.get(flag_name) != <span class="org-constant">null</span>):
            retval.values[flag_name] = params.get(field_name);
            <span class="org-keyword">break</span>
        <span class="org-keyword">case</span>(tolerances == <span class="org-constant">null</span>):
            retval.errors += [<span class="org-string">"No column named '</span><span class="org-variable-name">$field_name</span><span class="org-string">' in sdrf. Add one or supply  --</span><span class="org-variable-name">$flag_name</span><span class="org-string">"</span>];
            <span class="org-keyword">break</span>;
        <span class="org-keyword">case</span> (isallnull(tolerances) || !tolerances):
            retval.errors += [<span class="org-string">"All values are missing in  supplied in sdrf '</span><span class="org-variable-name">$field_name</span><span class="org-string">'. Enter these or supply CLI --</span><span class="org-variable-name">$flag_name</span><span class="org-string">];"</span>]
        <span class="org-keyword">case</span> (!isallnull(tolerances) &amp;&amp; !tolerances.every()):
            retval.errors += [<span class="org-string">"Some entries in sdrf '</span><span class="org-variable-name">$field_name</span><span class="org-string">'  are not given, fill in or supply </span><span class="org-variable-name">$flag_name</span><span class="org-string">"</span>];
        <span class="org-keyword">case</span>(tolerances.any({<span class="org-variable-name">it</span> &amp;&amp; <span class="org-variable-name">it</span>.split().size() &gt; 1 &amp;&amp; !(supported_units.contains(<span class="org-variable-name">it</span>.split()[1]))})):
            retval.errors += [<span class="org-string">"Some entries in sdrf '</span><span class="org-variable-name">$field_name</span><span class="org-string">' have an unsupported unit(supported units for </span><span class="org-variable-name">$field_name</span><span class="org-string">: </span><span class="org-variable-name">$supported_units</span><span class="org-string">)"</span>];
        <span class="org-keyword">case</span>(tolerances.any({<span class="org-variable-name">it</span> &amp;&amp; <span class="org-variable-name">it</span>.split().size() &lt; 2})):
            retval.warnings += [<span class="org-string">"Some entries in sdrd '</span><span class="org-variable-name">$field_name</span><span class="org-string">' do not have a unit, assuming "</span> + supported_units[0]]
        <span class="org-keyword">case</span>(tolerances.unique().size() != 1):
            retval.errors += [<span class="org-string">"Gladiator currently requires the </span><span class="org-variable-name">$field_name</span><span class="org-string"> to be the same for all the samples, change it or supply --</span><span class="org-variable-name">$flag_name</span><span class="org-string">"</span>];
    <span class="org-keyword">case</span>(!retval.errors):
            retval.values[flag_name] = tolerances.unique()[0].split()[0]
    }
    <span class="org-keyword">return</span> retval
}
</pre>
</div>
</div>
</div>



<div id="outline-container-org1af7273" class="outline-3">
<h3 id="org1af7273"><span class="section-number-3">7.3.</span> Utility functions</h3>
<div class="outline-text-3" id="text-7-3">
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">def</span> <span class="org-function-name">isallnull</span>(value)
{
    value == <span class="org-constant">null</span> || value.every({x-&gt; x==<span class="org-constant">null</span>})
}
<span class="org-keyword">def</span> <span class="org-function-name">issomenull</span>(value)
{
    value == <span class="org-constant">null</span> || value.any({x -&gt; x == <span class="org-constant">null</span>})
}
</pre>
</div>
</div>
</div>

<div id="outline-container-org9fd977a" class="outline-3">
<h3 id="org9fd977a"><span class="section-number-3">7.4.</span> Testing</h3>
<div class="outline-text-3" id="text-7-4">
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">print</span>(readSDRF(<span class="org-string">"annotated-projects/PXD003977/PXD003977.sdrf.tsv"</span>,params))
</pre>
</div>
</div>
</div>
<div id="outline-container-org44cd314" class="outline-3">
<h3 id="org44cd314"><span class="section-number-3">7.5.</span> Updating params</h3>
<div class="outline-text-3" id="text-7-5">
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-keyword">if</span> (params.sdrf){
    params = readSDRF(params.sdrf, params) + params
}
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org8de5dc7" class="outline-2">
<h2 id="org8de5dc7"><span class="section-number-2">8.</span> Putting it together</h2>
<div class="outline-text-2" id="text-8">
</div>
<div id="outline-container-org6d834a7" class="outline-3">
<h3 id="org6d834a7"><span class="section-number-3">8.1.</span> gwl</h3>
<div class="outline-text-3" id="text-8-1">
<div class="org-src-container">
<pre class="src src-scheme">(make-workflow
 (name <span class="org-string">"my-workflow"</span>)
 (processes
  (auto-connect
   (list
    <span class="org-type">&lt;&lt;gwl-proc&gt;&gt;</span>
    ))))
</pre>
</div>
</div>
</div>
<div id="outline-container-org2a47adf" class="outline-3">
<h3 id="org2a47adf"><span class="section-number-3">8.2.</span> nf</h3>
<div class="outline-text-3" id="text-8-2">
<p>
according to <code>nextflow/docs/process.rst</code>,
the cache looks at the timestamp for cache-correctness,
which can be inconsistent in on shared file systems,
setting <code>process.cache</code> to 'lenient' maybe will work around this?
See alos 
</p>
<blockquote>
<p>
``'lenient'``         Enable caching. Cache keys are created indexing input files path and size attributes (this policy provides a workaround for incorrect caching invalidation observed on shared file systems due to inconsistent files timestamps; requires version 0.32.x or later).
</p>
</blockquote>

<p>
there is also 'deep' hashing, based on file content
and the undocumented 'sha256' hashing bashed on the shasum of the file
See also here: <a href="https://www.nextflow.io/blog/2019/troubleshooting-nextflow-resume.html">https://www.nextflow.io/blog/2019/troubleshooting-nextflow-resume.html</a>
</p>

<div class="org-src-container">
<pre class="src src-nextflow">process.cache=<span class="org-string">'lenient'</span>
</pre>
</div>

<p>
though this will break if you replace input files with files with the same
path but different content.
'sha256' might be the best one?
</p>

<p>
The results files will all be output in the following directory
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">directory where the  results will be output to </span>
params.outdir = <span class="org-string">"./results"</span>
</pre>
</div>




<div class="org-src-container">
<pre class="src src-nextflow">&lt;&lt;nf-params&gt;&gt;
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-orgac9c4be" class="outline-2">
<h2 id="orgac9c4be"><span class="section-number-2">9.</span> File local variables</h2>
<div class="outline-text-2" id="text-9">
<p>
For this file
</p>

<p>
For the scheme file
</p>

<div class="org-src-container">
<pre class="src src-scheme"><span class="org-comment-delimiter">;; </span><span class="org-comment">Local Variables;</span>
<span class="org-comment-delimiter">;; </span><span class="org-comment">compile-command: "guix shell guix guile gwl --with-git-url=gwl=git://git.savannah.gnu.org/gwl.git --with-commit=gwl=e233be5cf0e2f9cb37e3daa299f5031bea56ba71 -- guix workflow run -v10 gwl-gladiator.scm</span>
<span class="org-comment-delimiter">;; </span><span class="org-comment">End:</span>
</pre>
</div>
<p>
For the nextflow file
</p>
<div class="org-src-container">
<pre class="src src-nextflow"><span class="org-comment-delimiter">// </span><span class="org-comment">Local Variables&#58;</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">compile-command: "module use /appl/user/modulefiles/ &amp;&amp; module load nextflow &amp;&amp; salloc nextflow run gladiator.nf"</span>
<span class="org-comment-delimiter">// </span><span class="org-comment">End:</span>
</pre>
</div>
</div>
</div>



<div id="outline-container-org84da839" class="outline-2">
<h2 id="org84da839"><span class="section-number-2">10.</span> Note</h2>
<div class="outline-text-2" id="text-10">
<p>
Pyprophet calling convention altered after v2.1.0,
check what changed
so can't literally copy invocation from <code>workflow.py</code>.
(S. said)
</p>
</div>
</div>
<div id="outline-container-org186f487" class="outline-2">
<h2 id="org186f487"><span class="section-number-2">11.</span> Launching</h2>
<div class="outline-text-2" id="text-11">
<pre class="example" id="orgb81f50c">
module use /appl/user/modulefiles/
module load nextflow
mkdir -p runs
RUNLOGDIR=`mktemp -p runs -d run.XXXX`
export TMPDIR=$PWD/tmp
echo $RUNLOGDIR
sbatch --partition=long  --output=$RUNLOGDIR/slurm.log --err=$RUNLOGDIR/slurm.log --time=13-11:59:00 nextflow -log $RUNLOGDIR/.nextflow.log run gladiator.nf -with-trace $RUNLOGDIR/trace.txt -with-report $RUNLOGDIR/report.html -with-dag $RUNLOGDIR/graph.dot --dda_assisted=false --fastafiles='/data/epouta1/B22003_Elixir_DIA_Nextflow/datasets/pxd014194/pxd014194/fasta/*.fasta' --diafiles='/data/epouta1/B22003_Elixir_DIA_Nextflow/datasets/pxd003497/DIA/MZML-pwiz/*.mzML'  --precursor_mass_tolerance=50 --fragment_mass_tolerance=0.1 --osw_use_cache=false -resume -c pxd000672.config 
</pre>

<pre class="example" id="org02270b1">
module use /appl/user/modulefiles/
module load nextflow
mkdir -p runs
RUNLOGDIR=`mktemp -p runs -d run.XXXX`
export TMPDIR=$PWD/tmp
echo $RUNLOGDIR
sbatch --output=$RUNLOGDIR/slurm.log --err=$RUNLOGDIR/slurm.logo nextflow -log $RUNLOGDIR/.nextflow.log run gladiator.nf -with-trace $RUNLOGDIR/trace.txt -with-report $RUNLOGDIR/report.html -with-dag $RUNLOGDIR/graph.dot --dda_assisted=false --fastafiles='/data/epouta1/B22003_Elixir_DIA_Nextflow/datasets/pxd014194/pxd014194/fasta/*.fasta' --diafiles='/data/epouta1/B22003_Elixir_DIA_Nextflow/datasets/bruderer/DIA/MZML-pwiz/B_D140314_SGSDSsample1_R01_MHRM_T0.mzML' --protFDR=0.01  --precursor_mass_tolerance=10 --fragment_mass_tolerance=0.02 -- 
</pre>



<div class="org-src-container">
<pre class="src src-emacs-lisp">(<span class="org-keyword">let</span> ((default-directory (concat (file-remote-p default-directory)
                                   <span class="org-string">"/data/epouta1/B22003_Elixir_DIA_Nextflow/playground/gladiator-nf"</span>))
        <span class="org-comment-delimiter">;; </span><span class="org-comment">dont ask about saving things</span>
        (compilation-save-buffers-predicate 'ignore))
  (compile nextflow-compile-command))

</pre>
</div>
</div>
</div>

<div id="outline-container-org99e0bbc" class="outline-2">
<h2 id="org99e0bbc"><span class="section-number-2">12.</span> Troubleshooting</h2>
<div class="outline-text-2" id="text-12">
<ul class="org-ul">
<li><a href="#org928fcbb">Podman specific Issues</a></li>
<li><a href="#orgeb39ba4">KeyError: 'HOME' in <code>pyprophet_subsample</code></a></li>
</ul>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="date">Created: 2025-07-03 Thu 11:25</p>
<p class="validation"><a href="https://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
